// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/clob/matches.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MatchPerpetualDeleveraging_Fill = exports.MatchPerpetualDeleveraging = exports.MatchPerpetualLiquidation = exports.MatchOrders = exports.MakerFill = exports.ClobMatch = void 0;
const order_1 = require("./order");
const subaccount_1 = require("../subaccounts/subaccount");
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../helpers");
function createBaseClobMatch() {
    return {
        matchOrders: undefined,
        matchPerpetualLiquidation: undefined,
        matchPerpetualDeleveraging: undefined
    };
}
exports.ClobMatch = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.matchOrders !== undefined) {
            exports.MatchOrders.encode(message.matchOrders, writer.uint32(10).fork()).ldelim();
        }
        if (message.matchPerpetualLiquidation !== undefined) {
            exports.MatchPerpetualLiquidation.encode(message.matchPerpetualLiquidation, writer.uint32(18).fork()).ldelim();
        }
        if (message.matchPerpetualDeleveraging !== undefined) {
            exports.MatchPerpetualDeleveraging.encode(message.matchPerpetualDeleveraging, writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseClobMatch();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.matchOrders = exports.MatchOrders.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.matchPerpetualLiquidation = exports.MatchPerpetualLiquidation.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.matchPerpetualDeleveraging = exports.MatchPerpetualDeleveraging.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseClobMatch();
        message.matchOrders = object.matchOrders !== undefined && object.matchOrders !== null ? exports.MatchOrders.fromPartial(object.matchOrders) : undefined;
        message.matchPerpetualLiquidation = object.matchPerpetualLiquidation !== undefined && object.matchPerpetualLiquidation !== null ? exports.MatchPerpetualLiquidation.fromPartial(object.matchPerpetualLiquidation) : undefined;
        message.matchPerpetualDeleveraging = object.matchPerpetualDeleveraging !== undefined && object.matchPerpetualDeleveraging !== null ? exports.MatchPerpetualDeleveraging.fromPartial(object.matchPerpetualDeleveraging) : undefined;
        return message;
    }
};
function createBaseMakerFill() {
    return {
        fillAmount: helpers_1.Long.UZERO,
        makerOrderId: undefined
    };
}
exports.MakerFill = {
    encode(message, writer = _m0.Writer.create()) {
        if (!message.fillAmount.isZero()) {
            writer.uint32(8).uint64(message.fillAmount);
        }
        if (message.makerOrderId !== undefined) {
            order_1.OrderId.encode(message.makerOrderId, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMakerFill();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.fillAmount = reader.uint64();
                    break;
                case 2:
                    message.makerOrderId = order_1.OrderId.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseMakerFill();
        message.fillAmount = object.fillAmount !== undefined && object.fillAmount !== null ? helpers_1.Long.fromValue(object.fillAmount) : helpers_1.Long.UZERO;
        message.makerOrderId = object.makerOrderId !== undefined && object.makerOrderId !== null ? order_1.OrderId.fromPartial(object.makerOrderId) : undefined;
        return message;
    }
};
function createBaseMatchOrders() {
    return {
        takerOrderId: undefined,
        fills: []
    };
}
exports.MatchOrders = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.takerOrderId !== undefined) {
            order_1.OrderId.encode(message.takerOrderId, writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.fills) {
            exports.MakerFill.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMatchOrders();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.takerOrderId = order_1.OrderId.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.fills.push(exports.MakerFill.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMatchOrders();
        message.takerOrderId = object.takerOrderId !== undefined && object.takerOrderId !== null ? order_1.OrderId.fromPartial(object.takerOrderId) : undefined;
        message.fills = ((_a = object.fills) === null || _a === void 0 ? void 0 : _a.map(e => exports.MakerFill.fromPartial(e))) || [];
        return message;
    }
};
function createBaseMatchPerpetualLiquidation() {
    return {
        liquidated: undefined,
        clobPairId: 0,
        perpetualId: 0,
        totalSize: helpers_1.Long.UZERO,
        isBuy: false,
        fills: []
    };
}
exports.MatchPerpetualLiquidation = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.liquidated !== undefined) {
            subaccount_1.SubaccountId.encode(message.liquidated, writer.uint32(10).fork()).ldelim();
        }
        if (message.clobPairId !== 0) {
            writer.uint32(16).uint32(message.clobPairId);
        }
        if (message.perpetualId !== 0) {
            writer.uint32(24).uint32(message.perpetualId);
        }
        if (!message.totalSize.isZero()) {
            writer.uint32(32).uint64(message.totalSize);
        }
        if (message.isBuy === true) {
            writer.uint32(40).bool(message.isBuy);
        }
        for (const v of message.fills) {
            exports.MakerFill.encode(v, writer.uint32(50).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMatchPerpetualLiquidation();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.liquidated = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.clobPairId = reader.uint32();
                    break;
                case 3:
                    message.perpetualId = reader.uint32();
                    break;
                case 4:
                    message.totalSize = reader.uint64();
                    break;
                case 5:
                    message.isBuy = reader.bool();
                    break;
                case 6:
                    message.fills.push(exports.MakerFill.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c, _d;
        const message = createBaseMatchPerpetualLiquidation();
        message.liquidated = object.liquidated !== undefined && object.liquidated !== null ? subaccount_1.SubaccountId.fromPartial(object.liquidated) : undefined;
        message.clobPairId = (_a = object.clobPairId) !== null && _a !== void 0 ? _a : 0;
        message.perpetualId = (_b = object.perpetualId) !== null && _b !== void 0 ? _b : 0;
        message.totalSize = object.totalSize !== undefined && object.totalSize !== null ? helpers_1.Long.fromValue(object.totalSize) : helpers_1.Long.UZERO;
        message.isBuy = (_c = object.isBuy) !== null && _c !== void 0 ? _c : false;
        message.fills = ((_d = object.fills) === null || _d === void 0 ? void 0 : _d.map(e => exports.MakerFill.fromPartial(e))) || [];
        return message;
    }
};
function createBaseMatchPerpetualDeleveraging() {
    return {
        liquidated: undefined,
        perpetualId: 0,
        fills: [],
        isFinalSettlement: false
    };
}
exports.MatchPerpetualDeleveraging = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.liquidated !== undefined) {
            subaccount_1.SubaccountId.encode(message.liquidated, writer.uint32(10).fork()).ldelim();
        }
        if (message.perpetualId !== 0) {
            writer.uint32(16).uint32(message.perpetualId);
        }
        for (const v of message.fills) {
            exports.MatchPerpetualDeleveraging_Fill.encode(v, writer.uint32(26).fork()).ldelim();
        }
        if (message.isFinalSettlement === true) {
            writer.uint32(32).bool(message.isFinalSettlement);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMatchPerpetualDeleveraging();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.liquidated = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.perpetualId = reader.uint32();
                    break;
                case 3:
                    message.fills.push(exports.MatchPerpetualDeleveraging_Fill.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.isFinalSettlement = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseMatchPerpetualDeleveraging();
        message.liquidated = object.liquidated !== undefined && object.liquidated !== null ? subaccount_1.SubaccountId.fromPartial(object.liquidated) : undefined;
        message.perpetualId = (_a = object.perpetualId) !== null && _a !== void 0 ? _a : 0;
        message.fills = ((_b = object.fills) === null || _b === void 0 ? void 0 : _b.map(e => exports.MatchPerpetualDeleveraging_Fill.fromPartial(e))) || [];
        message.isFinalSettlement = (_c = object.isFinalSettlement) !== null && _c !== void 0 ? _c : false;
        return message;
    }
};
function createBaseMatchPerpetualDeleveraging_Fill() {
    return {
        offsettingSubaccountId: undefined,
        fillAmount: helpers_1.Long.UZERO
    };
}
exports.MatchPerpetualDeleveraging_Fill = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.offsettingSubaccountId !== undefined) {
            subaccount_1.SubaccountId.encode(message.offsettingSubaccountId, writer.uint32(10).fork()).ldelim();
        }
        if (!message.fillAmount.isZero()) {
            writer.uint32(16).uint64(message.fillAmount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMatchPerpetualDeleveraging_Fill();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.offsettingSubaccountId = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.fillAmount = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseMatchPerpetualDeleveraging_Fill();
        message.offsettingSubaccountId = object.offsettingSubaccountId !== undefined && object.offsettingSubaccountId !== null ? subaccount_1.SubaccountId.fromPartial(object.offsettingSubaccountId) : undefined;
        message.fillAmount = object.fillAmount !== undefined && object.fillAmount !== null ? helpers_1.Long.fromValue(object.fillAmount) : helpers_1.Long.UZERO;
        return message;
    }
};
