// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/daemons/pricefeed/price_feed.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarketPriceUpdate = exports.ExchangePrice = exports.UpdateMarketPricesResponse = exports.UpdateMarketPricesRequest = void 0;
const timestamp_1 = require("../../../google/protobuf/timestamp");
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../../helpers");
function createBaseUpdateMarketPricesRequest() {
    return {
        marketPriceUpdates: []
    };
}
exports.UpdateMarketPricesRequest = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.marketPriceUpdates) {
            exports.MarketPriceUpdate.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateMarketPricesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.marketPriceUpdates.push(exports.MarketPriceUpdate.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseUpdateMarketPricesRequest();
        message.marketPriceUpdates = ((_a = object.marketPriceUpdates) === null || _a === void 0 ? void 0 : _a.map(e => exports.MarketPriceUpdate.fromPartial(e))) || [];
        return message;
    }
};
function createBaseUpdateMarketPricesResponse() {
    return {};
}
exports.UpdateMarketPricesResponse = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUpdateMarketPricesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseUpdateMarketPricesResponse();
        return message;
    }
};
function createBaseExchangePrice() {
    return {
        exchangeId: "",
        price: helpers_1.Long.UZERO,
        lastUpdateTime: undefined
    };
}
exports.ExchangePrice = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.exchangeId !== "") {
            writer.uint32(10).string(message.exchangeId);
        }
        if (!message.price.isZero()) {
            writer.uint32(16).uint64(message.price);
        }
        if (message.lastUpdateTime !== undefined) {
            timestamp_1.Timestamp.encode((0, helpers_1.toTimestamp)(message.lastUpdateTime), writer.uint32(26).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseExchangePrice();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.exchangeId = reader.string();
                    break;
                case 2:
                    message.price = reader.uint64();
                    break;
                case 3:
                    message.lastUpdateTime = (0, helpers_1.fromTimestamp)(timestamp_1.Timestamp.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseExchangePrice();
        message.exchangeId = (_a = object.exchangeId) !== null && _a !== void 0 ? _a : "";
        message.price = object.price !== undefined && object.price !== null ? helpers_1.Long.fromValue(object.price) : helpers_1.Long.UZERO;
        message.lastUpdateTime = (_b = object.lastUpdateTime) !== null && _b !== void 0 ? _b : undefined;
        return message;
    }
};
function createBaseMarketPriceUpdate() {
    return {
        marketId: 0,
        exchangePrices: []
    };
}
exports.MarketPriceUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.marketId !== 0) {
            writer.uint32(8).uint32(message.marketId);
        }
        for (const v of message.exchangePrices) {
            exports.ExchangePrice.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMarketPriceUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.marketId = reader.uint32();
                    break;
                case 2:
                    message.exchangePrices.push(exports.ExchangePrice.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMarketPriceUpdate();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : 0;
        message.exchangePrices = ((_b = object.exchangePrices) === null || _b === void 0 ? void 0 : _b.map(e => exports.ExchangePrice.fromPartial(e))) || [];
        return message;
    }
};
