// @dydxprotocol-indexer/postgres@18.19.39/build/src/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiquidityTierDoesNotExistError = exports.MarketDoesNotExistError = exports.PerpetualDoesNotExistError = exports.InvalidClobPairStatusError = exports.ValidationError = exports.RequiredFieldMissing = void 0;
/**
 * Custom error types.
 */
/**
 * Base class for custom errors.
 */
class CustomError extends Error {
    constructor(message) {
        super(message);
        // Set a more specific name. This will show up in e.g. console.log.
        this.name = this.constructor.name;
    }
}
class RequiredFieldMissing extends CustomError {
    constructor(field) {
        super(`Required field '${field}' is missing`);
    }
}
exports.RequiredFieldMissing = RequiredFieldMissing;
class ValidationError extends CustomError {
}
exports.ValidationError = ValidationError;
/**
 * Custom errors for converting protocol events into SQL.
 */
class InvalidClobPairStatusError extends Error {
    constructor(clobPairStatus) {
        super(`Invalid clob pair status: ${clobPairStatus}`);
        this.name = 'InvalidClobPairStatusError';
    }
}
exports.InvalidClobPairStatusError = InvalidClobPairStatusError;
class PerpetualDoesNotExistError extends Error {
    constructor(perpetualId, clobPairId) {
        super(`Perpetual with id ${perpetualId} does not exist. Referenced in clob pair with id ` +
            `${clobPairId}.`);
        this.name = 'PerpetualDoesNotExistError';
    }
}
exports.PerpetualDoesNotExistError = PerpetualDoesNotExistError;
class MarketDoesNotExistError extends Error {
    constructor(marketId, perpetualId) {
        super(`Market with id ${marketId} does not exist. Referenced in perpetual with id ` +
            `${perpetualId}.`);
        this.name = 'MarketDoesNotExistError';
    }
}
exports.MarketDoesNotExistError = MarketDoesNotExistError;
class LiquidityTierDoesNotExistError extends Error {
    constructor(liquidityTier, perpetualId) {
        super(`Liquidity tier with id ${liquidityTier} does not exist. Referenced in perpetual with id ` +
            `${perpetualId}.`);
        this.name = 'LiquidityTierDoesNotExistError';
    }
}
exports.LiquidityTierDoesNotExistError = LiquidityTierDoesNotExistError;
