// @cosmwasm/ts-codegen@0.21.1/main/utils/parse.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parser = void 0;

var _traverse = _interopRequireDefault(require("@babel/traverse"));

var _parser = require("@babel/parser");

var parser = function parser(codes) {
  var hash = {};
  codes.forEach(function (code) {
    var plugins = ['typescript'];
    var ast = (0, _parser.parse)(code, {
      sourceType: 'module',
      plugins: plugins
    });
    var visitor = visitorFn({
      addType: function addType(key, node) {
        hash[key] = node;
      }
    });
    (0, _traverse["default"])(ast, visitor);
  });
  return hash;
};

exports.parser = parser;

var visitorFn = function visitorFn(parser) {
  return {
    TSTypeAliasDeclaration: function TSTypeAliasDeclaration(path) {
      parser.addType(path.node.id.name, path.parentPath.node); // if (path.node.id.name.endsWith('For_Empty')) {
      //     const newName = path.node.id.name.replace(/For_Empty$/, '_for_Empty');
      //     path.parentPath.node.declaration.id.name = newName;
      //     parser.addType(newName, path.parentPath.node);
      // } else {
      //     parser.addType(path.node.id.name, path.parentPath.node);
      // }
    },
    TSInterfaceDeclaration: function TSInterfaceDeclaration(path) {
      parser.addType(path.node.id.name, path.parentPath.node);
    }
  };
};
