// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/epochs/query.lcd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LCDQueryClient = void 0;
const helpers_1 = require("../../helpers");
class LCDQueryClient {
    constructor({ requestClient }) {
        this.req = requestClient;
        this.epochInfo = this.epochInfo.bind(this);
        this.epochInfoAll = this.epochInfoAll.bind(this);
    }
    /* Queries a EpochInfo by name. */
    async epochInfo(params) {
        const endpoint = `dydxprotocol/v4/epochs/epoch_info/${params.name}`;
        return await this.req.get(endpoint);
    }
    /* Queries a list of EpochInfo items. */
    async epochInfoAll(params = {
        pagination: undefined
    }) {
        const options = {
            params: {}
        };
        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
            (0, helpers_1.setPaginationParams)(options, params.pagination);
        }
        const endpoint = `dydxprotocol/v4/epochs/epoch_info`;
        return await this.req.get(endpoint, options);
    }
}
exports.LCDQueryClient = LCDQueryClient;
