// @dydxprotocol-indexer/postgres@18.19.39/build/__tests__/stores/oracle-price-table.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../src/types");
const OraclePriceTable = __importStar(require("../../src/stores/oracle-price-table"));
const db_helpers_1 = require("../../src/helpers/db-helpers");
const mock_generators_1 = require("../helpers/mock-generators");
const constants_1 = require("../helpers/constants");
const BlockTable = __importStar(require("../../src/stores/block-table"));
const luxon_1 = require("luxon");
describe('Oracle price store', () => {
    const updatedHeight = '5';
    beforeEach(async () => {
        await (0, mock_generators_1.seedData)();
        await BlockTable.create({
            ...constants_1.defaultBlock,
            blockHeight: updatedHeight,
        });
    });
    beforeAll(async () => {
        await (0, db_helpers_1.migrate)();
    });
    afterEach(async () => {
        await (0, db_helpers_1.clearData)();
    });
    afterAll(async () => {
        await (0, db_helpers_1.teardown)();
    });
    it('Successfully creates an oracle price', async () => {
        await OraclePriceTable.create(constants_1.defaultOraclePrice);
    });
    it('Successfully creates multiple oracle prices', async () => {
        const oraclePrice2 = {
            marketId: constants_1.defaultMarket.id,
            price: '10000.05',
            effectiveAt: constants_1.createdDateTime.toISO(),
            effectiveAtHeight: updatedHeight,
        };
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create(oraclePrice2),
        ]);
        const oraclePrices = await OraclePriceTable.findAll({
            marketId: [constants_1.defaultMarket.id],
        }, [], {
            orderBy: [[types_1.OraclePriceColumns.effectiveAtHeight, types_1.Ordering.ASC]],
        });
        expect(oraclePrices.length).toEqual(2);
        expect(oraclePrices[0]).toEqual(expect.objectContaining(constants_1.defaultOraclePrice));
        expect(oraclePrices[1]).toEqual(expect.objectContaining(oraclePrice2));
    });
    it('Successfully finds all OraclePrices', async () => {
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create({
                ...constants_1.defaultOraclePrice,
                effectiveAtHeight: updatedHeight,
            }),
        ]);
        const oraclePrices = await OraclePriceTable.findAll({
            marketId: [constants_1.defaultMarket.id],
        }, [], {
            orderBy: [[types_1.OraclePriceColumns.effectiveAtHeight, types_1.Ordering.ASC]],
        });
        expect(oraclePrices.length).toEqual(2);
        expect(oraclePrices[0]).toEqual(expect.objectContaining(constants_1.defaultOraclePrice));
        expect(oraclePrices[1]).toEqual(expect.objectContaining({
            ...constants_1.defaultOraclePrice,
            effectiveAtHeight: updatedHeight,
        }));
    });
    it('Successfully finds OraclePrice with effectiveAtHeight', async () => {
        await OraclePriceTable.create(constants_1.defaultOraclePrice);
        const oraclePrices = await OraclePriceTable.findAll({
            effectiveAtHeight: constants_1.defaultOraclePrice.effectiveAtHeight,
        }, [], { readReplica: true });
        expect(oraclePrices.length).toEqual(1);
        expect(oraclePrices[0]).toEqual(expect.objectContaining({
            ...constants_1.defaultOraclePrice,
        }));
    });
    it('Successfully finds all OraclePrices effective before or after the height', async () => {
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create({
                ...constants_1.defaultOraclePrice,
                effectiveAtHeight: updatedHeight,
            }),
        ]);
        const oraclePrices = await OraclePriceTable.findAll({
            marketId: [constants_1.defaultMarket.id],
            effectiveBeforeOrAtHeight: constants_1.defaultOraclePrice.effectiveAtHeight,
        }, [], {});
        expect(oraclePrices.length).toEqual(1);
        expect(oraclePrices[0]).toEqual(expect.objectContaining(constants_1.defaultOraclePrice));
    });
    it('Successfully finds all OraclePrices effective before or after time', async () => {
        const oraclePrice2 = {
            ...constants_1.defaultOraclePrice,
            effectiveAtHeight: updatedHeight,
            effectiveAt: '1982-05-25T00:00:00.000Z',
        };
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create(oraclePrice2),
        ]);
        const oraclePrices = await OraclePriceTable.findAll({
            marketId: [constants_1.defaultMarket.id],
            effectiveBeforeOrAt: '2000-05-25T00:00:00.000Z',
        }, [], {});
        expect(oraclePrices.length).toEqual(1);
        expect(oraclePrices[0]).toEqual(expect.objectContaining(oraclePrice2));
    });
    it('Successfully finds an OraclePrice', async () => {
        await OraclePriceTable.create(constants_1.defaultOraclePrice);
        const oraclePrice = await OraclePriceTable.findById(constants_1.defaultOraclePriceId);
        expect(oraclePrice).toEqual(expect.objectContaining(constants_1.defaultOraclePrice));
    });
    it('Successfully finds oracle prices in reverse chronological order by market id', async () => {
        const oraclePrice2 = {
            ...constants_1.defaultOraclePrice,
            effectiveAtHeight: updatedHeight,
        };
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create(oraclePrice2),
        ]);
        const oraclePrices = await OraclePriceTable
            .findOraclePricesInReverseChronologicalOrder(constants_1.defaultMarket.id);
        expect(oraclePrices.length).toEqual(2);
        expect(oraclePrices[0]).toEqual(expect.objectContaining(oraclePrice2));
        expect(oraclePrices[1]).toEqual(expect.objectContaining(constants_1.defaultOraclePrice));
    });
    it('Successfully finds latest oracle price for market id', async () => {
        const oraclePrice2 = {
            ...constants_1.defaultOraclePrice,
            effectiveAtHeight: updatedHeight,
            effectiveAt: '1982-05-25T00:00:00.000Z',
        };
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create(oraclePrice2),
        ]);
        const oraclePrice = await OraclePriceTable
            .findMostRecentMarketOraclePrice(constants_1.defaultMarket.id);
        expect(oraclePrice).toEqual(expect.objectContaining(oraclePrice2));
    });
    it('Successfully finds latest prices by effectiveAtHeight', async () => {
        const oraclePrice2 = {
            ...constants_1.defaultOraclePrice,
            price: '10000.05',
            effectiveAtHeight: updatedHeight,
            effectiveAt: '1982-05-25T00:00:00.000Z',
        };
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create(oraclePrice2),
            OraclePriceTable.create(constants_1.defaultOraclePrice2),
        ]);
        const oraclePrices = await OraclePriceTable
            .findLatestPrices(updatedHeight);
        expect(oraclePrices).toEqual(expect.objectContaining({
            [constants_1.defaultOraclePrice.marketId]: oraclePrice2.price,
            [constants_1.defaultOraclePrice2.marketId]: constants_1.defaultOraclePrice2.price,
        }));
    });
    it('Successfully finds the latest price and the price 24h ago', async () => {
        const now = luxon_1.DateTime.local().toISO();
        const lessThan24HAgo = luxon_1.DateTime.local().minus({ hour: 23 }).toISO();
        const moreThan24HAgo = luxon_1.DateTime.local().minus({ hour: 24, minute: 5 }).toISO();
        const wayMoreThan24HAgo = luxon_1.DateTime.local().minus({ hour: 25 }).toISO();
        const oraclePrice3 = {
            ...constants_1.defaultOraclePrice,
            price: '3',
            effectiveAtHeight: '3',
            effectiveAt: lessThan24HAgo,
        };
        const oraclePrice4 = {
            ...constants_1.defaultOraclePrice,
            price: '4',
            effectiveAtHeight: '4',
            effectiveAt: moreThan24HAgo,
        };
        const oraclePrice5 = {
            ...constants_1.defaultOraclePrice,
            price: '5',
            effectiveAtHeight: '5',
            effectiveAt: wayMoreThan24HAgo,
        };
        const oraclePrice6 = {
            ...constants_1.defaultOraclePrice,
            price: '6',
            effectiveAtHeight: '6',
            effectiveAt: now,
        };
        const oraclePrice7 = {
            ...constants_1.defaultOraclePrice,
            marketId: constants_1.defaultMarket2.id,
            price: '7',
            effectiveAtHeight: '7',
            effectiveAt: lessThan24HAgo,
        };
        const blockHeights = ['3', '4', '6', '7'];
        const blockPromises = blockHeights.map((height) => BlockTable.create({
            ...constants_1.defaultBlock,
            blockHeight: height,
        }));
        await Promise.all(blockPromises);
        await Promise.all([
            OraclePriceTable.create(oraclePrice3),
            OraclePriceTable.create(oraclePrice4),
            OraclePriceTable.create(oraclePrice5),
            OraclePriceTable.create(oraclePrice6),
            OraclePriceTable.create(oraclePrice7),
        ]);
        const oraclePricesFrom24hAgo = await OraclePriceTable
            .getPricesFrom24hAgo();
        expect(oraclePricesFrom24hAgo).toEqual(expect.objectContaining({
            [constants_1.defaultOraclePrice.marketId]: oraclePrice4.price,
        }));
        const latestPrices = await OraclePriceTable
            .getLatestPrices();
        expect(latestPrices).toEqual(expect.objectContaining({
            [constants_1.defaultOraclePrice.marketId]: oraclePrice6.price,
            [constants_1.defaultMarket2.id]: oraclePrice7.price,
        }));
    });
    it('Successfully finds latest prices respecting effectiveAtHeight', async () => {
        const oraclePrice2 = {
            ...constants_1.defaultOraclePrice,
            price: '10000.05',
            effectiveAtHeight: updatedHeight,
            effectiveAt: '1982-05-25T00:00:00.000Z',
        };
        await Promise.all([
            OraclePriceTable.create(constants_1.defaultOraclePrice),
            OraclePriceTable.create(oraclePrice2),
        ]);
        const oraclePrices = await OraclePriceTable
            .findLatestPrices(constants_1.defaultOraclePrice.effectiveAtHeight);
        expect(oraclePrices).toEqual(expect.objectContaining({
            [constants_1.defaultOraclePrice.marketId]: constants_1.defaultOraclePrice.price,
        }));
    });
});
