// @dydxprotocol-indexer/postgres@18.19.39/build/__tests__/stores/perpetual-market-table.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../src/types");
const PerpetualMarketTable = __importStar(require("../../src/stores/perpetual-market-table"));
const LiquidityTiersTable = __importStar(require("../../src/stores/liquidity-tiers-table"));
const db_helpers_1 = require("../../src/helpers/db-helpers");
const constants_1 = require("../helpers/constants");
const MarketTable = __importStar(require("../../src/stores/market-table"));
describe('PerpetualMarket store', () => {
    beforeAll(async () => {
        await (0, db_helpers_1.migrate)();
    });
    beforeEach(async () => {
        await Promise.all([
            MarketTable.create(constants_1.defaultMarket),
            MarketTable.create(constants_1.defaultMarket2),
        ]);
        await Promise.all([
            LiquidityTiersTable.create(constants_1.defaultLiquidityTier),
            LiquidityTiersTable.create(constants_1.defaultLiquidityTier2),
        ]);
    });
    afterEach(async () => {
        await (0, db_helpers_1.clearData)();
    });
    afterAll(async () => {
        await (0, db_helpers_1.teardown)();
    });
    it('Successfully creates a PerpetualMarket', async () => {
        await PerpetualMarketTable.create(constants_1.defaultPerpetualMarket);
    });
    it('Successfully finds all PerpetualMarkets', async () => {
        await Promise.all([
            PerpetualMarketTable.create(constants_1.defaultPerpetualMarket),
            PerpetualMarketTable.create({
                ...constants_1.defaultPerpetualMarket,
                id: '1',
            }),
        ]);
        const perpetualMarkets = await PerpetualMarketTable.findAll({}, [], { readReplica: true });
        expect(perpetualMarkets.length).toEqual(2);
        expect(perpetualMarkets[0]).toEqual(expect.objectContaining(constants_1.defaultPerpetualMarket));
        expect(perpetualMarkets[1]).toEqual(expect.objectContaining({
            ...constants_1.defaultPerpetualMarket,
            id: '1',
        }));
    });
    it('Successfully finds a PerpetualMarket', async () => {
        await PerpetualMarketTable.create(constants_1.defaultPerpetualMarket);
        const perpetualMarket = await PerpetualMarketTable
            .findById('0');
        expect(perpetualMarket).toEqual(expect.objectContaining(constants_1.defaultPerpetualMarket));
    });
    it('Successfully finds a PerpetualMarket by market id', async () => {
        await PerpetualMarketTable.create(constants_1.defaultPerpetualMarket);
        const perpetualMarket = await PerpetualMarketTable
            .findByMarketId(constants_1.defaultMarket.id);
        expect(perpetualMarket).toEqual(expect.objectContaining(constants_1.defaultPerpetualMarket));
    });
    it('Successfully finds a PerpetualMarket by clob pair id', async () => {
        await PerpetualMarketTable.create(constants_1.defaultPerpetualMarket);
        const perpetualMarket = await PerpetualMarketTable
            .findByClobPairId('1');
        expect(perpetualMarket).toEqual(expect.objectContaining(constants_1.defaultPerpetualMarket));
    });
    it('Unable finds a PerpetualMarket', async () => {
        const perpetualMarket = await PerpetualMarketTable
            .findById('0');
        expect(perpetualMarket).toEqual(undefined);
    });
    it('Unable finds a PerpetualMarket by market id', async () => {
        const perpetualMarket = await PerpetualMarketTable
            .findByMarketId(constants_1.defaultMarket.id);
        expect(perpetualMarket).toEqual(undefined);
    });
    it.each([
        ['market with ticker exists', constants_1.defaultPerpetualMarket.ticker, constants_1.defaultPerpetualMarket],
        ['market with ticker does not exist', constants_1.invalidTicker, undefined],
    ])('Finds a PerpetualMarket by ticker: %s', async (_name, ticker, expectedPerpetualMarket) => {
        await PerpetualMarketTable.create(constants_1.defaultPerpetualMarket);
        const perpetualMarket = await PerpetualMarketTable
            .findByTicker(ticker);
        if (expectedPerpetualMarket !== undefined) {
            expect(perpetualMarket).toEqual(expect.objectContaining(constants_1.defaultPerpetualMarket));
        }
        else {
            expect(perpetualMarket).toEqual(expectedPerpetualMarket);
        }
    });
    it('Successfully updates a perpetual market', async () => {
        await PerpetualMarketTable.create(constants_1.defaultPerpetualMarket);
        const perpetualMarket = await PerpetualMarketTable
            .update({
            id: constants_1.defaultPerpetualMarket.id,
            trades24H: 100,
        });
        expect(perpetualMarket).toEqual(expect.objectContaining({
            ...constants_1.defaultPerpetualMarket,
            trades24H: 100,
        }));
    });
    it('Successfully winds down a perpetual market', async () => {
        await PerpetualMarketTable.create(constants_1.defaultPerpetualMarket);
        const perpetualMarket = await PerpetualMarketTable
            .update({
            id: constants_1.defaultPerpetualMarket.id,
            status: types_1.PerpetualMarketStatus.FINAL_SETTLEMENT,
        });
        expect(perpetualMarket).toEqual(expect.objectContaining({
            ...constants_1.defaultPerpetualMarket,
            status: types_1.PerpetualMarketStatus.FINAL_SETTLEMENT,
        }));
    });
    it('Successfully updates a perpetual market by market id', async () => {
        const market = {
            id: 5,
            pair: 'DYDX-USD',
            exponent: -5,
            minPriceChangePpm: 50,
        };
        await MarketTable.create(market);
        await PerpetualMarketTable.create({
            ...constants_1.defaultPerpetualMarket,
            marketId: 5,
        });
        const perpetualMarket = await PerpetualMarketTable
            .updateByMarketId({
            marketId: 5,
            trades24H: 100,
        });
        expect(perpetualMarket).toEqual(expect.objectContaining({
            ...constants_1.defaultPerpetualMarket,
            marketId: 5,
            trades24H: 100,
        }));
    });
});
