// @dydxprotocol-indexer/postgres@18.19.39/build/__tests__/stores/pnl-ticks-table.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../src/types");
const PnlTicksTable = __importStar(require("../../src/stores/pnl-ticks-table"));
const BlockTable = __importStar(require("../../src/stores/block-table"));
const db_helpers_1 = require("../../src/helpers/db-helpers");
const mock_generators_1 = require("../helpers/mock-generators");
const constants_1 = require("../helpers/constants");
const luxon_1 = require("luxon");
const constants_2 = require("../../src/constants");
describe('PnlTicks store', () => {
    beforeEach(async () => {
        await (0, mock_generators_1.seedData)();
    });
    beforeAll(async () => {
        await (0, db_helpers_1.migrate)();
    });
    afterEach(async () => {
        await (0, db_helpers_1.clearData)();
    });
    afterAll(async () => {
        await (0, db_helpers_1.teardown)();
    });
    it('Successfully creates a Pnl Tick', async () => {
        await PnlTicksTable.create(constants_1.defaultPnlTick);
    });
    it('Successfully creates multiple Pnl Ticks', async () => {
        await BlockTable.create({
            ...constants_1.defaultBlock,
            blockHeight: '5',
        });
        const pnlTick2 = {
            subaccountId: constants_1.defaultSubaccountId,
            equity: '5',
            totalPnl: '5',
            netTransfers: '5',
            createdAt: '2020-01-01T00:00:00.000Z',
            blockHeight: '5',
            blockTime: constants_1.defaultBlock.time,
        };
        await Promise.all([
            PnlTicksTable.create(constants_1.defaultPnlTick),
            PnlTicksTable.create(pnlTick2),
        ]);
        const pnlTicks = await PnlTicksTable.findAll({}, [], {
            orderBy: [[types_1.PnlTicksColumns.blockHeight, types_1.Ordering.ASC]],
        });
        expect(pnlTicks.length).toEqual(2);
        expect(pnlTicks[0]).toEqual(expect.objectContaining(constants_1.defaultPnlTick));
        expect(pnlTicks[1]).toEqual(expect.objectContaining(pnlTick2));
    });
    it('createMany Pnl Ticks', async () => {
        const pnlTick2 = {
            subaccountId: constants_1.defaultSubaccountId,
            equity: '5',
            totalPnl: '5',
            netTransfers: '5',
            createdAt: '2020-01-01T00:00:00.000Z',
            blockHeight: '5',
            blockTime: constants_1.defaultBlock.time,
        };
        await PnlTicksTable.createMany([constants_1.defaultPnlTick, pnlTick2]);
        const pnlTicks = await PnlTicksTable.findAll({}, [], {
            orderBy: [[types_1.PnlTicksColumns.blockHeight, types_1.Ordering.ASC]],
        });
        expect(pnlTicks.length).toEqual(2);
        expect(pnlTicks[0]).toEqual(expect.objectContaining(constants_1.defaultPnlTick));
        expect(pnlTicks[1]).toEqual(expect.objectContaining(pnlTick2));
    });
    it('Successfully finds PnlTicks with subaccountId', async () => {
        await Promise.all([
            PnlTicksTable.create(constants_1.defaultPnlTick),
            PnlTicksTable.create({
                ...constants_1.defaultPnlTick,
                createdAt: '2020-01-01T00:00:00.000Z',
            }),
            PnlTicksTable.create({
                ...constants_1.defaultPnlTick,
                subaccountId: constants_1.defaultSubaccountId2,
                createdAt: '2020-01-01T00:00:00.000Z',
            }),
        ]);
        const pnlTicks = await PnlTicksTable.findAll({
            subaccountId: [constants_1.defaultSubaccountId],
        }, [], { readReplica: true });
        expect(pnlTicks.length).toEqual(2);
    });
    it('Successfully finds latest block time', async () => {
        const blockTime = '2023-01-01T00:00:00.000Z';
        await Promise.all([
            PnlTicksTable.create(constants_1.defaultPnlTick),
            PnlTicksTable.create({
                ...constants_1.defaultPnlTick,
                createdAt: '2020-01-01T00:00:00.000Z',
                blockHeight: '1000',
                blockTime,
            }),
        ]);
        const latestBlocktime = await PnlTicksTable.findLatestProcessedBlocktime();
        expect(latestBlocktime).toEqual(blockTime);
    });
    it('Successfully finds latest block time without any pnl ticks', async () => {
        const latestBlocktime = await PnlTicksTable.findLatestProcessedBlocktime();
        expect(latestBlocktime).toEqual(constants_2.ZERO_TIME_ISO_8601);
    });
    it('createMany PnlTicks, find most recent pnl ticks for each account', async () => {
        await Promise.all([
            BlockTable.create({
                blockHeight: '3',
                time: constants_1.defaultBlock.time,
            }),
            BlockTable.create({
                blockHeight: '5',
                time: constants_1.defaultBlock.time,
            }),
        ]);
        await PnlTicksTable.createMany([
            {
                subaccountId: constants_1.defaultSubaccountId,
                equity: '1092',
                createdAt: luxon_1.DateTime.utc().minus({ hours: 1 }).toISO(),
                totalPnl: '1000',
                netTransfers: '50',
                blockHeight: constants_1.defaultBlock.blockHeight,
                blockTime: constants_1.defaultBlock.time,
            },
            {
                subaccountId: constants_1.defaultSubaccountId,
                equity: '1097',
                createdAt: luxon_1.DateTime.utc().minus({ hours: 3 }).toISO(),
                totalPnl: '1000',
                netTransfers: '50',
                blockHeight: '3',
                blockTime: constants_1.defaultBlock.time,
            },
            {
                subaccountId: constants_1.defaultSubaccountId,
                equity: '1011',
                createdAt: luxon_1.DateTime.utc().minus({ hours: 11 }).toISO(),
                totalPnl: '1000',
                netTransfers: '50',
                blockHeight: '5',
                blockTime: constants_1.defaultBlock.time,
            },
            {
                subaccountId: constants_1.defaultSubaccountId,
                equity: '1014',
                createdAt: luxon_1.DateTime.utc().minus({ hours: 9 }).toISO(),
                totalPnl: '1000',
                netTransfers: '50',
                blockHeight: '5',
                blockTime: constants_1.defaultBlock.time,
            },
            {
                subaccountId: constants_1.defaultSubaccountId2,
                equity: '100',
                createdAt: new Date().toISOString(),
                totalPnl: '1000',
                netTransfers: '50',
                blockHeight: '2',
                blockTime: constants_1.defaultBlock2.time,
            },
            {
                subaccountId: constants_1.defaultSubaccountId2,
                equity: '200',
                createdAt: luxon_1.DateTime.utc().minus({ hours: 9 }).toISO(),
                totalPnl: '1000',
                netTransfers: '50',
                blockHeight: '5',
                blockTime: constants_1.defaultBlock.time,
            },
        ]);
        const mostRecent = await PnlTicksTable.findMostRecentPnlTickForEachAccount('3');
        expect(mostRecent[constants_1.defaultSubaccountId].equity).toEqual('1014');
        expect(mostRecent[constants_1.defaultSubaccountId2].equity).toEqual('200');
    });
});
