// @dydxprotocol/v4-proto@3.0.0-dev.0/src/codegen/cosmos/bundle.ts
import * as _2 from "./app/runtime/v1alpha1/module";
import * as _3 from "./app/v1alpha1/config";
import * as _4 from "./app/v1alpha1/module";
import * as _5 from "./app/v1alpha1/query";
import * as _6 from "./auth/module/v1/module";
import * as _7 from "./auth/v1beta1/auth";
import * as _8 from "./auth/v1beta1/genesis";
import * as _9 from "./auth/v1beta1/query";
import * as _10 from "./auth/v1beta1/tx";
import * as _11 from "./authz/module/v1/module";
import * as _12 from "./authz/v1beta1/authz";
import * as _13 from "./authz/v1beta1/event";
import * as _14 from "./authz/v1beta1/genesis";
import * as _15 from "./authz/v1beta1/query";
import * as _16 from "./authz/v1beta1/tx";
import * as _17 from "./bank/module/v1/module";
import * as _18 from "./bank/v1beta1/authz";
import * as _19 from "./bank/v1beta1/bank";
import * as _20 from "./bank/v1beta1/genesis";
import * as _21 from "./bank/v1beta1/query";
import * as _22 from "./bank/v1beta1/tx";
import * as _23 from "./base/abci/v1beta1/abci";
import * as _24 from "./base/node/v1beta1/query";
import * as _25 from "./base/query/v1beta1/pagination";
import * as _26 from "./base/reflection/v1beta1/reflection";
import * as _27 from "./base/reflection/v2alpha1/reflection";
import * as _28 from "./base/tendermint/v1beta1/query";
import * as _29 from "./base/tendermint/v1beta1/types";
import * as _30 from "./base/v1beta1/coin";
import * as _31 from "./circuit/module/v1/module";
import * as _32 from "./circuit/v1/query";
import * as _33 from "./circuit/v1/tx";
import * as _34 from "./circuit/v1/types";
import * as _35 from "./consensus/module/v1/module";
import * as _36 from "./consensus/v1/query";
import * as _37 from "./consensus/v1/tx";
import * as _38 from "./crisis/module/v1/module";
import * as _39 from "./crisis/v1beta1/genesis";
import * as _40 from "./crisis/v1beta1/tx";
import * as _41 from "./crypto/ed25519/keys";
import * as _42 from "./crypto/hd/v1/hd";
import * as _43 from "./crypto/keyring/v1/record";
import * as _44 from "./crypto/multisig/keys";
import * as _45 from "./crypto/secp256k1/keys";
import * as _46 from "./crypto/secp256r1/keys";
import * as _47 from "./distribution/module/v1/module";
import * as _48 from "./distribution/v1beta1/distribution";
import * as _49 from "./distribution/v1beta1/genesis";
import * as _50 from "./distribution/v1beta1/query";
import * as _51 from "./distribution/v1beta1/tx";
import * as _52 from "./evidence/module/v1/module";
import * as _53 from "./evidence/v1beta1/evidence";
import * as _54 from "./evidence/v1beta1/genesis";
import * as _55 from "./evidence/v1beta1/query";
import * as _56 from "./evidence/v1beta1/tx";
import * as _57 from "./feegrant/module/v1/module";
import * as _58 from "./feegrant/v1beta1/feegrant";
import * as _59 from "./feegrant/v1beta1/genesis";
import * as _60 from "./feegrant/v1beta1/query";
import * as _61 from "./feegrant/v1beta1/tx";
import * as _62 from "./genutil/module/v1/module";
import * as _63 from "./genutil/v1beta1/genesis";
import * as _64 from "./gov/module/v1/module";
import * as _65 from "./gov/v1/genesis";
import * as _66 from "./gov/v1/gov";
import * as _67 from "./gov/v1/query";
import * as _68 from "./gov/v1/tx";
import * as _69 from "./gov/v1beta1/genesis";
import * as _70 from "./gov/v1beta1/gov";
import * as _71 from "./gov/v1beta1/query";
import * as _72 from "./gov/v1beta1/tx";
import * as _73 from "./group/module/v1/module";
import * as _74 from "./group/v1/events";
import * as _75 from "./group/v1/genesis";
import * as _76 from "./group/v1/query";
import * as _77 from "./group/v1/tx";
import * as _78 from "./group/v1/types";
import * as _79 from "./mint/module/v1/module";
import * as _80 from "./mint/v1beta1/genesis";
import * as _81 from "./mint/v1beta1/mint";
import * as _82 from "./mint/v1beta1/query";
import * as _83 from "./mint/v1beta1/tx";
import * as _84 from "./msg/textual/v1/textual";
import * as _85 from "./msg/v1/msg";
import * as _86 from "./nft/module/v1/module";
import * as _87 from "./nft/v1beta1/event";
import * as _88 from "./nft/v1beta1/genesis";
import * as _89 from "./nft/v1beta1/nft";
import * as _90 from "./nft/v1beta1/query";
import * as _91 from "./nft/v1beta1/tx";
import * as _92 from "./orm/module/v1alpha1/module";
import * as _93 from "./orm/query/v1alpha1/query";
import * as _94 from "./orm/v1/orm";
import * as _95 from "./orm/v1alpha1/schema";
import * as _96 from "./params/module/v1/module";
import * as _97 from "./params/v1beta1/params";
import * as _98 from "./params/v1beta1/query";
import * as _99 from "./query/v1/query";
import * as _100 from "./reflection/v1/reflection";
import * as _101 from "./slashing/module/v1/module";
import * as _102 from "./slashing/v1beta1/genesis";
import * as _103 from "./slashing/v1beta1/query";
import * as _104 from "./slashing/v1beta1/slashing";
import * as _105 from "./slashing/v1beta1/tx";
import * as _106 from "./staking/module/v1/module";
import * as _107 from "./staking/v1beta1/authz";
import * as _108 from "./staking/v1beta1/genesis";
import * as _109 from "./staking/v1beta1/query";
import * as _110 from "./staking/v1beta1/staking";
import * as _111 from "./staking/v1beta1/tx";
import * as _112 from "./store/internal/kv/v1beta1/kv";
import * as _113 from "./store/snapshots/v1/snapshot";
import * as _114 from "./store/streaming/abci/grpc";
import * as _115 from "./store/v1beta1/commit_info";
import * as _116 from "./store/v1beta1/listening";
import * as _117 from "./tx/config/v1/config";
import * as _118 from "./tx/signing/v1beta1/signing";
import * as _119 from "./tx/v1beta1/service";
import * as _120 from "./tx/v1beta1/tx";
import * as _121 from "./upgrade/module/v1/module";
import * as _122 from "./upgrade/v1beta1/query";
import * as _123 from "./upgrade/v1beta1/tx";
import * as _124 from "./upgrade/v1beta1/upgrade";
import * as _125 from "./vesting/module/v1/module";
import * as _126 from "./vesting/v1beta1/tx";
import * as _127 from "./vesting/v1beta1/vesting";
import * as _236 from "./auth/v1beta1/query.lcd";
import * as _237 from "./authz/v1beta1/query.lcd";
import * as _238 from "./bank/v1beta1/query.lcd";
import * as _239 from "./base/node/v1beta1/query.lcd";
import * as _240 from "./base/tendermint/v1beta1/query.lcd";
import * as _241 from "./circuit/v1/query.lcd";
import * as _242 from "./consensus/v1/query.lcd";
import * as _243 from "./distribution/v1beta1/query.lcd";
import * as _244 from "./evidence/v1beta1/query.lcd";
import * as _245 from "./feegrant/v1beta1/query.lcd";
import * as _246 from "./gov/v1/query.lcd";
import * as _247 from "./gov/v1beta1/query.lcd";
import * as _248 from "./group/v1/query.lcd";
import * as _249 from "./mint/v1beta1/query.lcd";
import * as _250 from "./nft/v1beta1/query.lcd";
import * as _251 from "./params/v1beta1/query.lcd";
import * as _252 from "./slashing/v1beta1/query.lcd";
import * as _253 from "./staking/v1beta1/query.lcd";
import * as _254 from "./tx/v1beta1/service.lcd";
import * as _255 from "./upgrade/v1beta1/query.lcd";
import * as _256 from "./app/v1alpha1/query.rpc.Query";
import * as _257 from "./auth/v1beta1/query.rpc.Query";
import * as _258 from "./authz/v1beta1/query.rpc.Query";
import * as _259 from "./bank/v1beta1/query.rpc.Query";
import * as _260 from "./base/node/v1beta1/query.rpc.Service";
import * as _261 from "./base/tendermint/v1beta1/query.rpc.Service";
import * as _262 from "./circuit/v1/query.rpc.Query";
import * as _263 from "./consensus/v1/query.rpc.Query";
import * as _264 from "./distribution/v1beta1/query.rpc.Query";
import * as _265 from "./evidence/v1beta1/query.rpc.Query";
import * as _266 from "./feegrant/v1beta1/query.rpc.Query";
import * as _267 from "./gov/v1/query.rpc.Query";
import * as _268 from "./gov/v1beta1/query.rpc.Query";
import * as _269 from "./group/v1/query.rpc.Query";
import * as _270 from "./mint/v1beta1/query.rpc.Query";
import * as _271 from "./nft/v1beta1/query.rpc.Query";
import * as _272 from "./orm/query/v1alpha1/query.rpc.Query";
import * as _273 from "./params/v1beta1/query.rpc.Query";
import * as _274 from "./slashing/v1beta1/query.rpc.Query";
import * as _275 from "./staking/v1beta1/query.rpc.Query";
import * as _276 from "./tx/v1beta1/service.rpc.Service";
import * as _277 from "./upgrade/v1beta1/query.rpc.Query";
import * as _278 from "./auth/v1beta1/tx.rpc.msg";
import * as _279 from "./authz/v1beta1/tx.rpc.msg";
import * as _280 from "./bank/v1beta1/tx.rpc.msg";
import * as _281 from "./circuit/v1/tx.rpc.msg";
import * as _282 from "./consensus/v1/tx.rpc.msg";
import * as _283 from "./crisis/v1beta1/tx.rpc.msg";
import * as _284 from "./distribution/v1beta1/tx.rpc.msg";
import * as _285 from "./evidence/v1beta1/tx.rpc.msg";
import * as _286 from "./feegrant/v1beta1/tx.rpc.msg";
import * as _287 from "./gov/v1/tx.rpc.msg";
import * as _288 from "./gov/v1beta1/tx.rpc.msg";
import * as _289 from "./group/v1/tx.rpc.msg";
import * as _290 from "./mint/v1beta1/tx.rpc.msg";
import * as _291 from "./nft/v1beta1/tx.rpc.msg";
import * as _292 from "./slashing/v1beta1/tx.rpc.msg";
import * as _293 from "./staking/v1beta1/tx.rpc.msg";
import * as _294 from "./upgrade/v1beta1/tx.rpc.msg";
import * as _295 from "./vesting/v1beta1/tx.rpc.msg";
import * as _337 from "./lcd";
import * as _338 from "./rpc.query";
import * as _339 from "./rpc.tx";
export namespace cosmos {
  export namespace app {
    export namespace runtime {
      export const v1alpha1 = { ..._2
      };
    }
    export const v1alpha1 = { ..._3,
      ..._4,
      ..._5,
      ..._256
    };
  }
  export namespace auth {
    export namespace module {
      export const v1 = { ..._6
      };
    }
    export const v1beta1 = { ..._7,
      ..._8,
      ..._9,
      ..._10,
      ..._236,
      ..._257,
      ..._278
    };
  }
  export namespace authz {
    export namespace module {
      export const v1 = { ..._11
      };
    }
    export const v1beta1 = { ..._12,
      ..._13,
      ..._14,
      ..._15,
      ..._16,
      ..._237,
      ..._258,
      ..._279
    };
  }
  export namespace bank {
    export namespace module {
      export const v1 = { ..._17
      };
    }
    export const v1beta1 = { ..._18,
      ..._19,
      ..._20,
      ..._21,
      ..._22,
      ..._238,
      ..._259,
      ..._280
    };
  }
  export namespace base {
    export namespace abci {
      export const v1beta1 = { ..._23
      };
    }
    export namespace node {
      export const v1beta1 = { ..._24,
        ..._239,
        ..._260
      };
    }
    export namespace query {
      export const v1beta1 = { ..._25
      };
    }
    export namespace reflection {
      export const v1beta1 = { ..._26
      };
      export const v2alpha1 = { ..._27
      };
    }
    export namespace tendermint {
      export const v1beta1 = { ..._28,
        ..._29,
        ..._240,
        ..._261
      };
    }
    export const v1beta1 = { ..._30
    };
  }
  export namespace circuit {
    export namespace module {
      export const v1 = { ..._31
      };
    }
    export const v1 = { ..._32,
      ..._33,
      ..._34,
      ..._241,
      ..._262,
      ..._281
    };
  }
  export namespace consensus {
    export namespace module {
      export const v1 = { ..._35
      };
    }
    export const v1 = { ..._36,
      ..._37,
      ..._242,
      ..._263,
      ..._282
    };
  }
  export namespace crisis {
    export namespace module {
      export const v1 = { ..._38
      };
    }
    export const v1beta1 = { ..._39,
      ..._40,
      ..._283
    };
  }
  export namespace crypto {
    export const ed25519 = { ..._41
    };
    export namespace hd {
      export const v1 = { ..._42
      };
    }
    export namespace keyring {
      export const v1 = { ..._43
      };
    }
    export const multisig = { ..._44
    };
    export const secp256k1 = { ..._45
    };
    export const secp256r1 = { ..._46
    };
  }
  export namespace distribution {
    export namespace module {
      export const v1 = { ..._47
      };
    }
    export const v1beta1 = { ..._48,
      ..._49,
      ..._50,
      ..._51,
      ..._243,
      ..._264,
      ..._284
    };
  }
  export namespace evidence {
    export namespace module {
      export const v1 = { ..._52
      };
    }
    export const v1beta1 = { ..._53,
      ..._54,
      ..._55,
      ..._56,
      ..._244,
      ..._265,
      ..._285
    };
  }
  export namespace feegrant {
    export namespace module {
      export const v1 = { ..._57
      };
    }
    export const v1beta1 = { ..._58,
      ..._59,
      ..._60,
      ..._61,
      ..._245,
      ..._266,
      ..._286
    };
  }
  export namespace genutil {
    export namespace module {
      export const v1 = { ..._62
      };
    }
    export const v1beta1 = { ..._63
    };
  }
  export namespace gov {
    export namespace module {
      export const v1 = { ..._64
      };
    }
    export const v1 = { ..._65,
      ..._66,
      ..._67,
      ..._68,
      ..._246,
      ..._267,
      ..._287
    };
    export const v1beta1 = { ..._69,
      ..._70,
      ..._71,
      ..._72,
      ..._247,
      ..._268,
      ..._288
    };
  }
  export namespace group {
    export namespace module {
      export const v1 = { ..._73
      };
    }
    export const v1 = { ..._74,
      ..._75,
      ..._76,
      ..._77,
      ..._78,
      ..._248,
      ..._269,
      ..._289
    };
  }
  export namespace mint {
    export namespace module {
      export const v1 = { ..._79
      };
    }
    export const v1beta1 = { ..._80,
      ..._81,
      ..._82,
      ..._83,
      ..._249,
      ..._270,
      ..._290
    };
  }
  export namespace msg {
    export namespace textual {
      export const v1 = { ..._84
      };
    }
    export const v1 = { ..._85
    };
  }
  export namespace nft {
    export namespace module {
      export const v1 = { ..._86
      };
    }
    export const v1beta1 = { ..._87,
      ..._88,
      ..._89,
      ..._90,
      ..._91,
      ..._250,
      ..._271,
      ..._291
    };
  }
  export namespace orm {
    export namespace module {
      export const v1alpha1 = { ..._92
      };
    }
    export namespace query {
      export const v1alpha1 = { ..._93,
        ..._272
      };
    }
    export const v1 = { ..._94
    };
    export const v1alpha1 = { ..._95
    };
  }
  export namespace params {
    export namespace module {
      export const v1 = { ..._96
      };
    }
    export const v1beta1 = { ..._97,
      ..._98,
      ..._251,
      ..._273
    };
  }
  export namespace query {
    export const v1 = { ..._99
    };
  }
  export namespace reflection {
    export const v1 = { ..._100
    };
  }
  export namespace slashing {
    export namespace module {
      export const v1 = { ..._101
      };
    }
    export const v1beta1 = { ..._102,
      ..._103,
      ..._104,
      ..._105,
      ..._252,
      ..._274,
      ..._292
    };
  }
  export namespace staking {
    export namespace module {
      export const v1 = { ..._106
      };
    }
    export const v1beta1 = { ..._107,
      ..._108,
      ..._109,
      ..._110,
      ..._111,
      ..._253,
      ..._275,
      ..._293
    };
  }
  export namespace store {
    export namespace internal {
      export namespace kv {
        export const v1beta1 = { ..._112
        };
      }
    }
    export namespace snapshots {
      export const v1 = { ..._113
      };
    }
    export namespace streaming {
      export const abci = { ..._114
      };
    }
    export const v1beta1 = { ..._115,
      ..._116
    };
  }
  export namespace tx {
    export namespace config {
      export const v1 = { ..._117
      };
    }
    export namespace signing {
      export const v1beta1 = { ..._118
      };
    }
    export const v1beta1 = { ..._119,
      ..._120,
      ..._254,
      ..._276
    };
  }
  export namespace upgrade {
    export namespace module {
      export const v1 = { ..._121
      };
    }
    export const v1beta1 = { ..._122,
      ..._123,
      ..._124,
      ..._255,
      ..._277,
      ..._294
    };
  }
  export namespace vesting {
    export namespace module {
      export const v1 = { ..._125
      };
    }
    export const v1beta1 = { ..._126,
      ..._127,
      ..._295
    };
  }
  export const ClientFactory = { ..._337,
    ..._338,
    ..._339
  };
}
