// @dydxprotocol-indexer/postgres@18.19.39/build/src/helpers/db-helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.teardown = exports.migrate = exports.reset = exports.clearSchema = exports.clearData = exports.createModelToJsonFunctions = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const config_1 = __importDefault(require("../config"));
const constants_1 = require("../constants");
const knex_1 = require("./knex");
const stores_helpers_1 = require("./stores-helpers");
const layer2Tables = [
    'perpetual_positions',
    'fills',
];
const layer1Tables = [
    'subaccounts',
    'markets',
    'orders',
    'perpetual_markets',
    'tendermint_events',
    'transactions',
    'blocks',
    'assets',
    'candles',
    'liquidity_tiers',
    'compliance_data',
];
/**
 * Returns the SQL statement that would convert the provided field and type to the type expected
 * by the model.
 *
 * Raises an error if an unknown conversion is requested.
 */
function getSqlConversionForDydxModelTypes(fieldName, type) {
    switch (type) {
        case 'integer':
            return `row_t."${fieldName}"::int`;
        case 'string':
            return `row_t."${fieldName}"::text`;
        case 'boolean':
            return `row_t."${fieldName}"::bool`;
        case 'hex-string':
            return `encode(row_t."${fieldName}", 'hex')`;
        case 'date-time':
            return `to_char(row_t."${fieldName}" at time zone 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS.MS"Z"')`;
        default:
            throw new Error(`Unknown type conversion for ${type}`);
    }
}
/**
 * Defines a `dydx_to_jsonb` function for each of the models in SQL_TO_JSON_DEFINED_MODELS and
 * loads them in Postgres. This allows for plpgsql functions to invoke `dydx_to_jsonb` on the
 * associated models table row type and convert the record into a JSON representation which
 * conforms to the models schema allowing conversion to the model type via the models `fromJson`
 * method.
 */
async function createModelToJsonFunctions() {
    await Promise.all(constants_1.SQL_TO_JSON_DEFINED_MODELS.map(async (model) => {
        const sqlProperties = Object.entries(model.sqlToJsonConversions)
            .map(([key, value]) => `'${key}', ${getSqlConversionForDydxModelTypes(key, value)}`);
        const sqlFn = `CREATE OR REPLACE FUNCTION dydx_to_jsonb(row_t ${model.tableName}) RETURNS jsonb AS $$
BEGIN
    RETURN jsonb_build_object(
        ${sqlProperties.join(',\n        ')}
        );
    END;
$$ LANGUAGE plpgsql IMMUTABLE PARALLEL SAFE;`;
        return (0, stores_helpers_1.rawQuery)(sqlFn, {}).catch((error) => {
            base_1.logger.error({
                at: 'dbHelpers#createModelToJsonFunctions',
                message: `Failed to create or replace function dydx_to_jsonb for model ${model.tableName}.`,
                error,
            });
            throw error;
        });
    }));
}
exports.createModelToJsonFunctions = createModelToJsonFunctions;
async function dropData() {
    await Promise.all(layer2Tables.map(async (table) => {
        return (0, knex_1.knexPrimary)(table).del();
    }));
    // need to use for... of to ensure tables are removed sequentially
    for (const table of layer1Tables) {
        await (0, knex_1.knexPrimary)(table).del();
    }
}
/**
 * Drops all functions named dydx_.* from the database.
 */
async function dropAllDydxFunctions() {
    await knex_1.knexPrimary.raw(`DO
$do$
DECLARE
   _sql text;
BEGIN
   SELECT INTO _sql
          string_agg(format('DROP %s %s;'
                          , CASE prokind
                              WHEN 'f' THEN 'FUNCTION'
                              WHEN 'a' THEN 'AGGREGATE'
                              WHEN 'p' THEN 'PROCEDURE'
                              WHEN 'w' THEN 'FUNCTION'  -- window function (rarely applicable)
                              -- ELSE NULL              -- not possible in pg 11
                            END
                          , oid::regprocedure)
                   , E'\\n')
   FROM   pg_proc
   WHERE  pronamespace = 'public'::regnamespace  -- schema name here!
   AND proname LIKE 'dydx_%';

   IF _sql IS NOT NULL THEN
      EXECUTE _sql;         -- uncomment payload once you are sure
   ELSE
      RAISE NOTICE 'No fuctions found in schema %', quote_ident('public');
   END IF;
END
$do$;`);
}
async function clearData() {
    for (const table of layer1Tables) {
        const tableExists = await knex_1.knexPrimary.schema.hasTable(table);
        if (tableExists) {
            await knex_1.knexPrimary.raw(`truncate table "${table}" cascade`);
        }
    }
    await Promise.all(layer2Tables.map(async (table) => {
        const tableExists = await knex_1.knexPrimary.schema.hasTable(table);
        if (tableExists) {
            return knex_1.knexPrimary.raw(`truncate table "${table}" cascade`);
        }
    }));
}
exports.clearData = clearData;
async function clearSchema() {
    await knex_1.knexPrimary.schema.raw('DROP SCHEMA public CASCADE');
    await knex_1.knexPrimary.schema.raw('CREATE SCHEMA public');
}
exports.clearSchema = clearSchema;
async function reset() {
    await dropData();
    await dropAllDydxFunctions();
    await rollback();
}
exports.reset = reset;
async function rollback() {
    await knex_1.knexPrimary.migrate.rollback({ loadExtensions: ['.js'] });
}
async function migrate() {
    return knex_1.knexPrimary.migrate.latest({ loadExtensions: ['.js'] });
}
exports.migrate = migrate;
async function teardown() {
    await dropAllDydxFunctions();
    await knex_1.knexPrimary.destroy();
    if (config_1.default.IS_USING_DB_READONLY) {
        await knex_1.knexReadReplica.getConnection().destroy();
    }
}
exports.teardown = teardown;
