// @dydxprotocol-indexer/postgres@18.19.39/build/src/helpers/stores-helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateBulkUpsertString = exports.generateBulkUpdateString = exports.setBulkRowsForUpdate = exports.rawQuery = exports.setupBaseQuery = exports.verifyAllInjectableVariables = exports.verifyAllRequiredFields = void 0;
const lodash_1 = __importDefault(require("lodash"));
const errors_1 = require("../lib/errors");
const knex_1 = require("./knex");
const transaction_1 = __importDefault(require("./transaction"));
function verifyAllRequiredFields(query, requiredFields) {
    const queryKeys = Object.keys(lodash_1.default.omitBy(query, lodash_1.default.isNil));
    requiredFields.forEach((field) => {
        if (!queryKeys.includes(field)) {
            throw new errors_1.RequiredFieldMissing(field);
        }
    });
}
exports.verifyAllRequiredFields = verifyAllRequiredFields;
function verifyAllInjectableVariables(vals) {
    vals.forEach((val) => {
        // Numbers and undefined are okay.
        if (val === undefined ||
            val === null ||
            typeof val === 'number') {
            return;
        }
        // Prevent possible query injections.
        if (typeof val === 'string' &&
            val.split(' ').length > 1) {
            throw Error(`Invalid value: ${val} could be a malicious query injection`);
        }
    });
}
exports.verifyAllInjectableVariables = verifyAllInjectableVariables;
function setupBaseQuery(model, options) {
    if (options.readReplica) {
        return model.bindKnex(knex_1.knexReadReplica.getConnection()).query(transaction_1.default.get(options.txId));
    }
    else {
        return model.query(transaction_1.default.get(options.txId));
    }
}
exports.setupBaseQuery = setupBaseQuery;
async function rawQuery(queryString, options) {
    const connection = options.readReplica ? knex_1.knexReadReplica.getConnection() : knex_1.knexPrimary;
    let queryBuilder = options.bindings === undefined
        ? connection.raw(queryString) : connection.raw(queryString, options.bindings);
    if (options.txId) {
        queryBuilder = queryBuilder.transacting(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        transaction_1.default.get(options.txId));
    }
    if (options.sqlOptions) {
        queryBuilder = queryBuilder.options(options.sqlOptions);
    }
    return queryBuilder;
}
exports.rawQuery = rawQuery;
/* ------- Bulk Helpers ------- */
function setBulkRowsForUpdate({ objectArray, columns, stringColumns, numericColumns, bigintColumns, timestampColumns, uuidColumns, booleanColumns, binaryColumns, }) {
    return objectArray.map((object) => columns.map((col) => {
        if (stringColumns && stringColumns.includes(col)) {
            return `'${castNull(object[col])}'`;
        }
        if (numericColumns && numericColumns.includes(col)) {
            return `${castNull(lodash_1.default.get(object, col))}`;
        }
        if (bigintColumns && bigintColumns.includes(col)) {
            return castValue(object[col], 'bigint');
        }
        if (timestampColumns && timestampColumns.includes(col)) {
            return castValue(object[col], 'timestamp');
        }
        if (uuidColumns && uuidColumns.includes(col)) {
            return castValue(object[col], 'uuid');
        }
        if (binaryColumns && binaryColumns.includes(col)) {
            return castBinaryValue(object[col]);
        }
        if (booleanColumns && booleanColumns.includes(col)) {
            return `${castNull(object[col])}`;
        }
        throw new Error(`Unsupported column for bulk update: ${col}`);
    }).join(', '));
}
exports.setBulkRowsForUpdate = setBulkRowsForUpdate;
/**
 * If the value is null || undefined, return 'NULL'
 */
function castNull(value) {
    if (value === null || value === undefined) {
        return 'NULL';
    }
    return `${value}`;
}
/**
 * If the value is null || undefined, return 'NULL' casted with typesuffix, otherwise return
 * the stringified value in quotes casted with typesuffix.
 */
function castValue(value, typeSuffix) {
    if (value === null || value === undefined) {
        return `NULL::${typeSuffix}`;
    }
    return `'${value}'::${typeSuffix}`;
}
function castBinaryValue(value) {
    const typeSuffix = 'bytea';
    if (value === null || value === undefined) {
        return `NULL::${typeSuffix}`;
    }
    return `'\\x${value.toString('hex')}'::${typeSuffix}`;
}
function generateBulkUpdateString({ table, objectRows, columns, isUuid, uniqueIdentifier = 'id', setFieldsToAppend, }) {
    const columnsToUpdate = lodash_1.default.without(columns, uniqueIdentifier);
    const setFields = columnsToUpdate.map((col) => {
        return `"${col}" = c."${col}"`;
    }).concat(setFieldsToAppend || []);
    return `
  UPDATE "${table}" SET
    ${setFields.join(',')}
  FROM (VALUES
    ${objectRows.map((object) => `(${object})`).join(', ')}
  ) AS c(${columns.map((c) => `"${c}"`).join(', ')})
  WHERE c."${uniqueIdentifier}"${isUuid ? '::uuid' : ''} = "${table}"."${uniqueIdentifier}";
`;
}
exports.generateBulkUpdateString = generateBulkUpdateString;
function generateBulkUpsertString({ table, objectRows, columns, uniqueIdentifiers = ['id'], }) {
    const columnsToUpdate = lodash_1.default.without(columns, ...uniqueIdentifiers);
    const idFields = uniqueIdentifiers.map((id) => { return `"${id}"`; }).join(',');
    const insertFields = columns.map((column) => { return `"${column}"`; }).join(',');
    const setFields = columnsToUpdate.map((col) => {
        return `"${col}" = excluded."${col}"`;
    });
    return `
  INSERT INTO "${table}" (${insertFields}) VALUES
    ${objectRows.map((object) => `(${object})`).join(',')}
  ON CONFLICT (${idFields}) DO UPDATE SET ${setFields.join(',')};
  `;
}
exports.generateBulkUpsertString = generateBulkUpsertString;
