// @dydxprotocol/v4-proto@3.0.0-dev.0/src/codegen/dydxprotocol/bundle.ts
import * as _128 from "./assets/asset";
import * as _129 from "./assets/genesis";
import * as _130 from "./assets/query";
import * as _131 from "./assets/tx";
import * as _132 from "./blocktime/blocktime";
import * as _133 from "./blocktime/genesis";
import * as _134 from "./blocktime/params";
import * as _135 from "./blocktime/query";
import * as _136 from "./blocktime/tx";
import * as _137 from "./bridge/bridge_event_info";
import * as _138 from "./bridge/bridge_event";
import * as _139 from "./bridge/genesis";
import * as _140 from "./bridge/params";
import * as _141 from "./bridge/query";
import * as _142 from "./bridge/tx";
import * as _143 from "./clob/block_rate_limit_config";
import * as _144 from "./clob/clob_pair";
import * as _145 from "./clob/equity_tier_limit_config";
import * as _146 from "./clob/genesis";
import * as _147 from "./clob/liquidations_config";
import * as _148 from "./clob/liquidations";
import * as _149 from "./clob/matches";
import * as _150 from "./clob/mev";
import * as _151 from "./clob/operation";
import * as _152 from "./clob/order_removals";
import * as _153 from "./clob/order";
import * as _154 from "./clob/process_proposer_matches_events";
import * as _155 from "./clob/query";
import * as _156 from "./clob/tx";
import * as _157 from "./daemons/bridge/bridge";
import * as _158 from "./daemons/liquidation/liquidation";
import * as _159 from "./daemons/pricefeed/price_feed";
import * as _160 from "./delaymsg/block_message_ids";
import * as _161 from "./delaymsg/delayed_message";
import * as _162 from "./delaymsg/genesis";
import * as _163 from "./delaymsg/query";
import * as _164 from "./delaymsg/tx";
import * as _165 from "./epochs/epoch_info";
import * as _166 from "./epochs/genesis";
import * as _167 from "./epochs/query";
import * as _168 from "./feetiers/genesis";
import * as _169 from "./feetiers/params";
import * as _170 from "./feetiers/query";
import * as _171 from "./feetiers/tx";
import * as _172 from "./indexer/events/events";
import * as _173 from "./indexer/indexer_manager/event";
import * as _174 from "./indexer/off_chain_updates/off_chain_updates";
import * as _175 from "./indexer/protocol/v1/clob";
import * as _176 from "./indexer/protocol/v1/subaccount";
import * as _177 from "./indexer/redis/redis_order";
import * as _178 from "./indexer/shared/removal_reason";
import * as _179 from "./indexer/socks/messages";
import * as _180 from "./perpetuals/genesis";
import * as _181 from "./perpetuals/params";
import * as _182 from "./perpetuals/perpetual";
import * as _183 from "./perpetuals/query";
import * as _184 from "./perpetuals/tx";
import * as _185 from "./prices/genesis";
import * as _186 from "./prices/market_param";
import * as _187 from "./prices/market_price";
import * as _188 from "./prices/query";
import * as _189 from "./prices/tx";
import * as _190 from "./ratelimit/capacity";
import * as _191 from "./ratelimit/genesis";
import * as _192 from "./ratelimit/limit_params";
import * as _193 from "./ratelimit/query";
import * as _194 from "./ratelimit/tx";
import * as _195 from "./rewards/genesis";
import * as _196 from "./rewards/params";
import * as _197 from "./rewards/query";
import * as _198 from "./rewards/reward_share";
import * as _199 from "./rewards/tx";
import * as _200 from "./sending/genesis";
import * as _201 from "./sending/query";
import * as _202 from "./sending/transfer";
import * as _203 from "./sending/tx";
import * as _204 from "./stats/genesis";
import * as _205 from "./stats/params";
import * as _206 from "./stats/query";
import * as _207 from "./stats/stats";
import * as _208 from "./stats/tx";
import * as _209 from "./subaccounts/asset_position";
import * as _210 from "./subaccounts/genesis";
import * as _211 from "./subaccounts/perpetual_position";
import * as _212 from "./subaccounts/query";
import * as _213 from "./subaccounts/subaccount";
import * as _214 from "./vest/genesis";
import * as _215 from "./vest/query";
import * as _216 from "./vest/tx";
import * as _217 from "./vest/vest_entry";
import * as _296 from "./assets/query.lcd";
import * as _297 from "./blocktime/query.lcd";
import * as _298 from "./bridge/query.lcd";
import * as _299 from "./clob/query.lcd";
import * as _300 from "./delaymsg/query.lcd";
import * as _301 from "./epochs/query.lcd";
import * as _302 from "./feetiers/query.lcd";
import * as _303 from "./perpetuals/query.lcd";
import * as _304 from "./prices/query.lcd";
import * as _305 from "./ratelimit/query.lcd";
import * as _306 from "./rewards/query.lcd";
import * as _307 from "./stats/query.lcd";
import * as _308 from "./subaccounts/query.lcd";
import * as _309 from "./vest/query.lcd";
import * as _310 from "./assets/query.rpc.Query";
import * as _311 from "./blocktime/query.rpc.Query";
import * as _312 from "./bridge/query.rpc.Query";
import * as _313 from "./clob/query.rpc.Query";
import * as _314 from "./delaymsg/query.rpc.Query";
import * as _315 from "./epochs/query.rpc.Query";
import * as _316 from "./feetiers/query.rpc.Query";
import * as _317 from "./perpetuals/query.rpc.Query";
import * as _318 from "./prices/query.rpc.Query";
import * as _319 from "./ratelimit/query.rpc.Query";
import * as _320 from "./rewards/query.rpc.Query";
import * as _321 from "./sending/query.rpc.Query";
import * as _322 from "./stats/query.rpc.Query";
import * as _323 from "./subaccounts/query.rpc.Query";
import * as _324 from "./vest/query.rpc.Query";
import * as _325 from "./blocktime/tx.rpc.msg";
import * as _326 from "./bridge/tx.rpc.msg";
import * as _327 from "./clob/tx.rpc.msg";
import * as _328 from "./delaymsg/tx.rpc.msg";
import * as _329 from "./feetiers/tx.rpc.msg";
import * as _330 from "./perpetuals/tx.rpc.msg";
import * as _331 from "./prices/tx.rpc.msg";
import * as _332 from "./ratelimit/tx.rpc.msg";
import * as _333 from "./rewards/tx.rpc.msg";
import * as _334 from "./sending/tx.rpc.msg";
import * as _335 from "./stats/tx.rpc.msg";
import * as _336 from "./vest/tx.rpc.msg";
import * as _340 from "./lcd";
import * as _341 from "./rpc.query";
import * as _342 from "./rpc.tx";
export namespace dydxprotocol {
  export const assets = { ..._128,
    ..._129,
    ..._130,
    ..._131,
    ..._296,
    ..._310
  };
  export const blocktime = { ..._132,
    ..._133,
    ..._134,
    ..._135,
    ..._136,
    ..._297,
    ..._311,
    ..._325
  };
  export const bridge = { ..._137,
    ..._138,
    ..._139,
    ..._140,
    ..._141,
    ..._142,
    ..._298,
    ..._312,
    ..._326
  };
  export const clob = { ..._143,
    ..._144,
    ..._145,
    ..._146,
    ..._147,
    ..._148,
    ..._149,
    ..._150,
    ..._151,
    ..._152,
    ..._153,
    ..._154,
    ..._155,
    ..._156,
    ..._299,
    ..._313,
    ..._327
  };
  export namespace daemons {
    export const bridge = { ..._157
    };
    export const liquidation = { ..._158
    };
    export const pricefeed = { ..._159
    };
  }
  export const delaymsg = { ..._160,
    ..._161,
    ..._162,
    ..._163,
    ..._164,
    ..._300,
    ..._314,
    ..._328
  };
  export const epochs = { ..._165,
    ..._166,
    ..._167,
    ..._301,
    ..._315
  };
  export const feetiers = { ..._168,
    ..._169,
    ..._170,
    ..._171,
    ..._302,
    ..._316,
    ..._329
  };
  export namespace indexer {
    export const events = { ..._172
    };
    export const indexer_manager = { ..._173
    };
    export const off_chain_updates = { ..._174
    };
    export namespace protocol {
      export const v1 = { ..._175,
        ..._176
      };
    }
    export const redis = { ..._177
    };
    export const shared = { ..._178
    };
    export const socks = { ..._179
    };
  }
  export const perpetuals = { ..._180,
    ..._181,
    ..._182,
    ..._183,
    ..._184,
    ..._303,
    ..._317,
    ..._330
  };
  export const prices = { ..._185,
    ..._186,
    ..._187,
    ..._188,
    ..._189,
    ..._304,
    ..._318,
    ..._331
  };
  export const ratelimit = { ..._190,
    ..._191,
    ..._192,
    ..._193,
    ..._194,
    ..._305,
    ..._319,
    ..._332
  };
  export const rewards = { ..._195,
    ..._196,
    ..._197,
    ..._198,
    ..._199,
    ..._306,
    ..._320,
    ..._333
  };
  export const sending = { ..._200,
    ..._201,
    ..._202,
    ..._203,
    ..._321,
    ..._334
  };
  export const stats = { ..._204,
    ..._205,
    ..._206,
    ..._207,
    ..._208,
    ..._307,
    ..._322,
    ..._335
  };
  export const subaccounts = { ..._209,
    ..._210,
    ..._211,
    ..._212,
    ..._213,
    ..._308,
    ..._323
  };
  export const vest = { ..._214,
    ..._215,
    ..._216,
    ..._217,
    ..._309,
    ..._324,
    ..._336
  };
  export const ClientFactory = { ..._340,
    ..._341,
    ..._342
  };
}
