// @dydxprotocol-indexer/kafka/build/src/batch-kafka-producer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchKafkaProducer = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const lodash_1 = __importDefault(require("lodash"));
/**
 * Groups messages for a single kafka topic into batches to send fewer ProducerRecords.
 */
class BatchKafkaProducer {
    constructor(topic, producer, 
    // Note that default parameters are bound during module load time making it difficult
    // to modify the parameter during a test so we explicitly require callers to pass in
    // config.KAFKA_MAX_BATCH_WEBSOCKET_MESSAGE_SIZE_BYTES.
    maxBatchSizeBytes) {
        this.maxBatchSizeBytes = maxBatchSizeBytes;
        this.producer = producer;
        this.topic = topic;
        this.producerMessages = [];
        this.producerPromises = [];
        this.currentSize = 0;
    }
    /**
     * Add a message to the current batch. If the message size would push the current batch size over
     * the maxBatchSizeBytes, the current batch (without this message) is flushed first, then the
     * message is added to a new batch.
     */
    addMessageAndMaybeFlush(message) {
        const keyByteLength = message.key === undefined ? 0 : message.key.byteLength;
        const msgBuffer = message.value;
        if (this.currentSize + msgBuffer.byteLength + keyByteLength > this.maxBatchSizeBytes) {
            this.sendBatch();
        }
        this.producerMessages.push({ key: message.key, value: msgBuffer });
        this.currentSize += msgBuffer.byteLength;
        this.currentSize += keyByteLength;
    }
    async flush() {
        this.sendBatch();
        // TODO(IND-198): Log an error when kafka producer fails
        return Promise.all(this.producerPromises);
    }
    sendBatch() {
        if (!lodash_1.default.isEmpty(this.producerMessages)) {
            this.producerPromises.push(this.producer.send({ topic: this.topic, messages: this.producerMessages }));
        }
        base_1.logger.info({
            at: 'BatchMessenger#sendBatch',
            message: 'Produced kafka batch',
            currentSize: this.currentSize,
            producerMessages: JSON.stringify(this.producerMessages),
            recalculatedCurrentSize: this.producerMessages.reduce((acc, msg) => acc + msg.value.byteLength, 0),
            topic: this.topic,
        });
        this.producerMessages = [];
        this.currentSize = 0;
    }
}
exports.BatchKafkaProducer = BatchKafkaProducer;
