// @dydxprotocol-indexer/kafka/build/src/config.js
"use strict";
/**
 * Environment variables required for Kafka.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.kafkaConfigSchema = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
exports.kafkaConfigSchema = {
    // Required to set groupId and clientId for kafka.
    SERVICE_NAME: (0, base_1.parseString)(),
    KAFKA_BROKER_URLS: (0, base_1.parseString)({
        default: 'localhost:9092',
        requireInEnv: [base_1.NodeEnv.PRODUCTION, base_1.NodeEnv.STAGING],
    }),
    KAFKA_CONNECTION_TIMEOUT_MS: (0, base_1.parseInteger)({ default: 5000 }),
    KAFKA_SESSION_TIMEOUT_MS: (0, base_1.parseInteger)({ default: 60000 }),
    KAFKA_REBALANCE_TIMEOUT_MS: (0, base_1.parseInteger)({ default: 50000 }),
    KAFKA_HEARTBEAT_INTERVAL_MS: (0, base_1.parseInteger)({ default: 5000 }),
    KAFKA_CONCURRENT_PARTITIONS: (0, base_1.parseInteger)({ default: 1 }),
    // If true, consumers will have unique group ids, and SERVICE_NAME will be a common prefix for
    // the consumer group ids.
    KAFKA_ENABLE_UNIQUE_CONSUMER_GROUP_IDS: (0, base_1.parseBoolean)({ default: false }),
    // Set to a number smaller than the max message size for the Kafka broker
    KAFKA_MAX_BATCH_WEBSOCKET_MESSAGE_SIZE_BYTES: (0, base_1.parseInteger)({
        default: 900000, // ~900 kB, 100 kB smaller than the 1 MB default max size of messages in Kafka
    }),
};
exports.default = (0, base_1.parseSchema)(exports.kafkaConfigSchema);
