// @dydxprotocol-indexer/kafka/build/src/consumer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startConsumer = exports.stopConsumer = exports.updateOnBatchFunction = exports.updateOnMessageFunction = exports.consumer = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const uuid_1 = require("uuid");
const config_1 = __importDefault(require("./config"));
const kafka_1 = require("./kafka");
const groupIdPrefix = config_1.default.SERVICE_NAME;
const groupIdSuffix = config_1.default.KAFKA_ENABLE_UNIQUE_CONSUMER_GROUP_IDS ? `_${(0, uuid_1.v4)()}` : '';
const groupId = `${groupIdPrefix}${groupIdSuffix}`;
exports.consumer = kafka_1.kafka.consumer({
    groupId,
    sessionTimeout: config_1.default.KAFKA_SESSION_TIMEOUT_MS,
    rebalanceTimeout: config_1.default.KAFKA_REBALANCE_TIMEOUT_MS,
    heartbeatInterval: config_1.default.KAFKA_HEARTBEAT_INTERVAL_MS,
    readUncommitted: false,
    maxBytes: 4194304, // 4MB
});
// List of functions to run per message consumed.
let onMessageFunction;
// List of function to be run per batch consumed.
let onBatchFunction;
/**
 * Overwrite function to be run on each kafka message
 * @param onMessage
 */
function updateOnMessageFunction(onMessage) {
    onMessageFunction = onMessage;
}
exports.updateOnMessageFunction = updateOnMessageFunction;
/**
 * Overwrite function to be run on each kafka batch
 */
function updateOnBatchFunction(onBatch) {
    onBatchFunction = onBatch;
}
exports.updateOnBatchFunction = updateOnBatchFunction;
// Whether the consumer is stopped.
let stopped = false;
exports.consumer.on('consumer.disconnect', async () => {
    base_1.logger.info({
        at: 'consumers#disconnect',
        message: 'Kafka consumer disconnected',
        groupId,
    });
    if (!stopped) {
        await exports.consumer.connect();
        base_1.logger.info({
            at: 'kafka-consumer#disconnect',
            message: 'Kafka consumer reconnected',
            groupId,
        });
    }
    else {
        base_1.logger.info({
            at: 'kafka-consumer#disconnect',
            message: 'Not reconnecting since task is shutting down',
            groupId,
        });
    }
});
async function stopConsumer() {
    base_1.logger.info({
        at: 'kafka-consumer#stop',
        message: 'Stopping kafka consumer',
        groupId,
    });
    stopped = true;
    await exports.consumer.disconnect();
}
exports.stopConsumer = stopConsumer;
async function startConsumer(batchProcessing = false) {
    const consumerRunConfig = {
        // The last offset of each batch will be committed if processing does not error.
        // The commit will still happen if the number of messages in the batch < autoCommitThreshold.
        eachBatchAutoResolve: true,
        partitionsConsumedConcurrently: config_1.default.KAFKA_CONCURRENT_PARTITIONS,
        autoCommit: true,
    };
    if (batchProcessing) {
        consumerRunConfig.eachBatch = onBatchFunction;
    }
    else {
        consumerRunConfig.eachMessage = async ({ topic, message }) => {
            await onMessageFunction(topic, message);
        };
    }
    await exports.consumer.run(consumerRunConfig);
    base_1.logger.info({
        at: 'consumers#connect',
        message: 'Started kafka consumer',
        groupId,
    });
}
exports.startConsumer = startConsumer;
