// @dydxprotocol-indexer/kafka/build/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KafkaTopics = exports.WebsocketTopics = void 0;
var WebsocketTopics;
(function (WebsocketTopics) {
    WebsocketTopics["TO_WEBSOCKETS_ORDERBOOKS"] = "to-websockets-orderbooks";
    WebsocketTopics["TO_WEBSOCKETS_SUBACCOUNTS"] = "to-websockets-subaccounts";
    WebsocketTopics["TO_WEBSOCKETS_TRADES"] = "to-websockets-trades";
    WebsocketTopics["TO_WEBSOCKETS_MARKETS"] = "to-websockets-markets";
    WebsocketTopics["TO_WEBSOCKETS_CANDLES"] = "to-websockets-candles";
})(WebsocketTopics = exports.WebsocketTopics || (exports.WebsocketTopics = {}));
var KafkaTopics;
(function (KafkaTopics) {
    KafkaTopics["TO_ENDER"] = "to-ender";
    KafkaTopics["TO_VULCAN"] = "to-vulcan";
    KafkaTopics["TO_WEBSOCKETS_ORDERBOOKS"] = "to-websockets-orderbooks";
    KafkaTopics["TO_WEBSOCKETS_SUBACCOUNTS"] = "to-websockets-subaccounts";
    KafkaTopics["TO_WEBSOCKETS_TRADES"] = "to-websockets-trades";
    KafkaTopics["TO_WEBSOCKETS_MARKETS"] = "to-websockets-markets";
    KafkaTopics["TO_WEBSOCKETS_CANDLES"] = "to-websockets-candles";
})(KafkaTopics = exports.KafkaTopics || (exports.KafkaTopics = {}));
