// @cosmwasm/ts-codegen@0.21.1/main/generators/react-query.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _case = require("case");

var _header = require("../utils/header");

var _path = require("path");

var _mkdirp = require("mkdirp");

var w = _interopRequireWildcard(require("wasm-ast-types"));

var t = _interopRequireWildcard(require("@babel/types"));

var _fs = require("fs");

var _generator = _interopRequireDefault(require("@babel/generator"));

var _utils = require("../utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var _default = /*#__PURE__*/function () {
  var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(contractName, contractInfo, outPath, reactQueryOptions) {
    var schemas, context, options, localname, ContractFile, TypesFile, QueryMsg, ExecuteMsg, typeHash, ExecuteClient, QueryClient, body, clientImports, shouldGenerateMutationHooks, imports, code;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            schemas = contractInfo.schemas;
            context = new w.RenderContext(contractInfo, {
              reactQuery: reactQueryOptions !== null && reactQueryOptions !== void 0 ? reactQueryOptions : {}
            });
            options = context.options.reactQuery;
            localname = (0, _case.pascal)("".concat(contractName)) + '.react-query.ts';
            ContractFile = (0, _case.pascal)("".concat(contractName)) + '.client';
            TypesFile = (0, _case.pascal)("".concat(contractName)) + '.types';
            QueryMsg = (0, _utils.findQueryMsg)(schemas);
            ExecuteMsg = (0, _utils.findExecuteMsg)(schemas);
            _context.next = 10;
            return (0, _utils.findAndParseTypes)(schemas);

          case 10:
            typeHash = _context.sent;
            ExecuteClient = (0, _case.pascal)("".concat(contractName, "Client"));
            QueryClient = (0, _case.pascal)("".concat(contractName, "QueryClient"));
            body = [];
            clientImports = [];
            QueryMsg && clientImports.push(QueryClient); // check that there are commands within the exec msg

            shouldGenerateMutationHooks = ExecuteMsg && (options === null || options === void 0 ? void 0 : options.version) === 'v4' && (options === null || options === void 0 ? void 0 : options.mutations) && (0, w.getMessageProperties)(ExecuteMsg).length > 0;

            if (shouldGenerateMutationHooks) {
              clientImports.push(ExecuteClient);
            } // general contract imports


            body.push(w.importStmt(Object.keys(typeHash), "./".concat(TypesFile))); // client imports

            body.push(w.importStmt(clientImports, "./".concat(ContractFile))); // query messages

            if (QueryMsg) {
              [].push.apply(body, w.createReactQueryHooks({
                context: context,
                queryMsg: QueryMsg,
                contractName: contractName,
                QueryClient: QueryClient
              }));
            }

            if (shouldGenerateMutationHooks) {
              [].push.apply(body, w.createReactQueryMutationHooks({
                context: context,
                execMsg: ExecuteMsg,
                contractName: contractName,
                ExecuteClient: ExecuteClient
              }));
            }

            if (typeHash.hasOwnProperty('Coin')) {
              // @ts-ignore
              delete context.utils.Coin;
            }

            imports = context.getImports();
            code = _header.header + (0, _generator["default"])(t.program([].concat((0, _toConsumableArray2["default"])(imports), body))).code;
            (0, _mkdirp.sync)(outPath);
            (0, _fs.writeFileSync)((0, _path.join)(outPath, localname), code);
            return _context.abrupt("return", [{
              type: 'react-query',
              contract: contractName,
              localname: localname,
              filename: (0, _path.join)(outPath, localname)
            }]);

          case 28:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function (_x, _x2, _x3, _x4) {
    return _ref.apply(this, arguments);
  };
}();

exports["default"] = _default;
