// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/sending/transfer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSendFromModuleToAccount = exports.MsgWithdrawFromSubaccount = exports.MsgDepositToSubaccount = exports.Transfer = void 0;
const subaccount_1 = require("../subaccounts/subaccount");
const coin_1 = require("../../cosmos/base/v1beta1/coin");
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseTransfer() {
    return {
        sender: undefined,
        recipient: undefined,
        assetId: 0,
        amount: helpers_1.Long.UZERO
    };
}
exports.Transfer = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.sender !== undefined) {
            subaccount_1.SubaccountId.encode(message.sender, writer.uint32(10).fork()).ldelim();
        }
        if (message.recipient !== undefined) {
            subaccount_1.SubaccountId.encode(message.recipient, writer.uint32(18).fork()).ldelim();
        }
        if (message.assetId !== 0) {
            writer.uint32(24).uint32(message.assetId);
        }
        if (!message.amount.isZero()) {
            writer.uint32(32).uint64(message.amount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTransfer();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.recipient = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.assetId = reader.uint32();
                    break;
                case 4:
                    message.amount = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseTransfer();
        message.sender = object.sender !== undefined && object.sender !== null ? subaccount_1.SubaccountId.fromPartial(object.sender) : undefined;
        message.recipient = object.recipient !== undefined && object.recipient !== null ? subaccount_1.SubaccountId.fromPartial(object.recipient) : undefined;
        message.assetId = (_a = object.assetId) !== null && _a !== void 0 ? _a : 0;
        message.amount = object.amount !== undefined && object.amount !== null ? helpers_1.Long.fromValue(object.amount) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMsgDepositToSubaccount() {
    return {
        sender: "",
        recipient: undefined,
        assetId: 0,
        quantums: helpers_1.Long.UZERO
    };
}
exports.MsgDepositToSubaccount = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.recipient !== undefined) {
            subaccount_1.SubaccountId.encode(message.recipient, writer.uint32(18).fork()).ldelim();
        }
        if (message.assetId !== 0) {
            writer.uint32(24).uint32(message.assetId);
        }
        if (!message.quantums.isZero()) {
            writer.uint32(32).uint64(message.quantums);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgDepositToSubaccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = reader.string();
                    break;
                case 2:
                    message.recipient = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.assetId = reader.uint32();
                    break;
                case 4:
                    message.quantums = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMsgDepositToSubaccount();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.recipient = object.recipient !== undefined && object.recipient !== null ? subaccount_1.SubaccountId.fromPartial(object.recipient) : undefined;
        message.assetId = (_b = object.assetId) !== null && _b !== void 0 ? _b : 0;
        message.quantums = object.quantums !== undefined && object.quantums !== null ? helpers_1.Long.fromValue(object.quantums) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMsgWithdrawFromSubaccount() {
    return {
        sender: undefined,
        recipient: "",
        assetId: 0,
        quantums: helpers_1.Long.UZERO
    };
}
exports.MsgWithdrawFromSubaccount = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.sender !== undefined) {
            subaccount_1.SubaccountId.encode(message.sender, writer.uint32(18).fork()).ldelim();
        }
        if (message.recipient !== "") {
            writer.uint32(10).string(message.recipient);
        }
        if (message.assetId !== 0) {
            writer.uint32(24).uint32(message.assetId);
        }
        if (!message.quantums.isZero()) {
            writer.uint32(32).uint64(message.quantums);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgWithdrawFromSubaccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 2:
                    message.sender = subaccount_1.SubaccountId.decode(reader, reader.uint32());
                    break;
                case 1:
                    message.recipient = reader.string();
                    break;
                case 3:
                    message.assetId = reader.uint32();
                    break;
                case 4:
                    message.quantums = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMsgWithdrawFromSubaccount();
        message.sender = object.sender !== undefined && object.sender !== null ? subaccount_1.SubaccountId.fromPartial(object.sender) : undefined;
        message.recipient = (_a = object.recipient) !== null && _a !== void 0 ? _a : "";
        message.assetId = (_b = object.assetId) !== null && _b !== void 0 ? _b : 0;
        message.quantums = object.quantums !== undefined && object.quantums !== null ? helpers_1.Long.fromValue(object.quantums) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMsgSendFromModuleToAccount() {
    return {
        authority: "",
        senderModuleName: "",
        recipient: "",
        coin: undefined
    };
}
exports.MsgSendFromModuleToAccount = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.authority !== "") {
            writer.uint32(10).string(message.authority);
        }
        if (message.senderModuleName !== "") {
            writer.uint32(18).string(message.senderModuleName);
        }
        if (message.recipient !== "") {
            writer.uint32(26).string(message.recipient);
        }
        if (message.coin !== undefined) {
            coin_1.Coin.encode(message.coin, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgSendFromModuleToAccount();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.authority = reader.string();
                    break;
                case 2:
                    message.senderModuleName = reader.string();
                    break;
                case 3:
                    message.recipient = reader.string();
                    break;
                case 4:
                    message.coin = coin_1.Coin.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseMsgSendFromModuleToAccount();
        message.authority = (_a = object.authority) !== null && _a !== void 0 ? _a : "";
        message.senderModuleName = (_b = object.senderModuleName) !== null && _b !== void 0 ? _b : "";
        message.recipient = (_c = object.recipient) !== null && _c !== void 0 ? _c : "";
        message.coin = object.coin !== undefined && object.coin !== null ? coin_1.Coin.fromPartial(object.coin) : undefined;
        return message;
    }
};
