// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/sending/tx.rpc.msg.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClientImpl = void 0;
const transfer_1 = require("./transfer");
const _m0 = __importStar(require("protobufjs/minimal"));
const tx_1 = require("./tx");
class MsgClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.createTransfer = this.createTransfer.bind(this);
        this.depositToSubaccount = this.depositToSubaccount.bind(this);
        this.withdrawFromSubaccount = this.withdrawFromSubaccount.bind(this);
        this.sendFromModuleToAccount = this.sendFromModuleToAccount.bind(this);
    }
    createTransfer(request) {
        const data = tx_1.MsgCreateTransfer.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.sending.Msg", "CreateTransfer", data);
        return promise.then(data => tx_1.MsgCreateTransferResponse.decode(new _m0.Reader(data)));
    }
    depositToSubaccount(request) {
        const data = transfer_1.MsgDepositToSubaccount.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.sending.Msg", "DepositToSubaccount", data);
        return promise.then(data => tx_1.MsgDepositToSubaccountResponse.decode(new _m0.Reader(data)));
    }
    withdrawFromSubaccount(request) {
        const data = transfer_1.MsgWithdrawFromSubaccount.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.sending.Msg", "WithdrawFromSubaccount", data);
        return promise.then(data => tx_1.MsgWithdrawFromSubaccountResponse.decode(new _m0.Reader(data)));
    }
    sendFromModuleToAccount(request) {
        const data = transfer_1.MsgSendFromModuleToAccount.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.sending.Msg", "SendFromModuleToAccount", data);
        return promise.then(data => tx_1.MsgSendFromModuleToAccountResponse.decode(new _m0.Reader(data)));
    }
}
exports.MsgClientImpl = MsgClientImpl;
