// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/bridge/query.rpc.Query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
const stargate_1 = require("@cosmjs/stargate");
const query_1 = require("./query");
class QueryClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.eventParams = this.eventParams.bind(this);
        this.proposeParams = this.proposeParams.bind(this);
        this.safetyParams = this.safetyParams.bind(this);
        this.acknowledgedEventInfo = this.acknowledgedEventInfo.bind(this);
        this.recognizedEventInfo = this.recognizedEventInfo.bind(this);
        this.delayedCompleteBridgeMessages = this.delayedCompleteBridgeMessages.bind(this);
    }
    eventParams(request = {}) {
        const data = query_1.QueryEventParamsRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Query", "EventParams", data);
        return promise.then(data => query_1.QueryEventParamsResponse.decode(new _m0.Reader(data)));
    }
    proposeParams(request = {}) {
        const data = query_1.QueryProposeParamsRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Query", "ProposeParams", data);
        return promise.then(data => query_1.QueryProposeParamsResponse.decode(new _m0.Reader(data)));
    }
    safetyParams(request = {}) {
        const data = query_1.QuerySafetyParamsRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Query", "SafetyParams", data);
        return promise.then(data => query_1.QuerySafetyParamsResponse.decode(new _m0.Reader(data)));
    }
    acknowledgedEventInfo(request = {}) {
        const data = query_1.QueryAcknowledgedEventInfoRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Query", "AcknowledgedEventInfo", data);
        return promise.then(data => query_1.QueryAcknowledgedEventInfoResponse.decode(new _m0.Reader(data)));
    }
    recognizedEventInfo(request = {}) {
        const data = query_1.QueryRecognizedEventInfoRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Query", "RecognizedEventInfo", data);
        return promise.then(data => query_1.QueryRecognizedEventInfoResponse.decode(new _m0.Reader(data)));
    }
    delayedCompleteBridgeMessages(request) {
        const data = query_1.QueryDelayedCompleteBridgeMessagesRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Query", "DelayedCompleteBridgeMessages", data);
        return promise.then(data => query_1.QueryDelayedCompleteBridgeMessagesResponse.decode(new _m0.Reader(data)));
    }
}
exports.QueryClientImpl = QueryClientImpl;
const createRpcQueryExtension = (base) => {
    const rpc = (0, stargate_1.createProtobufRpcClient)(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        eventParams(request) {
            return queryService.eventParams(request);
        },
        proposeParams(request) {
            return queryService.proposeParams(request);
        },
        safetyParams(request) {
            return queryService.safetyParams(request);
        },
        acknowledgedEventInfo(request) {
            return queryService.acknowledgedEventInfo(request);
        },
        recognizedEventInfo(request) {
            return queryService.recognizedEventInfo(request);
        },
        delayedCompleteBridgeMessages(request) {
            return queryService.delayedCompleteBridgeMessages(request);
        }
    };
};
exports.createRpcQueryExtension = createRpcQueryExtension;
