// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/bridge/tx.rpc.msg.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClientImpl = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
const tx_1 = require("./tx");
class MsgClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.acknowledgeBridges = this.acknowledgeBridges.bind(this);
        this.completeBridge = this.completeBridge.bind(this);
        this.updateEventParams = this.updateEventParams.bind(this);
        this.updateProposeParams = this.updateProposeParams.bind(this);
        this.updateSafetyParams = this.updateSafetyParams.bind(this);
    }
    acknowledgeBridges(request) {
        const data = tx_1.MsgAcknowledgeBridges.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Msg", "AcknowledgeBridges", data);
        return promise.then(data => tx_1.MsgAcknowledgeBridgesResponse.decode(new _m0.Reader(data)));
    }
    completeBridge(request) {
        const data = tx_1.MsgCompleteBridge.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Msg", "CompleteBridge", data);
        return promise.then(data => tx_1.MsgCompleteBridgeResponse.decode(new _m0.Reader(data)));
    }
    updateEventParams(request) {
        const data = tx_1.MsgUpdateEventParams.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Msg", "UpdateEventParams", data);
        return promise.then(data => tx_1.MsgUpdateEventParamsResponse.decode(new _m0.Reader(data)));
    }
    updateProposeParams(request) {
        const data = tx_1.MsgUpdateProposeParams.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Msg", "UpdateProposeParams", data);
        return promise.then(data => tx_1.MsgUpdateProposeParamsResponse.decode(new _m0.Reader(data)));
    }
    updateSafetyParams(request) {
        const data = tx_1.MsgUpdateSafetyParams.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.bridge.Msg", "UpdateSafetyParams", data);
        return promise.then(data => tx_1.MsgUpdateSafetyParamsResponse.decode(new _m0.Reader(data)));
    }
}
exports.MsgClientImpl = MsgClientImpl;
