// @dydxprotocol-indexer/postgres@18.19.39/build/src/db/helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateNormalizedCandleStartTime = exports.getTransferType = exports.getUnrealizedPnl = exports.getUnsettledFunding = exports.getMaintenanceMarginPpm = exports.ppmToString = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const big_js_1 = __importDefault(require("big.js"));
const luxon_1 = require("luxon");
const constants_1 = require("../constants");
const types_1 = require("../types");
/**
 * Converts a parts-per-million value to the string representation of the number. 1 ppm, or
 * parts-per-million is equal to 10^-6 (0.000001).
 * @param ppm Parts-per-million value.
 * @returns String representation of the parts-per-million value as a floating point number.
 */
function ppmToString(ppm) {
    return (0, big_js_1.default)(ppm).div(1000000).toFixed();
}
exports.ppmToString = ppmToString;
/**
 * Calculates maintenance margin based on initial margin and maintenance fraction.
 * maintenance margin = initial margin * maintenance fraction
 * @param initialMarginPpm Initial margin in parts-per-million.
 * @param maintenanceFractionPpm Maintenance fraction in parts-per-million.
 * @returns Maintenance margin in parts-per-million.
 */
function getMaintenanceMarginPpm(initialMarginPpm, maintenanceFractionPpm) {
    return (0, big_js_1.default)(initialMarginPpm).times(maintenanceFractionPpm).div(constants_1.ONE_MILLION).toNumber();
}
exports.getMaintenanceMarginPpm = getMaintenanceMarginPpm;
/**
 * Computes the unsettled funding for a position.
 *
 * To compute the net USDC balance for a subaccount, sum the result of this function for all
 * open perpetual positions, and add it to the latest USDC asset position for
 * this subaccount.
 *
 * When funding index is increasing, shorts get paid & unsettled funding for shorts should
 * be positive, vice versa for longs.
 * When funding index is decreasing, longs get paid & unsettled funding for longs should
 * be positive, vice versa for shorts.
 *
 * @param position
 * @param latestFundingIndex
 * @param lastUpdateFundingIndex
 */
function getUnsettledFunding(position, latestFundingIndexMap, lastUpdateFundingIndexMap) {
    return (0, big_js_1.default)(position.size).times(lastUpdateFundingIndexMap[position.perpetualId].minus(latestFundingIndexMap[position.perpetualId]));
}
exports.getUnsettledFunding = getUnsettledFunding;
/**
 * Get unrealized pnl for a perpetual position. If the perpetual market is not found in the
 * markets map or the oracle price is not found in the market, return 0.
 *
 * @param position Perpetual position object from the database, or the updated
 * perpetual position subaccountKafkaObject.
 * @param perpetualMarketsMap Map of perpetual ids to perpetual market objects from the database.
 * @returns Unrealized pnl of the position.
 */
function getUnrealizedPnl(position, perpetualMarket, marketsMap) {
    if (marketsMap[perpetualMarket.marketId] === undefined) {
        base_1.logger.error({
            at: 'getUnrealizedPnl',
            message: 'Market is undefined',
            marketId: perpetualMarket.marketId,
        });
        return (0, big_js_1.default)(0).toFixed();
    }
    if (marketsMap[perpetualMarket.marketId].oraclePrice === undefined) {
        base_1.logger.error({
            at: 'getUnrealizedPnl',
            message: 'Oracle price is undefined for market',
            marketId: perpetualMarket.marketId,
        });
        return (0, big_js_1.default)(0).toFixed();
    }
    return ((0, big_js_1.default)(position.size).times((0, big_js_1.default)(marketsMap[perpetualMarket.marketId].oraclePrice).minus(position.entryPrice))).toFixed();
}
exports.getUnrealizedPnl = getUnrealizedPnl;
/**
 * Gets the transfer type for a subaccount.
 *
 * If sender/recipient are both subaccounts, then it is a transfer_in/transfer_out.
 * If sender/recipient are wallet addresses, then it is a deposit/withdrawal.
 *
 * @param transfer
 * @param subaccountId
 */
function getTransferType(transfer, subaccountId) {
    if (transfer.senderSubaccountId === subaccountId) {
        if (transfer.recipientSubaccountId) {
            return types_1.TransferType.TRANSFER_OUT;
        }
        else {
            return types_1.TransferType.WITHDRAWAL;
        }
    }
    else if (transfer.recipientSubaccountId === subaccountId) {
        if (transfer.senderSubaccountId) {
            return types_1.TransferType.TRANSFER_IN;
        }
        else {
            return types_1.TransferType.DEPOSIT;
        }
    }
    throw new Error(`Transfer ${transfer.id} does not involve subaccount ${subaccountId}`);
}
exports.getTransferType = getTransferType;
function calculateNormalizedCandleStartTime(time, resolution) {
    const epochSeconds = Math.floor(time.toUTC().toSeconds());
    const normalizedTimeSeconds = epochSeconds - (epochSeconds % constants_1.NUM_SECONDS_IN_CANDLE_RESOLUTIONS[resolution]);
    return luxon_1.DateTime.fromSeconds(normalizedTimeSeconds).toUTC();
}
exports.calculateNormalizedCandleStartTime = calculateNormalizedCandleStartTime;
