// @dydxprotocol-indexer/postgres@18.19.39/build/src/db/knexfile.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.knexReadReplicaConfigForEnv = exports.knexPrimaryConfigForEnv = exports.getConfigForHost = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const config_1 = __importDefault(require("../config"));
require("./pg-config");
const SUPPORTED_ENVIRONMENTS = Object.values(base_1.NodeEnv);
const environment = config_1.default.NODE_ENV || base_1.NodeEnv.DEVELOPMENT;
if (!SUPPORTED_ENVIRONMENTS.includes(environment)) {
    throw new Error(`Unknown node environment ${environment}`);
}
function getConfigForHost(host) {
    return {
        client: 'pg',
        connection: {
            host,
            port: config_1.default.DB_PORT,
            database: config_1.default.DB_NAME,
            user: config_1.default.DB_USERNAME,
            password: config_1.default.DB_PASSWORD,
        },
        pool: {
            min: config_1.default.PG_POOL_MIN,
            max: config_1.default.PG_POOL_MAX,
            acquireTimeoutMillis: config_1.default.PG_ACQUIRE_CONNECTION_TIMEOUT_MS,
            createTimeoutMillis: config_1.default.PG_ACQUIRE_CONNECTION_TIMEOUT_MS,
        },
        acquireConnectionTimeout: config_1.default.PG_ACQUIRE_CONNECTION_TIMEOUT_MS,
        migrations: {
            directory: `${__dirname}/migrations/migration_files/`,
        },
    };
}
exports.getConfigForHost = getConfigForHost;
exports.knexPrimaryConfigForEnv = getConfigForHost(config_1.default.DB_HOSTNAME);
exports.knexReadReplicaConfigForEnv = getConfigForHost(config_1.default.DB_READONLY_HOSTNAME);
