// wasm-ast-types@0.15.0/module/context/context.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import { convertUtilsToImportList, getImportStatements } from "./imports";
import deepmerge from "deepmerge"; /// Plugin Types

;
export const defaultOptions = {
  types: {
    enabled: true,
    aliasExecuteMsg: false
  },
  client: {
    enabled: true,
    execExtendsQuery: true,
    noImplicitOverride: false
  },
  recoil: {
    enabled: false
  },
  messageComposer: {
    enabled: false
  },
  reactQuery: {
    enabled: false,
    optionalClient: false,
    version: 'v3',
    mutations: false,
    camelize: true,
    queryKeys: false
  }
};
export const getDefinitionSchema = schemas => {
  const aggregateSchema = {
    definitions: {//
    }
  };
  schemas.forEach(schema => {
    schema.definitions = schema.definitions || {};
    aggregateSchema.definitions = _objectSpread(_objectSpread({}, aggregateSchema.definitions), schema.definitions);
  });
  return aggregateSchema;
};
export class RenderContext {
  constructor(contract, options) {
    _defineProperty(this, "contract", void 0);

    _defineProperty(this, "utils", []);

    _defineProperty(this, "schema", void 0);

    this.contract = contract;
    this.schema = getDefinitionSchema(contract.schemas);
    this.options = deepmerge(defaultOptions, options ?? {});
  }

  refLookup($ref) {
    const refName = $ref.replace('#/definitions/', '');
    return this.schema.definitions?.[refName];
  }

  addUtil(util) {
    this.utils[util] = true;
  }

  getImports() {
    return getImportStatements(convertUtilsToImportList(this, Object.keys(this.utils)));
  }

}
