// @dydxprotocol-indexer/v4-proto-parser/build/__tests__/position-helpers.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const position_helpers_1 = require("../src/position-helpers");
const bytes_helpers_1 = require("../src/bytes-helpers");
const long_1 = __importDefault(require("long"));
const TEST_CASES = [
    [BigInt('0'), false],
    [BigInt('-0'), false],
    [BigInt('1000'), true],
    [BigInt('123456'), true],
    [BigInt('-20'), false],
    [BigInt('-1'), false],
    [BigInt('-54321'), false],
];
describe('getPositionIsLong', () => {
    it('Asset position', () => {
        TEST_CASES.forEach(([i, isLong]) => {
            const assetPosition = {
                assetId: 0,
                index: new long_1.default(0),
                quantums: (0, bytes_helpers_1.bigIntToBytes)(i),
            };
            expect((0, position_helpers_1.getPositionIsLong)(assetPosition)).toEqual(isLong);
        });
    });
    it('Perpetual position', () => {
        TEST_CASES.forEach(([i, isLong]) => {
            const perpetualPosition = {
                perpetualId: 0,
                fundingIndex: (0, bytes_helpers_1.bigIntToBytes)(BigInt(0)),
                quantums: (0, bytes_helpers_1.bigIntToBytes)(i),
                fundingPayment: (0, bytes_helpers_1.bigIntToBytes)(BigInt(0)),
            };
            expect((0, position_helpers_1.getPositionIsLong)(perpetualPosition)).toEqual(isLong);
        });
    });
});
