// @dydxprotocol-indexer/postgres@18.19.39/build/src/config.js
"use strict";
/**
 * Environment variables required by postgres module.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.postgresConfigSchema = exports.DEFAULT_KMS_KEY = exports.configSecrets = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
exports.configSecrets = [
    'DB_PASSWORD',
];
exports.DEFAULT_KMS_KEY = 'default kms key';
exports.postgresConfigSchema = {
    ...base_1.baseConfigSchema,
    // Required environment variables.
    DB_HOSTNAME: (0, base_1.parseString)({ default: 'localhost' }),
    DB_READONLY_HOSTNAME: (0, base_1.parseString)({ default: 'localhost' }),
    IS_USING_DB_READONLY: (0, base_1.parseBoolean)({ default: true }),
    DB_PORT: (0, base_1.parseInteger)({ default: 5435 }),
    DB_NAME: (0, base_1.parseString)({ default: 'dydx_dev' }),
    DB_USERNAME: (0, base_1.parseString)({ default: 'dydx_dev' }),
    DB_PASSWORD: (0, base_1.parseString)({ default: 'dydxserver123' }),
    PG_POOL_MIN: (0, base_1.parseInteger)({ default: 1 }),
    PG_POOL_MAX: (0, base_1.parseInteger)({ default: 2 }),
    PG_ACQUIRE_CONNECTION_TIMEOUT_MS: (0, base_1.parseInteger)({ default: 10000 }),
    PERPETUAL_MARKETS_REFRESHER_INTERVAL_MS: (0, base_1.parseInteger)({ default: 30000 }),
    ASSET_REFRESHER_INTERVAL_MS: (0, base_1.parseInteger)({ default: 30000 }),
    MARKET_REFRESHER_INTERVAL_MS: (0, base_1.parseInteger)({ default: 30000 }),
    LIQUIDITY_TIER_REFRESHER_INTERVAL_MS: (0, base_1.parseInteger)({ default: 30000 }),
    USE_READ_REPLICA: (0, base_1.parseBoolean)({ default: false }),
    // Optional environment variables.
    NODE_ENV: (0, base_1.parseString)({ default: null }),
};
exports.default = (0, base_1.parseSchema)(exports.postgresConfigSchema);
