// @dydxprotocol-indexer/postgres@18.19.39/build/src/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHILD_SUBACCOUNT_MULTIPLIER = exports.MAX_PARENT_SUBACCOUNTS = exports.DEFAULT_POSTGRES_OPTIONS = exports.CLOB_STATUS_TO_MARKET_STATUS = exports.SQL_TO_JSON_DEFINED_MODELS = exports.TIME_IN_FORCE_TO_API_TIME_IN_FORCE = exports.FUNDING_RATE_FROM_PROTOCOL_IN_HOURS = exports.PPM_EXPONENT = exports.USDC_ASSET_ID = exports.PROTO_TO_CANDLE_RESOLUTION = exports.CANDLE_RESOLUTION_TO_PROTO = exports.NUM_SECONDS_IN_CANDLE_RESOLUTIONS = exports.ONE_MILLION = exports.ZERO_TIME_ISO_8601 = exports.USDC_DENOM = exports.USDC_SYMBOL = exports.QUOTE_CURRENCY_ATOMIC_RESOLUTION = exports.BUFFER_ENCODING_UTF_8 = void 0;
const v4_protos_1 = require("@dydxprotocol-indexer/v4-protos");
const config_1 = __importDefault(require("./config"));
const asset_model_1 = __importDefault(require("./models/asset-model"));
const asset_position_model_1 = __importDefault(require("./models/asset-position-model"));
const fill_model_1 = __importDefault(require("./models/fill-model"));
const funding_index_updates_model_1 = __importDefault(require("./models/funding-index-updates-model"));
const liquidity_tiers_model_1 = __importDefault(require("./models/liquidity-tiers-model"));
const market_model_1 = __importDefault(require("./models/market-model"));
const oracle_price_model_1 = __importDefault(require("./models/oracle-price-model"));
const order_model_1 = __importDefault(require("./models/order-model"));
const perpetual_market_model_1 = __importDefault(require("./models/perpetual-market-model"));
const perpetual_position_model_1 = __importDefault(require("./models/perpetual-position-model"));
const subaccount_model_1 = __importDefault(require("./models/subaccount-model"));
const transfer_model_1 = __importDefault(require("./models/transfer-model"));
const types_1 = require("./types");
exports.BUFFER_ENCODING_UTF_8 = 'utf-8';
// Sourced from protocol https://github.com/dydxprotocol/v4/blob/main/lib/constants.go#L6
exports.QUOTE_CURRENCY_ATOMIC_RESOLUTION = -6;
exports.USDC_SYMBOL = 'USDC';
exports.USDC_DENOM = 'ibc/xxx';
exports.ZERO_TIME_ISO_8601 = '1970-01-01T00:00:00.000Z';
exports.ONE_MILLION = 1000000;
exports.NUM_SECONDS_IN_CANDLE_RESOLUTIONS = {
    [types_1.CandleResolution.ONE_DAY]: 60 * 60 * 24,
    [types_1.CandleResolution.FOUR_HOURS]: 60 * 60 * 4,
    [types_1.CandleResolution.ONE_HOUR]: 60 * 60,
    [types_1.CandleResolution.THIRTY_MINUTES]: 60 * 30,
    [types_1.CandleResolution.FIFTEEN_MINUTES]: 60 * 15,
    [types_1.CandleResolution.FIVE_MINUTES]: 60 * 5,
    [types_1.CandleResolution.ONE_MINUTE]: 60,
};
exports.CANDLE_RESOLUTION_TO_PROTO = {
    [types_1.CandleResolution.ONE_DAY]: v4_protos_1.CandleMessage_Resolution.ONE_DAY,
    [types_1.CandleResolution.FOUR_HOURS]: v4_protos_1.CandleMessage_Resolution.FOUR_HOURS,
    [types_1.CandleResolution.ONE_HOUR]: v4_protos_1.CandleMessage_Resolution.ONE_HOUR,
    [types_1.CandleResolution.THIRTY_MINUTES]: v4_protos_1.CandleMessage_Resolution.THIRTY_MINUTES,
    [types_1.CandleResolution.FIFTEEN_MINUTES]: v4_protos_1.CandleMessage_Resolution.FIFTEEN_MINUTES,
    [types_1.CandleResolution.FIVE_MINUTES]: v4_protos_1.CandleMessage_Resolution.FIVE_MINUTES,
    [types_1.CandleResolution.ONE_MINUTE]: v4_protos_1.CandleMessage_Resolution.ONE_MINUTE,
};
exports.PROTO_TO_CANDLE_RESOLUTION = {
    [v4_protos_1.CandleMessage_Resolution.ONE_DAY]: types_1.CandleResolution.ONE_DAY,
    [v4_protos_1.CandleMessage_Resolution.FOUR_HOURS]: types_1.CandleResolution.FOUR_HOURS,
    [v4_protos_1.CandleMessage_Resolution.ONE_HOUR]: types_1.CandleResolution.ONE_HOUR,
    [v4_protos_1.CandleMessage_Resolution.THIRTY_MINUTES]: types_1.CandleResolution.THIRTY_MINUTES,
    [v4_protos_1.CandleMessage_Resolution.FIFTEEN_MINUTES]: types_1.CandleResolution.FIFTEEN_MINUTES,
    [v4_protos_1.CandleMessage_Resolution.FIVE_MINUTES]: types_1.CandleResolution.FIVE_MINUTES,
    [v4_protos_1.CandleMessage_Resolution.ONE_MINUTE]: types_1.CandleResolution.ONE_MINUTE,
};
exports.USDC_ASSET_ID = '0';
// Parts-per-million exponent, 1 million = 10 ^ 6, exponent to convert from ppm units is 10 ^ -6
exports.PPM_EXPONENT = -6;
// On the protocol, funding is returned in 8 hour increments but we want to store funding in 1 hour
// increments
exports.FUNDING_RATE_FROM_PROTOCOL_IN_HOURS = 8;
exports.TIME_IN_FORCE_TO_API_TIME_IN_FORCE = {
    [types_1.TimeInForce.GTT]: types_1.APITimeInForce.GTT,
    [types_1.TimeInForce.IOC]: types_1.APITimeInForce.IOC,
    [types_1.TimeInForce.FOK]: types_1.APITimeInForce.FOK,
    [types_1.TimeInForce.POST_ONLY]: types_1.APITimeInForce.GTT,
};
// A list of models that have sqlToJsonConversions defined.
exports.SQL_TO_JSON_DEFINED_MODELS = [
    asset_model_1.default,
    asset_position_model_1.default,
    fill_model_1.default,
    funding_index_updates_model_1.default,
    liquidity_tiers_model_1.default,
    market_model_1.default,
    oracle_price_model_1.default,
    order_model_1.default,
    perpetual_market_model_1.default,
    perpetual_position_model_1.default,
    subaccount_model_1.default,
    transfer_model_1.default,
];
exports.CLOB_STATUS_TO_MARKET_STATUS = {
    [v4_protos_1.ClobPairStatus.CLOB_PAIR_STATUS_ACTIVE]: types_1.PerpetualMarketStatus.ACTIVE,
    [v4_protos_1.ClobPairStatus.CLOB_PAIR_STATUS_CANCEL_ONLY]: types_1.PerpetualMarketStatus.CANCEL_ONLY,
    [v4_protos_1.ClobPairStatus.CLOB_PAIR_STATUS_PAUSED]: types_1.PerpetualMarketStatus.PAUSED,
    [v4_protos_1.ClobPairStatus.CLOB_PAIR_STATUS_POST_ONLY]: types_1.PerpetualMarketStatus.POST_ONLY,
    [v4_protos_1.ClobPairStatus.CLOB_PAIR_STATUS_INITIALIZING]: types_1.PerpetualMarketStatus.INITIALIZING,
    [v4_protos_1.ClobPairStatus.CLOB_PAIR_STATUS_FINAL_SETTLEMENT]: types_1.PerpetualMarketStatus.FINAL_SETTLEMENT,
};
exports.DEFAULT_POSTGRES_OPTIONS = config_1.default.USE_READ_REPLICA
    ? {
        readReplica: true,
    } : {};
exports.MAX_PARENT_SUBACCOUNTS = 128;
exports.CHILD_SUBACCOUNT_MULTIPLIER = 1000;
