// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/bridge/bridge_event_info.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeEventInfo = void 0;
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseBridgeEventInfo() {
    return {
        nextId: 0,
        ethBlockHeight: helpers_1.Long.UZERO
    };
}
exports.BridgeEventInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.nextId !== 0) {
            writer.uint32(8).uint32(message.nextId);
        }
        if (!message.ethBlockHeight.isZero()) {
            writer.uint32(16).uint64(message.ethBlockHeight);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBridgeEventInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.nextId = reader.uint32();
                    break;
                case 2:
                    message.ethBlockHeight = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseBridgeEventInfo();
        message.nextId = (_a = object.nextId) !== null && _a !== void 0 ? _a : 0;
        message.ethBlockHeight = object.ethBlockHeight !== undefined && object.ethBlockHeight !== null ? helpers_1.Long.fromValue(object.ethBlockHeight) : helpers_1.Long.UZERO;
        return message;
    }
};
