// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/bridge/genesis.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenesisState = void 0;
const params_1 = require("./params");
const bridge_event_info_1 = require("./bridge_event_info");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseGenesisState() {
    return {
        eventParams: undefined,
        proposeParams: undefined,
        safetyParams: undefined,
        acknowledgedEventInfo: undefined
    };
}
exports.GenesisState = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.eventParams !== undefined) {
            params_1.EventParams.encode(message.eventParams, writer.uint32(10).fork()).ldelim();
        }
        if (message.proposeParams !== undefined) {
            params_1.ProposeParams.encode(message.proposeParams, writer.uint32(18).fork()).ldelim();
        }
        if (message.safetyParams !== undefined) {
            params_1.SafetyParams.encode(message.safetyParams, writer.uint32(26).fork()).ldelim();
        }
        if (message.acknowledgedEventInfo !== undefined) {
            bridge_event_info_1.BridgeEventInfo.encode(message.acknowledgedEventInfo, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.eventParams = params_1.EventParams.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.proposeParams = params_1.ProposeParams.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.safetyParams = params_1.SafetyParams.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.acknowledgedEventInfo = bridge_event_info_1.BridgeEventInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseGenesisState();
        message.eventParams = object.eventParams !== undefined && object.eventParams !== null ? params_1.EventParams.fromPartial(object.eventParams) : undefined;
        message.proposeParams = object.proposeParams !== undefined && object.proposeParams !== null ? params_1.ProposeParams.fromPartial(object.proposeParams) : undefined;
        message.safetyParams = object.safetyParams !== undefined && object.safetyParams !== null ? params_1.SafetyParams.fromPartial(object.safetyParams) : undefined;
        message.acknowledgedEventInfo = object.acknowledgedEventInfo !== undefined && object.acknowledgedEventInfo !== null ? bridge_event_info_1.BridgeEventInfo.fromPartial(object.acknowledgedEventInfo) : undefined;
        return message;
    }
};
