// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/bridge/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SafetyParams = exports.ProposeParams = exports.EventParams = void 0;
const duration_1 = require("../../google/protobuf/duration");
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseEventParams() {
    return {
        denom: "",
        ethChainId: helpers_1.Long.UZERO,
        ethAddress: ""
    };
}
exports.EventParams = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.denom !== "") {
            writer.uint32(10).string(message.denom);
        }
        if (!message.ethChainId.isZero()) {
            writer.uint32(16).uint64(message.ethChainId);
        }
        if (message.ethAddress !== "") {
            writer.uint32(26).string(message.ethAddress);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEventParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.denom = reader.string();
                    break;
                case 2:
                    message.ethChainId = reader.uint64();
                    break;
                case 3:
                    message.ethAddress = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseEventParams();
        message.denom = (_a = object.denom) !== null && _a !== void 0 ? _a : "";
        message.ethChainId = object.ethChainId !== undefined && object.ethChainId !== null ? helpers_1.Long.fromValue(object.ethChainId) : helpers_1.Long.UZERO;
        message.ethAddress = (_b = object.ethAddress) !== null && _b !== void 0 ? _b : "";
        return message;
    }
};
function createBaseProposeParams() {
    return {
        maxBridgesPerBlock: 0,
        proposeDelayDuration: undefined,
        skipRatePpm: 0,
        skipIfBlockDelayedByDuration: undefined
    };
}
exports.ProposeParams = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.maxBridgesPerBlock !== 0) {
            writer.uint32(8).uint32(message.maxBridgesPerBlock);
        }
        if (message.proposeDelayDuration !== undefined) {
            duration_1.Duration.encode(message.proposeDelayDuration, writer.uint32(18).fork()).ldelim();
        }
        if (message.skipRatePpm !== 0) {
            writer.uint32(24).uint32(message.skipRatePpm);
        }
        if (message.skipIfBlockDelayedByDuration !== undefined) {
            duration_1.Duration.encode(message.skipIfBlockDelayedByDuration, writer.uint32(34).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseProposeParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.maxBridgesPerBlock = reader.uint32();
                    break;
                case 2:
                    message.proposeDelayDuration = duration_1.Duration.decode(reader, reader.uint32());
                    break;
                case 3:
                    message.skipRatePpm = reader.uint32();
                    break;
                case 4:
                    message.skipIfBlockDelayedByDuration = duration_1.Duration.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseProposeParams();
        message.maxBridgesPerBlock = (_a = object.maxBridgesPerBlock) !== null && _a !== void 0 ? _a : 0;
        message.proposeDelayDuration = object.proposeDelayDuration !== undefined && object.proposeDelayDuration !== null ? duration_1.Duration.fromPartial(object.proposeDelayDuration) : undefined;
        message.skipRatePpm = (_b = object.skipRatePpm) !== null && _b !== void 0 ? _b : 0;
        message.skipIfBlockDelayedByDuration = object.skipIfBlockDelayedByDuration !== undefined && object.skipIfBlockDelayedByDuration !== null ? duration_1.Duration.fromPartial(object.skipIfBlockDelayedByDuration) : undefined;
        return message;
    }
};
function createBaseSafetyParams() {
    return {
        isDisabled: false,
        delayBlocks: 0
    };
}
exports.SafetyParams = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.isDisabled === true) {
            writer.uint32(8).bool(message.isDisabled);
        }
        if (message.delayBlocks !== 0) {
            writer.uint32(16).uint32(message.delayBlocks);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseSafetyParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.isDisabled = reader.bool();
                    break;
                case 2:
                    message.delayBlocks = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseSafetyParams();
        message.isDisabled = (_a = object.isDisabled) !== null && _a !== void 0 ? _a : false;
        message.delayBlocks = (_b = object.delayBlocks) !== null && _b !== void 0 ? _b : 0;
        return message;
    }
};
