// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/bridge/query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelayedCompleteBridgeMessage = exports.QueryDelayedCompleteBridgeMessagesResponse = exports.QueryDelayedCompleteBridgeMessagesRequest = exports.QueryRecognizedEventInfoResponse = exports.QueryRecognizedEventInfoRequest = exports.QueryAcknowledgedEventInfoResponse = exports.QueryAcknowledgedEventInfoRequest = exports.QuerySafetyParamsResponse = exports.QuerySafetyParamsRequest = exports.QueryProposeParamsResponse = exports.QueryProposeParamsRequest = exports.QueryEventParamsResponse = exports.QueryEventParamsRequest = void 0;
const params_1 = require("./params");
const bridge_event_info_1 = require("./bridge_event_info");
const tx_1 = require("./tx");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseQueryEventParamsRequest() {
    return {};
}
exports.QueryEventParamsRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryEventParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryEventParamsRequest();
        return message;
    }
};
function createBaseQueryEventParamsResponse() {
    return {
        params: undefined
    };
}
exports.QueryEventParamsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.params !== undefined) {
            params_1.EventParams.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryEventParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = params_1.EventParams.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryEventParamsResponse();
        message.params = object.params !== undefined && object.params !== null ? params_1.EventParams.fromPartial(object.params) : undefined;
        return message;
    }
};
function createBaseQueryProposeParamsRequest() {
    return {};
}
exports.QueryProposeParamsRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryProposeParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryProposeParamsRequest();
        return message;
    }
};
function createBaseQueryProposeParamsResponse() {
    return {
        params: undefined
    };
}
exports.QueryProposeParamsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.params !== undefined) {
            params_1.ProposeParams.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryProposeParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = params_1.ProposeParams.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryProposeParamsResponse();
        message.params = object.params !== undefined && object.params !== null ? params_1.ProposeParams.fromPartial(object.params) : undefined;
        return message;
    }
};
function createBaseQuerySafetyParamsRequest() {
    return {};
}
exports.QuerySafetyParamsRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQuerySafetyParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQuerySafetyParamsRequest();
        return message;
    }
};
function createBaseQuerySafetyParamsResponse() {
    return {
        params: undefined
    };
}
exports.QuerySafetyParamsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.params !== undefined) {
            params_1.SafetyParams.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQuerySafetyParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = params_1.SafetyParams.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQuerySafetyParamsResponse();
        message.params = object.params !== undefined && object.params !== null ? params_1.SafetyParams.fromPartial(object.params) : undefined;
        return message;
    }
};
function createBaseQueryAcknowledgedEventInfoRequest() {
    return {};
}
exports.QueryAcknowledgedEventInfoRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryAcknowledgedEventInfoRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryAcknowledgedEventInfoRequest();
        return message;
    }
};
function createBaseQueryAcknowledgedEventInfoResponse() {
    return {
        info: undefined
    };
}
exports.QueryAcknowledgedEventInfoResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.info !== undefined) {
            bridge_event_info_1.BridgeEventInfo.encode(message.info, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryAcknowledgedEventInfoResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.info = bridge_event_info_1.BridgeEventInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryAcknowledgedEventInfoResponse();
        message.info = object.info !== undefined && object.info !== null ? bridge_event_info_1.BridgeEventInfo.fromPartial(object.info) : undefined;
        return message;
    }
};
function createBaseQueryRecognizedEventInfoRequest() {
    return {};
}
exports.QueryRecognizedEventInfoRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryRecognizedEventInfoRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryRecognizedEventInfoRequest();
        return message;
    }
};
function createBaseQueryRecognizedEventInfoResponse() {
    return {
        info: undefined
    };
}
exports.QueryRecognizedEventInfoResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.info !== undefined) {
            bridge_event_info_1.BridgeEventInfo.encode(message.info, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryRecognizedEventInfoResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.info = bridge_event_info_1.BridgeEventInfo.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryRecognizedEventInfoResponse();
        message.info = object.info !== undefined && object.info !== null ? bridge_event_info_1.BridgeEventInfo.fromPartial(object.info) : undefined;
        return message;
    }
};
function createBaseQueryDelayedCompleteBridgeMessagesRequest() {
    return {
        address: ""
    };
}
exports.QueryDelayedCompleteBridgeMessagesRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.address !== "") {
            writer.uint32(10).string(message.address);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryDelayedCompleteBridgeMessagesRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.address = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryDelayedCompleteBridgeMessagesRequest();
        message.address = (_a = object.address) !== null && _a !== void 0 ? _a : "";
        return message;
    }
};
function createBaseQueryDelayedCompleteBridgeMessagesResponse() {
    return {
        messages: []
    };
}
exports.QueryDelayedCompleteBridgeMessagesResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.messages) {
            exports.DelayedCompleteBridgeMessage.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryDelayedCompleteBridgeMessagesResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.messages.push(exports.DelayedCompleteBridgeMessage.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryDelayedCompleteBridgeMessagesResponse();
        message.messages = ((_a = object.messages) === null || _a === void 0 ? void 0 : _a.map(e => exports.DelayedCompleteBridgeMessage.fromPartial(e))) || [];
        return message;
    }
};
function createBaseDelayedCompleteBridgeMessage() {
    return {
        message: undefined,
        blockHeight: 0
    };
}
exports.DelayedCompleteBridgeMessage = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.message !== undefined) {
            tx_1.MsgCompleteBridge.encode(message.message, writer.uint32(10).fork()).ldelim();
        }
        if (message.blockHeight !== 0) {
            writer.uint32(16).uint32(message.blockHeight);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDelayedCompleteBridgeMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.message = tx_1.MsgCompleteBridge.decode(reader, reader.uint32());
                    break;
                case 2:
                    message.blockHeight = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseDelayedCompleteBridgeMessage();
        message.message = object.message !== undefined && object.message !== null ? tx_1.MsgCompleteBridge.fromPartial(object.message) : undefined;
        message.blockHeight = (_a = object.blockHeight) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
};
