// @dydxprotocol-indexer/postgres@18.19.39/build/src/types/candle-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CandleColumns = exports.CandleResolution = void 0;
var CandleResolution;
(function (CandleResolution) {
    CandleResolution["ONE_MINUTE"] = "1MIN";
    CandleResolution["FIVE_MINUTES"] = "5MINS";
    CandleResolution["FIFTEEN_MINUTES"] = "15MINS";
    CandleResolution["THIRTY_MINUTES"] = "30MINS";
    CandleResolution["ONE_HOUR"] = "1HOUR";
    CandleResolution["FOUR_HOURS"] = "4HOURS";
    CandleResolution["ONE_DAY"] = "1DAY";
})(CandleResolution = exports.CandleResolution || (exports.CandleResolution = {}));
var CandleColumns;
(function (CandleColumns) {
    CandleColumns["id"] = "id";
    CandleColumns["startedAt"] = "startedAt";
    CandleColumns["ticker"] = "ticker";
    CandleColumns["resolution"] = "resolution";
    CandleColumns["low"] = "low";
    CandleColumns["high"] = "high";
    CandleColumns["open"] = "open";
    CandleColumns["close"] = "close";
    CandleColumns["baseTokenVolume"] = "baseTokenVolume";
    CandleColumns["usdVolume"] = "usdVolume";
    CandleColumns["trades"] = "trades";
    CandleColumns["startingOpenInterest"] = "startingOpenInterest";
})(CandleColumns = exports.CandleColumns || (exports.CandleColumns = {}));
