// @dydxprotocol-indexer/postgres@18.19.39/build/src/types/fill-types.js
"use strict";
/* ------- FILL TYPES ------- */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FillColumns = exports.FillType = exports.Liquidity = void 0;
var Liquidity;
(function (Liquidity) {
    Liquidity["TAKER"] = "TAKER";
    Liquidity["MAKER"] = "MAKER";
})(Liquidity = exports.Liquidity || (exports.Liquidity = {}));
var FillType;
(function (FillType) {
    // LIMIT is the fill type for a fill with a limit taker order.
    FillType["LIMIT"] = "LIMIT";
    // LIQUIDATED is for the taker side of the fill where the subaccount was liquidated.
    // The subaccountId associated with this fill is the liquidated subaccount.
    FillType["LIQUIDATED"] = "LIQUIDATED";
    // LIQUIDATION is for the maker side of the fill, never used for orders
    FillType["LIQUIDATION"] = "LIQUIDATION";
    // DELEVERAGED is for the subaccount that was deleveraged in a deleveraging event.
    // The fill type will be set to taker.
    FillType["DELEVERAGED"] = "DELEVERAGED";
    // OFFSETTING is for the offsetting subaccount in a deleveraging event.
    // The fill type will be set to maker.
    FillType["OFFSETTING"] = "OFFSETTING";
})(FillType = exports.FillType || (exports.FillType = {}));
var FillColumns;
(function (FillColumns) {
    FillColumns["id"] = "id";
    FillColumns["subaccountId"] = "subaccountId";
    FillColumns["side"] = "side";
    FillColumns["liquidity"] = "liquidity";
    FillColumns["type"] = "type";
    FillColumns["clobPairId"] = "clobPairId";
    FillColumns["orderId"] = "orderId";
    FillColumns["size"] = "size";
    FillColumns["price"] = "price";
    FillColumns["quoteAmount"] = "quoteAmount";
    FillColumns["eventId"] = "eventId";
    FillColumns["transactionHash"] = "transactionHash";
    FillColumns["createdAt"] = "createdAt";
    FillColumns["createdAtHeight"] = "createdAtHeight";
    FillColumns["clientMetadata"] = "clientMetadata";
    FillColumns["fee"] = "fee";
})(FillColumns = exports.FillColumns || (exports.FillColumns = {}));
