// @dydxprotocol-indexer/postgres@18.19.39/build/src/types/perpetual-market-types.js
"use strict";
/* ------- PERPETUAL MARKET TYPES ------- */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerpetualMarketType = exports.PerpetualMarketStatus = exports.PerpetualMarketColumns = void 0;
var PerpetualMarketColumns;
(function (PerpetualMarketColumns) {
    PerpetualMarketColumns["id"] = "id";
    PerpetualMarketColumns["clobPairId"] = "clobPairId";
    PerpetualMarketColumns["ticker"] = "ticker";
    PerpetualMarketColumns["marketId"] = "marketId";
    PerpetualMarketColumns["status"] = "status";
    PerpetualMarketColumns["priceChange24H"] = "priceChange24H";
    PerpetualMarketColumns["volume24H"] = "volume24H";
    PerpetualMarketColumns["trades24H"] = "trades24H";
    PerpetualMarketColumns["nextFundingRate"] = "nextFundingRate";
    PerpetualMarketColumns["openInterest"] = "openInterest";
    PerpetualMarketColumns["quantumConversionExponent"] = "quantumConversionExponent";
    PerpetualMarketColumns["atomicResolution"] = "atomicResolution";
    PerpetualMarketColumns["subticksPerTick"] = "subticksPerTick";
    PerpetualMarketColumns["stepBaseQuantums"] = "stepBaseQuantums";
    PerpetualMarketColumns["liquidityTierId"] = "liquidityTierId";
})(PerpetualMarketColumns = exports.PerpetualMarketColumns || (exports.PerpetualMarketColumns = {}));
var PerpetualMarketStatus;
(function (PerpetualMarketStatus) {
    PerpetualMarketStatus["ACTIVE"] = "ACTIVE";
    PerpetualMarketStatus["PAUSED"] = "PAUSED";
    PerpetualMarketStatus["CANCEL_ONLY"] = "CANCEL_ONLY";
    PerpetualMarketStatus["POST_ONLY"] = "POST_ONLY";
    PerpetualMarketStatus["INITIALIZING"] = "INITIALIZING";
    PerpetualMarketStatus["FINAL_SETTLEMENT"] = "FINAL_SETTLEMENT";
})(PerpetualMarketStatus = exports.PerpetualMarketStatus || (exports.PerpetualMarketStatus = {}));
var PerpetualMarketType;
(function (PerpetualMarketType) {
    PerpetualMarketType["CROSS"] = "CROSS";
    PerpetualMarketType["ISOLATED"] = "ISOLATED";
})(PerpetualMarketType = exports.PerpetualMarketType || (exports.PerpetualMarketType = {}));
