// @dydxprotocol-indexer/postgres@18.19.39/build/src/types/perpetual-position-types.js
"use strict";
/* ------- PERPETUAL POSITION TYPES ------- */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerpetualPositionColumns = exports.PerpetualPositionStatus = void 0;
var PerpetualPositionStatus;
(function (PerpetualPositionStatus) {
    PerpetualPositionStatus["OPEN"] = "OPEN";
    PerpetualPositionStatus["CLOSED"] = "CLOSED";
    PerpetualPositionStatus["LIQUIDATED"] = "LIQUIDATED";
})(PerpetualPositionStatus = exports.PerpetualPositionStatus || (exports.PerpetualPositionStatus = {}));
var PerpetualPositionColumns;
(function (PerpetualPositionColumns) {
    PerpetualPositionColumns["id"] = "id";
    PerpetualPositionColumns["subaccountId"] = "subaccountId";
    PerpetualPositionColumns["perpetualId"] = "perpetualId";
    PerpetualPositionColumns["side"] = "side";
    PerpetualPositionColumns["status"] = "status";
    PerpetualPositionColumns["size"] = "size";
    PerpetualPositionColumns["maxSize"] = "maxSize";
    PerpetualPositionColumns["entryPrice"] = "entryPrice";
    PerpetualPositionColumns["exitPrice"] = "exitPrice";
    PerpetualPositionColumns["sumOpen"] = "sumOpen";
    PerpetualPositionColumns["sumClose"] = "sumClose";
    PerpetualPositionColumns["createdAt"] = "createdAt";
    PerpetualPositionColumns["closedAt"] = "closedAt";
    PerpetualPositionColumns["createdAtHeight"] = "createdAtHeight";
    PerpetualPositionColumns["closedAtHeight"] = "closedAtHeight";
    PerpetualPositionColumns["openEventId"] = "openEventId";
    PerpetualPositionColumns["closeEventId"] = "closeEventId";
    PerpetualPositionColumns["lastEventId"] = "lastEventId";
    PerpetualPositionColumns["settledFunding"] = "settledFunding";
})(PerpetualPositionColumns = exports.PerpetualPositionColumns || (exports.PerpetualPositionColumns = {}));
