// @dydxprotocol-indexer/postgres@18.19.39/build/src/types/query-types.js
"use strict";
/* ------- QUERY TYPES ------- */
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryableField = void 0;
var QueryableField;
(function (QueryableField) {
    QueryableField["LIMIT"] = "limit";
    QueryableField["ID"] = "id";
    QueryableField["ADDRESS"] = "address";
    QueryableField["SUBACCOUNT_NUMBER"] = "subaccountNumber";
    QueryableField["SUBACCOUNT_ID"] = "subaccountId";
    QueryableField["SENDER_SUBACCOUNT_ID"] = "senderSubaccountId";
    QueryableField["RECIPIENT_SUBACCOUNT_ID"] = "recipientSubaccountId";
    QueryableField["SENDER_WALLET_ADDRESS"] = "senderWalletAddress";
    QueryableField["RECIPIENT_WALLET_ADDRESS"] = "recipientWalletAddress";
    QueryableField["CLIENT_ID"] = "clientId";
    QueryableField["CLOB_PAIR_ID"] = "clobPairId";
    QueryableField["SIDE"] = "side";
    QueryableField["SIZE"] = "size";
    QueryableField["TOTAL_FILLED"] = "totalFilled";
    QueryableField["PRICE"] = "price";
    QueryableField["TYPE"] = "type";
    QueryableField["STATUS"] = "status";
    QueryableField["STATUSES"] = "statuses";
    QueryableField["POST_ONLY"] = "postOnly";
    QueryableField["REDUCE_ONLY"] = "reduceOnly";
    QueryableField["PERPETUAL_ID"] = "perpetualId";
    QueryableField["LIQUIDITY"] = "liquidity";
    QueryableField["CREATED_BEFORE_OR_AT"] = "createdBeforeOrAt";
    QueryableField["CREATED_BEFORE_OR_AT_HEIGHT"] = "createdBeforeOrAtHeight";
    QueryableField["CREATED_ON_OR_AFTER"] = "createdOnOrAfter";
    QueryableField["CREATED_ON_OR_AFTER_HEIGHT"] = "createdOnOrAfterHeight";
    QueryableField["CREATED_AFTER"] = "createdAfter";
    QueryableField["CREATED_AFTER_HEIGHT"] = "createdAfterHeight";
    QueryableField["CREATED_AT"] = "createdAt";
    QueryableField["CREATED_AT_HEIGHT"] = "createdAtHeight";
    QueryableField["EVENT_ID"] = "eventId";
    QueryableField["TRANSACTION_HASH"] = "transactionHash";
    QueryableField["BLOCK_HEIGHT"] = "blockHeight";
    QueryableField["BLOCK_TIME"] = "blockTime";
    QueryableField["TRANSACTION_INDEX"] = "transactionIndex";
    QueryableField["EVENT_INDEX"] = "eventIndex";
    QueryableField["SYMBOL"] = "symbol";
    QueryableField["ATOMIC_RESOLUTION"] = "atomicResolution";
    QueryableField["HAS_MARKET"] = "hasMarket";
    QueryableField["MARKET_ID"] = "marketId";
    QueryableField["ASSET_ID"] = "assetId";
    QueryableField["IS_LONG"] = "isLong";
    QueryableField["PAIR"] = "pair";
    QueryableField["EFFECTIVE_AT"] = "effectiveAt";
    QueryableField["EFFECTIVE_AT_HEIGHT"] = "effectiveAtHeight";
    QueryableField["EFFECTIVE_BEFORE_OR_AT"] = "effectiveBeforeOrAt";
    QueryableField["EFFECTIVE_BEFORE_OR_AT_HEIGHT"] = "effectiveBeforeOrAtHeight";
    QueryableField["GOOD_TIL_BLOCK_BEFORE_OR_AT"] = "goodTilBlockBeforeOrAt";
    QueryableField["GOOD_TIL_BLOCK_TIME_BEFORE_OR_AT"] = "goodTilBlockTimeBeforeOrAt";
    QueryableField["TICKER"] = "ticker";
    QueryableField["RESOLUTION"] = "resolution";
    QueryableField["FROM_ISO"] = "fromISO";
    QueryableField["TO_ISO"] = "toISO";
    QueryableField["CREATED_BEFORE_OR_AT_BLOCK_HEIGHT"] = "createdBeforeOrAtBlockHeight";
    QueryableField["CREATED_ON_OR_AFTER_BLOCK_HEIGHT"] = "createdOnOrAfterBlockHeight";
    QueryableField["ORDER_FLAGS"] = "orderFlags";
    QueryableField["CLIENT_METADATA"] = "clientMetadata";
    QueryableField["LIQUIDITY_TIER_ID"] = "liquidityTierId";
    QueryableField["FEE"] = "fee";
    QueryableField["TRIGGER_PRICE"] = "triggerPrice";
    QueryableField["UPDATED_BEFORE_OR_AT"] = "updatedBeforeOrAt";
    QueryableField["UPDATED_ON_OR_AFTER"] = "updatedOnOrAfter";
    QueryableField["PROVIDER"] = "provider";
    QueryableField["BLOCKED"] = "blocked";
    QueryableField["BLOCK_TIME_BEFORE_OR_AT"] = "blockTimeBeforeOrAt";
    QueryableField["STARTED_AT_HEIGHT"] = "startedAtHeight";
    QueryableField["PERIOD"] = "period";
    QueryableField["STARTED_AT_HEIGHT_OR_AFTER"] = "startedAtHeightOrAfter";
    QueryableField["BLOCK_TIME_AFTER_OR_AT"] = "blockTimeAfterOrAt";
    QueryableField["BLOCK_TIME_BEFORE"] = "blockTimeBefore";
    QueryableField["ADDRESSES"] = "addresses";
    QueryableField["BLOCK_HEIGHT_BEFORE_OR_AT"] = "blockHeightBeforeOrAt";
    QueryableField["STARTED_AT_BEFORE_OR_AT"] = "startedAtBeforeOrAt";
    QueryableField["STARTED_AT_HEIGHT_BEFORE_OR_AT"] = "startedAtHeightBeforeOrAt";
})(QueryableField = exports.QueryableField || (exports.QueryableField = {}));
