// @dydxprotocol-indexer/postgres@18.19.39/build/src/types/transfer-types.js
"use strict";
/* ------- TRANSFER TYPES ------- */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferType = exports.TransferColumns = void 0;
var TransferColumns;
(function (TransferColumns) {
    TransferColumns["id"] = "id";
    TransferColumns["senderSubaccountId"] = "senderSubaccountId";
    TransferColumns["recipientSubaccountId"] = "recipientSubaccountId";
    TransferColumns["senderWalletAddress"] = "senderWalletAddress";
    TransferColumns["recipientWalletAddress"] = "recipientWalletAddress";
    TransferColumns["assetId"] = "assetId";
    TransferColumns["size"] = "size";
    TransferColumns["eventId"] = "eventId";
    TransferColumns["transactionHash"] = "transactionHash";
    TransferColumns["createdAt"] = "createdAt";
    TransferColumns["createdAtHeight"] = "createdAtHeight";
})(TransferColumns = exports.TransferColumns || (exports.TransferColumns = {}));
var TransferType;
(function (TransferType) {
    TransferType["TRANSFER_IN"] = "TRANSFER_IN";
    TransferType["TRANSFER_OUT"] = "TRANSFER_OUT";
    TransferType["DEPOSIT"] = "DEPOSIT";
    TransferType["WITHDRAWAL"] = "WITHDRAWAL";
})(TransferType = exports.TransferType || (exports.TransferType = {}));
