// @dydxprotocol-indexer/postgres@18.19.39/build/src/types/websocket-message-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIOrderStatusEnum = exports.BestEffortOpenedStatus = exports.APITimeInForce = void 0;
const order_types_1 = require("./order-types");
// TODO(DEC-1659): Change this to match API specification devised by Product team.
// API does not have POST_ONLY as a valid TimeInForce value
// Chose to implement manually instead of with `Exclude<TimeInForce, TimeInForce.POST_ONLY>`
// because Exclude breaks conversion from swagger to markdown
var APITimeInForce;
(function (APITimeInForce) {
    // GTT represents Good-Til-Time, where an order will first match with existing orders on the book
    // and any remaining size will be added to the book as a maker order, which will expire at a
    // given expiry time.
    APITimeInForce["GTT"] = "GTT";
    // FOK represents Fill-Or-KILl where it's enforced that an order will either be filled
    // completely and immediately by maker orders on the book or canceled if the entire amount can't
    // be filled.
    APITimeInForce["FOK"] = "FOK";
    // IOC represents Immediate-Or-Cancel, where it's enforced that an order only be matched with
    // maker orders on the book. If the order has remaining size after matching with existing orders
    // on the book, the remaining size is not placed on the book.
    APITimeInForce["IOC"] = "IOC";
})(APITimeInForce = exports.APITimeInForce || (exports.APITimeInForce = {}));
// Create a superset of the `OrderStatus` type including `BEST_EFFORT_OPENED` which should only
// exist in API responses.
var BestEffortOpenedStatus;
(function (BestEffortOpenedStatus) {
    BestEffortOpenedStatus["BEST_EFFORT_OPENED"] = "BEST_EFFORT_OPENED";
})(BestEffortOpenedStatus = exports.BestEffortOpenedStatus || (exports.BestEffortOpenedStatus = {}));
exports.APIOrderStatusEnum = {
    ...order_types_1.OrderStatus,
    ...BestEffortOpenedStatus,
};
