// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/prices/query.lcd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LCDQueryClient = void 0;
const helpers_1 = require("../../helpers");
class LCDQueryClient {
    constructor({ requestClient }) {
        this.req = requestClient;
        this.marketPrice = this.marketPrice.bind(this);
        this.allMarketPrices = this.allMarketPrices.bind(this);
        this.marketParam = this.marketParam.bind(this);
        this.allMarketParams = this.allMarketParams.bind(this);
    }
    /* Queries a MarketPrice by id. */
    async marketPrice(params) {
        const endpoint = `dydxprotocol/prices/market/${params.id}`;
        return await this.req.get(endpoint);
    }
    /* Queries a list of MarketPrice items. */
    async allMarketPrices(params = {
        pagination: undefined
    }) {
        const options = {
            params: {}
        };
        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
            (0, helpers_1.setPaginationParams)(options, params.pagination);
        }
        const endpoint = `dydxprotocol/prices/market`;
        return await this.req.get(endpoint, options);
    }
    /* Queries a MarketParam by id. */
    async marketParam(params) {
        const endpoint = `dydxprotocol/prices/params/market/${params.id}`;
        return await this.req.get(endpoint);
    }
    /* Queries a list of MarketParam items. */
    async allMarketParams(params = {
        pagination: undefined
    }) {
        const options = {
            params: {}
        };
        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
            (0, helpers_1.setPaginationParams)(options, params.pagination);
        }
        const endpoint = `dydxprotocol/prices/params/market`;
        return await this.req.get(endpoint, options);
    }
}
exports.LCDQueryClient = LCDQueryClient;
