// @osmonauts/proto-parser@0.34.0/main/services.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseServiceUrl = exports.parseService = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var parseServiceUrl = function parseServiceUrl(options) {
  var method = 'get';
  var url = options === null || options === void 0 ? void 0 : options['(google.api.http).get'];

  if (!url) {
    var _options$GoogleApi;

    method = 'post';
    url = options === null || options === void 0 ? void 0 : (_options$GoogleApi = options['(google.api.http)']) === null || _options$GoogleApi === void 0 ? void 0 : _options$GoogleApi.post;
  }

  if (!url) {
    return;
  }

  var match = url.match(/\{([^\}]*)\}/g); // NOTE pathParams are the original casing!
  // e.g.: 
  // "(google.api.http).get": "/cosmwasm/wasm/v1/code/{code_id}",
  // "(google.api.http).get": "/osmosis/gamm/v1beta1/pools/{poolId}/total_pool_liquidity",

  return {
    method: method,
    url: url,
    pathParams: match !== null && match !== void 0 && match.length ? match.map(function (el) {
      return el.replace('{', '').replace('}', '');
    }) : []
  };
}; // this basically takes in a ServiceMethod object! not any!


exports.parseServiceUrl = parseServiceUrl;

var parseService = function parseService(obj) {
  var _obj$fields, _obj$options;

  var fields = (_obj$fields = obj.fields) !== null && _obj$fields !== void 0 ? _obj$fields : {};
  var options = (_obj$options = obj.options) !== null && _obj$options !== void 0 ? _obj$options : {};
  var pathInfo = parseServiceUrl(options);
  if (!pathInfo) return;
  var mapping = {};
  var allParams = Object.keys(fields).map(function (field) {
    // NOTE pathParams are the original casing!
    var name = fields[field].options['(telescope:orig)'];
    mapping[name] = fields[field].options['(telescope:name)'];
    return name;
  });
  var queryParams = allParams.filter(function (param) {
    return !pathInfo.pathParams.includes(param);
  });
  var paramMap = mapping;
  return _objectSpread(_objectSpread({}, pathInfo), {}, {
    queryParams: queryParams,
    paramMap: paramMap,
    // TODO paramMap === casing?
    casing: {}
  });
}; // ProtoField:
// validatorAddress: {
//     parsedType: { name: 'string', type: 'native' },
//     isScalar: true,
//     typeNum: 13,
//     type: 'string',
//     id: 1,
//     options: {
//       '(cosmos_proto.scalar)': 'cosmos.AddressString',
//       '(telescope:name)': 'validatorAddress',
//       '(telescope:orig)': 'validator_address',
//       '(telescope:camel)': 'validatorAddress'
//     },
//     comment: 'validator_address defines the validator address to query for.'
//   }


exports.parseService = parseService;
