// @dydxprotocol-indexer/postgres@18.19.39/build/src/loops/asset-refresher.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAsset = exports.clear = exports.getAssetsMap = exports.getAssetFromId = exports.updateAssets = exports.start = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const config_1 = __importDefault(require("../config"));
const AssetTable = __importStar(require("../stores/asset-table"));
let idToAsset = {};
/**
 * Refresh loop to cache the list of all assets from the database in-memory.
 */
async function start() {
    for (;;) {
        await updateAssets();
        await (0, base_1.delay)(config_1.default.ASSET_REFRESHER_INTERVAL_MS);
    }
}
exports.start = start;
/**
 * Updates in-memory map of assets.
 */
async function updateAssets(options) {
    const startTime = Date.now();
    const assets = await AssetTable.findAll({}, [], options || { readReplica: true });
    const tmpIdToAsset = {};
    assets.forEach((asset) => {
        tmpIdToAsset[asset.id] = asset;
    });
    idToAsset = tmpIdToAsset;
    base_1.stats.timing(`${config_1.default.SERVICE_NAME}.loops.update_assets`, Date.now() - startTime);
}
exports.updateAssets = updateAssets;
/**
 * Gets the perpetual market for a given id.
 */
function getAssetFromId(id) {
    const asset = idToAsset[id];
    if (asset === undefined) {
        const message = `Unable to find asset with assetId: ${id}`;
        base_1.logger.error({
            at: 'asset-refresher#getAssetFromId',
            message,
        });
        throw new Error(message);
    }
    return asset;
}
exports.getAssetFromId = getAssetFromId;
function getAssetsMap() {
    return idToAsset;
}
exports.getAssetsMap = getAssetsMap;
function clear() {
    if (config_1.default.NODE_ENV !== base_1.NodeEnv.TEST) {
        throw new Error('clear cannot be used in non-test env');
    }
    idToAsset = {};
}
exports.clear = clear;
function addAsset(asset) {
    if (asset.id in idToAsset) {
        throw new Error(`Asset with id ${asset.id} already exists`);
    }
    idToAsset[asset.id] = asset;
}
exports.addAsset = addAsset;
