// wasm-ast-types@0.15.0/main/context/imports.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getImportStatements = exports.convertUtilsToImportList = exports.UTILS = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var t = _interopRequireWildcard(require("@babel/types"));

var _utils = require("../utils");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var makeReactQuerySwitch = function makeReactQuerySwitch(varName) {
  return function (context) {
    switch (context.options.reactQuery.version) {
      case 'v4':
        return {
          type: 'import',
          path: '@tanstack/react-query',
          name: varName
        };

      case 'v3':
      default:
        return {
          type: 'import',
          path: 'react-query',
          name: varName
        };
    }
  };
};

var UTILS = {
  MsgExecuteContract: 'cosmjs-types/cosmwasm/wasm/v1/tx',
  MsgExecuteContractEncodeObject: 'cosmwasm',
  Coin: '@cosmjs/amino',
  toUtf8: '@cosmjs/encoding',
  selectorFamily: 'recoil',
  StdFee: '@cosmjs/amino',
  CosmWasmClient: '@cosmjs/cosmwasm-stargate',
  ExecuteResult: '@cosmjs/cosmwasm-stargate',
  SigningCosmWasmClient: '@cosmjs/cosmwasm-stargate',
  // react-query
  useQuery: makeReactQuerySwitch('useQuery'),
  UseQueryOptions: makeReactQuerySwitch('UseQueryOptions'),
  useMutation: makeReactQuerySwitch('useMutation'),
  UseMutationOptions: makeReactQuerySwitch('UseMutationOptions')
};
exports.UTILS = UTILS;

var convertUtilsToImportList = function convertUtilsToImportList(context, utils) {
  return utils.map(function (util) {
    if (!UTILS.hasOwnProperty(util)) throw new Error("missing Util! ::[".concat(util, "]"));

    if (typeof UTILS[util] === 'string') {
      return {
        type: 'import',
        path: UTILS[util],
        name: util
      };
    } else if (typeof UTILS[util] === 'function') {
      return UTILS[util](context);
    } else {
      UTILS[util];
    }
  });
};

exports.convertUtilsToImportList = convertUtilsToImportList;

var getImportStatements = function getImportStatements(list) {
  var imports = list.reduce(function (m, obj) {
    m[obj.path] = m[obj.path] || [];
    var exists = m[obj.path].find(function (el) {
      return el.type === obj.type && el.path === obj.path && el.name === obj.name;
    }); // TODO some have google.protobuf.Any shows up... figure out the better way to handle this

    if (/\./.test(obj.name)) {
      obj.name = obj.name.split('.')[obj.name.split('.').length - 1];
    }

    if (!exists) m[obj.path].push(obj);
    return m;
  }, {});
  return Object.entries(imports).reduce(function (m, _ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
        importPath = _ref2[0],
        imports = _ref2[1];

    var defaultImports = imports.filter(function (a) {
      return a.type === 'default';
    });

    if (defaultImports.length) {
      if (defaultImports.length > 1) throw new Error('more than one default name NOT allowed.');
      m.push(t.importDeclaration([t.importDefaultSpecifier(t.identifier(defaultImports[0].name))], t.stringLiteral(defaultImports[0].path)));
    }

    var namedImports = imports.filter(function (a) {
      return a.type === 'import' && (!a.importAs || a.name === a.importAs);
    });

    if (namedImports.length) {
      m.push((0, _utils.importStmt)(namedImports.map(function (i) {
        return i.name;
      }), namedImports[0].path));
    }

    var aliasNamedImports = imports.filter(function (a) {
      return a.type === 'import' && a.importAs && a.name !== a.importAs;
    });
    aliasNamedImports.forEach(function (imp) {
      m.push((0, _utils.importAs)(imp.name, imp.importAs, imp.path));
    });
    var namespaced = imports.filter(function (a) {
      return a.type === 'namespace';
    });

    if (namespaced.length) {
      if (namespaced.length > 1) throw new Error('more than one namespaced name NOT allowed.');
      m.push(t.importDeclaration([t.importNamespaceSpecifier(t.identifier(namespaced[0].name))], t.stringLiteral(namespaced[0].path)));
    }

    return m;
  }, []);
};

exports.getImportStatements = getImportStatements;
