// @osmonauts/telescope@0.80.0/main/utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRoot = exports.getRelativePath = exports.fixlocalpaths = exports.UTIL_HELPERS = exports.UTILS = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _path = require("path");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var getRoot = function getRoot(ref) {
  if (ref.traversed) return ref.traversed;
  return ref.proto;
}; // default example:
// Long: { type: 'default', path: 'long', name: 'Long ' },
// namespaced:
// _m0: { type: 'namespace', path: 'protobufjs/minimal', name: '_m0' },


exports.getRoot = getRoot;
var UTILS = {
  _m0: {
    type: 'namespace',
    path: 'protobufjs/minimal',
    name: '_m0'
  },
  AminoHeight: '__helpers__',
  AminoMsg: '@cosmjs/amino',
  AminoTypes: '@cosmjs/stargate',
  base64FromBytes: '__helpers__',
  bytesFromBase64: '__helpers__',
  createProtobufRpcClient: '@cosmjs/stargate',
  decodeBech32Pubkey: '@cosmjs/amino',
  DeepPartial: '__helpers__',
  defaultRegistryTypes: '@cosmjs/stargate',
  encodeBech32Pubkey: '@cosmjs/amino',
  Exact: '__helpers__',
  fromBase64: '@cosmjs/encoding',
  fromBech32: '@cosmjs/encoding',
  fromDuration: '__helpers__',
  fromHex: '@cosmjs/encoding',
  fromJsonTimestamp: '__helpers__',
  fromTimestamp: '__helpers__',
  fromUtf8: '@cosmjs/encoding',
  GeneratedType: '@cosmjs/proto-signing',
  getRpcClient: '__extern__',
  getRpcEndpointKey: '__extern__',
  HttpEndpoint: '@cosmjs/tendermint-rpc',
  isObject: '__helpers__',
  isSet: '__helpers__',
  LCDClient: '@osmonauts/lcd',
  Long: '__helpers__',
  OfflineSigner: '@cosmjs/proto-signing',
  omitDefault: '__helpers__',
  ProtobufRpcClient: '@cosmjs/stargate',
  QueryClient: '@cosmjs/stargate',
  Registry: '@cosmjs/proto-signing',
  Rpc: '__helpers__',
  setPaginationParams: '__helpers__',
  SigningStargateClient: '@cosmjs/stargate',
  Tendermint34Client: '@cosmjs/tendermint-rpc',
  toBase64: '@cosmjs/encoding',
  toDuration: '__helpers__',
  toTimestamp: '__helpers__',
  toUtf8: '@cosmjs/encoding',
  useQuery: '@tanstack/react-query',
  useRpcEndpoint: '__react-query__',
  useRpcClient: '__react-query__',
  useTendermintClient: '__react-query__',
  ReactQueryParams: '__react-query__',
  UseQueryOptions: '@tanstack/react-query'
};
exports.UTILS = UTILS;
var UTIL_HELPERS = ['__helpers__', '__extern__', '__react-query__'];
exports.UTIL_HELPERS = UTIL_HELPERS;

var fixlocalpaths = function fixlocalpaths(imports) {
  return imports.map(function (imp) {
    return _objectSpread(_objectSpread({}, imp), {}, {
      path: UTIL_HELPERS.includes(imp.path) || imp.path.startsWith('.') || imp.path.startsWith('@') ? imp.path : "./".concat(imp.path)
    });
  });
};

exports.fixlocalpaths = fixlocalpaths;

var getRelativePath = function getRelativePath(f1, f2) {
  var rel = (0, _path.relative)((0, _path.dirname)(f1), f2);
  var importPath = rel.replace((0, _path.extname)(rel), '');
  if (!/^\./.test(importPath)) importPath = "./".concat(importPath);
  return importPath;
};

exports.getRelativePath = getRelativePath;
