// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/assets/query.lcd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LCDQueryClient = void 0;
const helpers_1 = require("../../helpers");
class LCDQueryClient {
    constructor({ requestClient }) {
        this.req = requestClient;
        this.asset = this.asset.bind(this);
        this.allAssets = this.allAssets.bind(this);
    }
    /* Queries a Asset by id. */
    async asset(params) {
        const endpoint = `dydxprotocol/assets/asset/${params.id}`;
        return await this.req.get(endpoint);
    }
    /* Queries a list of Asset items. */
    async allAssets(params = {
        pagination: undefined
    }) {
        const options = {
            params: {}
        };
        if (typeof (params === null || params === void 0 ? void 0 : params.pagination) !== "undefined") {
            (0, helpers_1.setPaginationParams)(options, params.pagination);
        }
        const endpoint = `dydxprotocol/assets/asset`;
        return await this.req.get(endpoint, options);
    }
}
exports.LCDQueryClient = LCDQueryClient;
