// @dydxprotocol-indexer/postgres@18.19.39/build/__tests__/db/helpers.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../src/types");
const helpers_1 = require("../../src/db/helpers");
const constants_1 = require("../helpers/constants");
const PerpetualPositionTable = __importStar(require("../../src/stores/perpetual-position-table"));
const MarketTable = __importStar(require("../../src/stores/market-table"));
const big_js_1 = __importDefault(require("big.js"));
const mock_generators_1 = require("../helpers/mock-generators");
const lodash_1 = __importDefault(require("lodash"));
const db_helpers_1 = require("../../src/helpers/db-helpers");
describe('helpers', () => {
    afterEach(async () => {
        await (0, db_helpers_1.clearData)();
    });
    describe('getUnsettledFunding', () => {
        const position = {
            ...constants_1.defaultPerpetualPosition,
            id: constants_1.defaultPerpetualPositionId,
            entryPrice: constants_1.defaultPerpetualPosition.entryPrice,
            sumOpen: constants_1.defaultPerpetualPosition.sumOpen,
            sumClose: constants_1.defaultPerpetualPosition.sumClose,
        };
        it('compute unsettled funding for long position', () => {
            expect((0, helpers_1.getUnsettledFunding)(position, {
                [constants_1.defaultFundingIndexUpdate.perpetualId]: (0, big_js_1.default)('12050'),
            }, {
                [constants_1.defaultFundingIndexUpdate.perpetualId]: (0, big_js_1.default)('10050'),
            })).toEqual((0, big_js_1.default)('-20000')); // 10 * (10050-12050). longs pay shorts when funding index is increasing.
        });
        it('compute unsettled funding for short position', () => {
            const shortPosition = {
                ...position,
                side: types_1.PositionSide.SHORT,
                size: '-10',
            };
            expect((0, helpers_1.getUnsettledFunding)(shortPosition, {
                [constants_1.defaultFundingIndexUpdate.perpetualId]: (0, big_js_1.default)('12050'),
            }, {
                [constants_1.defaultFundingIndexUpdate.perpetualId]: (0, big_js_1.default)('10050'),
            })).toEqual((0, big_js_1.default)('20000')); // -10 * (10050-12050). longs pay shorts when funding index is increasing.
        });
        it('compute unsettled funding for decimal position', () => {
            const shortPosition = {
                ...position,
                size: '1.35',
            };
            expect((0, helpers_1.getUnsettledFunding)(shortPosition, {
                [constants_1.defaultFundingIndexUpdate.perpetualId]: (0, big_js_1.default)('12050.124'),
            }, {
                [constants_1.defaultFundingIndexUpdate.perpetualId]: (0, big_js_1.default)('10050'),
            })).toEqual((0, big_js_1.default)('-2700.1674')); // 1.35 * (10050-12050.124). longs pay shorts when funding index is increasing.
        });
    });
    describe('getUnrealizedPnl', () => {
        it('getUnrealizedPnl long', async () => {
            await (0, mock_generators_1.seedData)();
            const perpetualPosition = await PerpetualPositionTable.create(constants_1.defaultPerpetualPosition);
            const markets = await MarketTable.findAll({}, []);
            const marketIdToMarket = lodash_1.default.keyBy(markets, types_1.MarketColumns.id);
            const unrealizedPnl = (0, helpers_1.getUnrealizedPnl)(perpetualPosition, constants_1.defaultPerpetualMarket, marketIdToMarket);
            expect(unrealizedPnl).toEqual((0, big_js_1.default)(-50000).toFixed());
        });
        it('getUnrealizedPnl short', async () => {
            await (0, mock_generators_1.seedData)();
            const perpetualPosition = await PerpetualPositionTable.create({
                ...constants_1.defaultPerpetualPosition,
                side: types_1.PositionSide.SHORT,
                size: '-10',
            });
            const markets = await MarketTable.findAll({}, []);
            const marketIdToMarket = lodash_1.default.keyBy(markets, types_1.MarketColumns.id);
            const unrealizedPnl = (0, helpers_1.getUnrealizedPnl)(perpetualPosition, constants_1.defaultPerpetualMarket, marketIdToMarket);
            expect(unrealizedPnl).toEqual((0, big_js_1.default)(50000).toFixed());
        });
    });
    describe('getMaintenanceMarginPpm', () => {
        it('5% initial margin, 60% maintenance fraction', () => {
            expect((0, helpers_1.getMaintenanceMarginPpm)(50000, 600000)).toEqual(30000);
        });
        it('25% initial margin, 100% maintenance fraction', () => {
            expect((0, helpers_1.getMaintenanceMarginPpm)(250000, 1000000)).toEqual(250000);
        });
        it('100% initial margin, 0% maintenance fraction', () => {
            expect((0, helpers_1.getMaintenanceMarginPpm)(1000000, 0)).toEqual(0);
        });
        it('0% initial margin, 100% maintenance fraction', () => {
            expect((0, helpers_1.getMaintenanceMarginPpm)(0, 1000000)).toEqual(0);
        });
        it('0% initial margin, 0% maintenance fraction', () => {
            expect((0, helpers_1.getMaintenanceMarginPpm)(0, 0)).toEqual(0);
        });
    });
});
