// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/epochs/epoch_info.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EpochInfo = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseEpochInfo() {
    return {
        name: "",
        nextTick: 0,
        duration: 0,
        currentEpoch: 0,
        currentEpochStartBlock: 0,
        isInitialized: false,
        fastForwardNextTick: false
    };
}
exports.EpochInfo = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.nextTick !== 0) {
            writer.uint32(16).uint32(message.nextTick);
        }
        if (message.duration !== 0) {
            writer.uint32(24).uint32(message.duration);
        }
        if (message.currentEpoch !== 0) {
            writer.uint32(32).uint32(message.currentEpoch);
        }
        if (message.currentEpochStartBlock !== 0) {
            writer.uint32(40).uint32(message.currentEpochStartBlock);
        }
        if (message.isInitialized === true) {
            writer.uint32(48).bool(message.isInitialized);
        }
        if (message.fastForwardNextTick === true) {
            writer.uint32(56).bool(message.fastForwardNextTick);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEpochInfo();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.name = reader.string();
                    break;
                case 2:
                    message.nextTick = reader.uint32();
                    break;
                case 3:
                    message.duration = reader.uint32();
                    break;
                case 4:
                    message.currentEpoch = reader.uint32();
                    break;
                case 5:
                    message.currentEpochStartBlock = reader.uint32();
                    break;
                case 6:
                    message.isInitialized = reader.bool();
                    break;
                case 7:
                    message.fastForwardNextTick = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g;
        const message = createBaseEpochInfo();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : "";
        message.nextTick = (_b = object.nextTick) !== null && _b !== void 0 ? _b : 0;
        message.duration = (_c = object.duration) !== null && _c !== void 0 ? _c : 0;
        message.currentEpoch = (_d = object.currentEpoch) !== null && _d !== void 0 ? _d : 0;
        message.currentEpochStartBlock = (_e = object.currentEpochStartBlock) !== null && _e !== void 0 ? _e : 0;
        message.isInitialized = (_f = object.isInitialized) !== null && _f !== void 0 ? _f : false;
        message.fastForwardNextTick = (_g = object.fastForwardNextTick) !== null && _g !== void 0 ? _g : false;
        return message;
    }
};
