// @osmonauts/ast@0.68.0/main/encoding/proto/from-sdk/utils.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fromSDK = exports.arrayTypes = void 0;

var t = _interopRequireWildcard(require("@babel/types"));

var _utils = require("../../../utils");

var _types2 = require("../../types");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var fromSDK = {
  scalar: function scalar(args) {
    var _getFieldNames = (0, _types2.getFieldNames)(args.field),
        propName = _getFieldNames.propName,
        origName = _getFieldNames.origName;

    args.context.addUtil('isSet');
    return t.objectProperty(t.identifier(propName), t.conditionalExpression(t.callExpression(t.identifier('isSet'), [t.memberExpression(t.identifier('object'), t.identifier(origName))]), t.memberExpression(t.identifier('object'), t.identifier(origName)), // getDefaultTSTypeFromProtoType(args.context, args.field, args.isOneOf)
    t.identifier('undefined')));
  },
  string: function string(args) {
    return fromSDK.scalar(args);
  },
  number: function number(args) {
    return fromSDK.scalar(args);
  },
  "double": function double(args) {
    return fromSDK.scalar(args);
  },
  "float": function float(args) {
    return fromSDK.scalar(args);
  },
  int32: function int32(args) {
    return fromSDK.scalar(args);
  },
  sint32: function sint32(args) {
    return fromSDK.scalar(args);
  },
  uint32: function uint32(args) {
    return fromSDK.scalar(args);
  },
  fixed32: function fixed32(args) {
    return fromSDK.scalar(args);
  },
  sfixed32: function sfixed32(args) {
    return fromSDK.scalar(args);
  },
  bool: function bool(args) {
    return fromSDK.scalar(args);
  },
  "long": function long(args) {
    return fromSDK.scalar(args);
  },
  int64: function int64(args) {
    return fromSDK.scalar(args);
  },
  uint64: function uint64(args) {
    return fromSDK.scalar(args);
  },
  sint64: function sint64(args) {
    return fromSDK.scalar(args);
  },
  fixed64: function fixed64(args) {
    return fromSDK.scalar(args);
  },
  sfixed64: function sfixed64(args) {
    return fromSDK.scalar(args);
  },
  type: function type(args) {
    var _getFieldNames2 = (0, _types2.getFieldNames)(args.field),
        propName = _getFieldNames2.propName,
        origName = _getFieldNames2.origName;

    var name = args.context.getTypeName(args.field);
    args.context.addUtil('isSet');
    return t.objectProperty(t.identifier(propName), t.conditionalExpression(t.callExpression(t.identifier('isSet'), [t.memberExpression(t.identifier('object'), t.identifier(origName))]), t.callExpression(t.memberExpression(t.identifier(name), t.identifier('fromSDK')), [t.memberExpression(t.identifier('object'), t.identifier(origName))]), t.identifier('undefined')));
  },
  "enum": function _enum(args) {
    var _getFieldNames3 = (0, _types2.getFieldNames)(args.field),
        propName = _getFieldNames3.propName,
        origName = _getFieldNames3.origName;

    args.context.addUtil('isSet');
    var fromSDKFuncName = args.context.getFromEnum(args.field);
    return t.objectProperty(t.identifier(propName), t.conditionalExpression(t.callExpression(t.identifier('isSet'), [t.memberExpression(t.identifier('object'), t.identifier(origName))]), t.callExpression(t.identifier(fromSDKFuncName), [t.memberExpression(t.identifier('object'), t.identifier(origName))]), args.isOptional ? t.identifier('undefined') : t.numericLiteral(0)));
  },
  bytes: function bytes(args) {
    return fromSDK.scalar(args);
  },
  duration: function duration(args) {
    return fromSDK.type(args);
  },
  timestamp: function timestamp(args) {
    return fromSDK.type(args);
  },
  //  labels: isObject(object.labels) ? Object.entries(object.labels).reduce<{
  //     [key: string]: string;
  //   }>((acc, [key, value]) => {
  //     acc[key] = String(value);
  //     return acc;
  //   }, {}) : {},
  //   referenceMap: isObject(object.referenceMap) ? Object.entries(object.referenceMap).reduce<{
  //     [key: Long]: Reference;
  //   }>((acc, [key, value]) => {
  //     acc[Number(key)] = Reference.fromSDK(value);
  //     return acc;
  //   }, {}) : {},
  keyHash: function keyHash(args) {
    var _getFieldNames4 = (0, _types2.getFieldNames)(args.field),
        propName = _getFieldNames4.propName,
        origName = _getFieldNames4.origName;

    var keyType = args.field.keyType;
    var valueType = args.field.parsedType.name;
    args.context.addUtil('isObject');
    var fromSDK = null; // valueTypeType: string for identifier

    var valueTypeType = valueType;

    switch (valueType) {
      case 'string':
        fromSDK = t.callExpression(t.identifier('String'), [t.identifier('value')]);
        break;

      case 'int32':
      case 'uint32':
        valueTypeType = 'number';
        fromSDK = t.callExpression(t.identifier('Number'), [t.identifier('value')]);
        break;

      case 'int64':
      case 'uint64':
        valueTypeType = 'Long';
        fromSDK = t.callExpression(t.memberExpression(t.identifier('Long'), t.identifier('fromValue')), [t.tsAsExpression(t.identifier('value'), t.tsUnionType([t.tsTypeReference(t.identifier('Long')), t.tsStringKeyword()]))]);
        break;

      default:
        fromSDK = t.callExpression(t.memberExpression(t.identifier(valueType), t.identifier('fromSDK')), [t.identifier('value')]);
    }

    var wrapKey = null;
    var keyTypeType = null;

    switch (keyType) {
      case 'string':
        wrapKey = function wrapKey(a) {
          return a;
        };

        keyTypeType = t.tsStringKeyword();
        break;

      case 'int64':
      case 'uint64':
        wrapKey = function wrapKey(a) {
          return t.callExpression(t.identifier('Number'), [a]);
        };

        keyTypeType = t.tsTypeReference(t.identifier('Long'));
        break;

      case 'uint32':
      case 'int32':
        wrapKey = function wrapKey(a) {
          return t.callExpression(t.identifier('Number'), [a]);
        };

        keyTypeType = t.tsTypeReference(t.identifier('number'));
        break;

      default:
        throw new Error('keyHash requires new type. Ask maintainers.');
    }

    return t.objectProperty(t.identifier(propName), t.conditionalExpression(t.callExpression(t.identifier('isObject'), [t.memberExpression(t.identifier('object'), t.identifier(origName))]), (0, _utils.callExpression)(t.memberExpression(t.callExpression(t.memberExpression(t.identifier('Object'), t.identifier('entries')), [t.memberExpression(t.identifier('object'), t.identifier(origName))]), t.identifier('reduce')), [t.arrowFunctionExpression([t.identifier('acc'), t.arrayPattern([t.identifier('key'), t.identifier('value')])], t.blockStatement([t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('acc'), wrapKey(t.identifier('key')), true), fromSDK)), t.returnStatement(t.identifier('acc'))])), t.objectExpression([])], t.tsTypeParameterInstantiation([t.tsTypeLiteral([t.tsIndexSignature([(0, _utils.identifier)('key', t.tsTypeAnnotation(keyTypeType))], t.tsTypeAnnotation(t.tsTypeReference(t.identifier(valueTypeType))))])])), t.objectExpression([])));
  },
  // codeIds: Array.isArray(object?.codeIds) ? object.codeIds.map((e: any) => Long.fromString(e)) : [],
  array: function array(args, expr) {
    var _getFieldNames5 = (0, _types2.getFieldNames)(args.field),
        propName = _getFieldNames5.propName,
        origName = _getFieldNames5.origName;

    return t.objectProperty(t.identifier(propName), t.conditionalExpression(t.callExpression(t.memberExpression(t.identifier('Array'), t.identifier('isArray')), [t.optionalMemberExpression(t.identifier('object'), t.identifier(origName), false, true)]), t.callExpression(t.memberExpression(t.memberExpression(t.identifier('object'), t.identifier(origName)), t.identifier('map')), [t.arrowFunctionExpression([(0, _utils.identifier)('e', t.tsTypeAnnotation(t.tsAnyKeyword()))], expr, false)]), t.arrayExpression([])));
  }
};
exports.fromSDK = fromSDK;
var arrayTypes = {
  scalar: function scalar() {
    return t.identifier('e');
  },
  string: function string() {
    return arrayTypes.scalar();
  },
  bool: function bool() {
    return arrayTypes.scalar();
  },
  bytes: function bytes(args) {
    return arrayTypes.scalar();
  },
  "long": function long() {
    return arrayTypes.scalar();
  },
  uint64: function uint64() {
    return arrayTypes.scalar();
  },
  int64: function int64() {
    return arrayTypes.scalar();
  },
  sint64: function sint64() {
    return arrayTypes.scalar();
  },
  fixed64: function fixed64() {
    return arrayTypes.scalar();
  },
  sfixed64: function sfixed64() {
    return arrayTypes.scalar();
  },
  number: function number() {
    return arrayTypes.scalar();
  },
  uint32: function uint32() {
    return arrayTypes.scalar();
  },
  int32: function int32() {
    return arrayTypes.scalar();
  },
  sint32: function sint32() {
    return arrayTypes.scalar();
  },
  fixed32: function fixed32() {
    return arrayTypes.scalar();
  },
  sfixed32: function sfixed32() {
    return arrayTypes.scalar();
  },
  "double": function double() {
    return arrayTypes.scalar();
  },
  "float": function float() {
    return arrayTypes.scalar();
  },
  "enum": function _enum(args) {
    var fromSDKFuncName = args.context.getFromEnum(args.field);
    return t.callExpression(t.identifier(fromSDKFuncName), [t.identifier('e')]);
  },
  // tokenInMaxs: Array.isArray(object?.tokenInMaxs) ? object.tokenInMaxs.map((e: any) => Coin.fromSDK(e)) : []
  type: function type(args) {
    var name = args.context.getTypeName(args.field);
    return t.callExpression(t.memberExpression(t.identifier(name), t.identifier('fromSDK')), [t.identifier('e')]);
  }
};
exports.arrayTypes = arrayTypes;
