// @cosmwasm/ts-codegen@0.21.1/main/builder/builder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TSBuilder = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _wasmAstTypes = require("wasm-ast-types");

var _header = require("../utils/header");

var _path = require("path");

var _fs = require("fs");

var _mkdirp = require("mkdirp");

var _messageComposer = _interopRequireDefault(require("../generators/message-composer"));

var _types = _interopRequireDefault(require("../generators/types"));

var _reactQuery = _interopRequireDefault(require("../generators/react-query"));

var _recoil = _interopRequireDefault(require("../generators/recoil"));

var _client = _interopRequireDefault(require("../generators/client"));

var _utils = require("../utils");

var _deepmerge = _interopRequireDefault(require("deepmerge"));

var _case = require("case");

var _bundler = require("../bundler");

var _generator = _interopRequireDefault(require("@babel/generator"));

var t = _interopRequireWildcard(require("@babel/types"));

var _excluded = ["enabled"],
    _excluded2 = ["enabled"],
    _excluded3 = ["enabled"],
    _excluded4 = ["enabled"],
    _excluded5 = ["enabled"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var defaultOpts = {
  bundle: {
    enabled: true,
    scope: 'contracts',
    bundleFile: 'bundle.ts'
  }
};
;
;
;

var TSBuilder = /*#__PURE__*/function () {
  function TSBuilder(_ref) {
    var contracts = _ref.contracts,
        outPath = _ref.outPath,
        options = _ref.options;
    (0, _classCallCheck2["default"])(this, TSBuilder);
    (0, _defineProperty2["default"])(this, "contracts", void 0);
    (0, _defineProperty2["default"])(this, "outPath", void 0);
    (0, _defineProperty2["default"])(this, "options", void 0);
    (0, _defineProperty2["default"])(this, "files", []);
    this.contracts = contracts;
    this.outPath = outPath;
    this.options = (0, _deepmerge["default"])((0, _deepmerge["default"])(_wasmAstTypes.defaultOptions, defaultOpts), options !== null && options !== void 0 ? options : {});
  }

  (0, _createClass2["default"])(TSBuilder, [{
    key: "getContracts",
    value: function getContracts() {
      return this.contracts.map(function (contractOpt) {
        if (typeof contractOpt === 'string') {
          var name = (0, _path.basename)(contractOpt);
          var contractName = (0, _case.pascal)(name);
          return {
            name: contractName,
            dir: contractOpt
          };
        }

        return {
          name: (0, _case.pascal)(contractOpt.name),
          dir: contractOpt.dir
        };
      });
    }
  }, {
    key: "renderTypes",
    value: function () {
      var _renderTypes = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(contract) {
        var _this$options$types, enabled, options, contractInfo, files;

        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$options$types = this.options.types, enabled = _this$options$types.enabled, options = (0, _objectWithoutProperties2["default"])(_this$options$types, _excluded);

                if (enabled) {
                  _context.next = 3;
                  break;
                }

                return _context.abrupt("return");

              case 3:
                _context.next = 5;
                return (0, _utils.readSchemas)({
                  schemaDir: contract.dir
                });

              case 5:
                contractInfo = _context.sent;
                _context.next = 8;
                return (0, _types["default"])(contract.name, contractInfo, this.outPath, options);

              case 8:
                files = _context.sent;
                [].push.apply(this.files, files);

              case 10:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function renderTypes(_x) {
        return _renderTypes.apply(this, arguments);
      }

      return renderTypes;
    }()
  }, {
    key: "renderClient",
    value: function () {
      var _renderClient = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(contract) {
        var _this$options$client, enabled, options, contractInfo, files;

        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _this$options$client = this.options.client, enabled = _this$options$client.enabled, options = (0, _objectWithoutProperties2["default"])(_this$options$client, _excluded2);

                if (enabled) {
                  _context2.next = 3;
                  break;
                }

                return _context2.abrupt("return");

              case 3:
                _context2.next = 5;
                return (0, _utils.readSchemas)({
                  schemaDir: contract.dir
                });

              case 5:
                contractInfo = _context2.sent;
                _context2.next = 8;
                return (0, _client["default"])(contract.name, contractInfo, this.outPath, options);

              case 8:
                files = _context2.sent;
                [].push.apply(this.files, files);

              case 10:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function renderClient(_x2) {
        return _renderClient.apply(this, arguments);
      }

      return renderClient;
    }()
  }, {
    key: "renderRecoil",
    value: function () {
      var _renderRecoil = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(contract) {
        var _this$options$recoil, enabled, options, contractInfo, files;

        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _this$options$recoil = this.options.recoil, enabled = _this$options$recoil.enabled, options = (0, _objectWithoutProperties2["default"])(_this$options$recoil, _excluded3);

                if (enabled) {
                  _context3.next = 3;
                  break;
                }

                return _context3.abrupt("return");

              case 3:
                _context3.next = 5;
                return (0, _utils.readSchemas)({
                  schemaDir: contract.dir
                });

              case 5:
                contractInfo = _context3.sent;
                _context3.next = 8;
                return (0, _recoil["default"])(contract.name, contractInfo, this.outPath, options);

              case 8:
                files = _context3.sent;
                [].push.apply(this.files, files);

              case 10:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function renderRecoil(_x3) {
        return _renderRecoil.apply(this, arguments);
      }

      return renderRecoil;
    }()
  }, {
    key: "renderReactQuery",
    value: function () {
      var _renderReactQuery = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(contract) {
        var _this$options$reactQu, enabled, options, contractInfo, files;

        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _this$options$reactQu = this.options.reactQuery, enabled = _this$options$reactQu.enabled, options = (0, _objectWithoutProperties2["default"])(_this$options$reactQu, _excluded4);

                if (enabled) {
                  _context4.next = 3;
                  break;
                }

                return _context4.abrupt("return");

              case 3:
                _context4.next = 5;
                return (0, _utils.readSchemas)({
                  schemaDir: contract.dir
                });

              case 5:
                contractInfo = _context4.sent;
                _context4.next = 8;
                return (0, _reactQuery["default"])(contract.name, contractInfo, this.outPath, options);

              case 8:
                files = _context4.sent;
                [].push.apply(this.files, files);

              case 10:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function renderReactQuery(_x4) {
        return _renderReactQuery.apply(this, arguments);
      }

      return renderReactQuery;
    }()
  }, {
    key: "renderMessageComposer",
    value: function () {
      var _renderMessageComposer = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(contract) {
        var _this$options$message, enabled, options, contractInfo, files;

        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _this$options$message = this.options.messageComposer, enabled = _this$options$message.enabled, options = (0, _objectWithoutProperties2["default"])(_this$options$message, _excluded5);

                if (enabled) {
                  _context5.next = 3;
                  break;
                }

                return _context5.abrupt("return");

              case 3:
                _context5.next = 5;
                return (0, _utils.readSchemas)({
                  schemaDir: contract.dir
                });

              case 5:
                contractInfo = _context5.sent;
                _context5.next = 8;
                return (0, _messageComposer["default"])(contract.name, contractInfo, this.outPath, options);

              case 8:
                files = _context5.sent;
                [].push.apply(this.files, files);

              case 10:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function renderMessageComposer(_x5) {
        return _renderMessageComposer.apply(this, arguments);
      }

      return renderMessageComposer;
    }()
  }, {
    key: "build",
    value: function () {
      var _build = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
        var contracts, c, contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                contracts = this.getContracts();
                c = 0;

              case 2:
                if (!(c < contracts.length)) {
                  _context6.next = 17;
                  break;
                }

                contract = contracts[c];
                _context6.next = 6;
                return this.renderTypes(contract);

              case 6:
                _context6.next = 8;
                return this.renderClient(contract);

              case 8:
                _context6.next = 10;
                return this.renderMessageComposer(contract);

              case 10:
                _context6.next = 12;
                return this.renderReactQuery(contract);

              case 12:
                _context6.next = 14;
                return this.renderRecoil(contract);

              case 14:
                c++;
                _context6.next = 2;
                break;

              case 17:
                if (this.options.bundle.enabled) {
                  this.bundle();
                }

              case 18:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function build() {
        return _build.apply(this, arguments);
      }

      return build;
    }()
  }, {
    key: "bundle",
    value: function () {
      var _bundle = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7() {
        var _this = this;

        var allFiles, bundleFile, bundleVariables, importPaths, ast, code;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                allFiles = this.files;
                bundleFile = this.options.bundle.bundleFile;
                bundleVariables = {};
                importPaths = [];
                allFiles.forEach(function (file) {
                  (0, _bundler.createFileBundle)("".concat(_this.options.bundle.scope, ".").concat(file.contract), file.localname, bundleFile, importPaths, bundleVariables);
                });
                ast = (0, _bundler.recursiveModuleBundle)(bundleVariables);
                code = (0, _generator["default"])(t.program([].concat(importPaths, (0, _toConsumableArray2["default"])(ast)))).code;
                (0, _mkdirp.sync)(this.outPath);
                if (code.trim() === '') code = 'export {};';
                (0, _fs.writeFileSync)((0, _path.join)(this.outPath, bundleFile), _header.header + code);

              case 10:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function bundle() {
        return _bundle.apply(this, arguments);
      }

      return bundle;
    }()
  }]);
  return TSBuilder;
}();

exports.TSBuilder = TSBuilder;
