// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/rewards/params.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseParams() {
    return {
        treasuryAccount: "",
        denom: "",
        denomExponent: 0,
        marketId: 0,
        feeMultiplierPpm: 0
    };
}
exports.Params = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.treasuryAccount !== "") {
            writer.uint32(10).string(message.treasuryAccount);
        }
        if (message.denom !== "") {
            writer.uint32(18).string(message.denom);
        }
        if (message.denomExponent !== 0) {
            writer.uint32(24).sint32(message.denomExponent);
        }
        if (message.marketId !== 0) {
            writer.uint32(32).uint32(message.marketId);
        }
        if (message.feeMultiplierPpm !== 0) {
            writer.uint32(40).uint32(message.feeMultiplierPpm);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.treasuryAccount = reader.string();
                    break;
                case 2:
                    message.denom = reader.string();
                    break;
                case 3:
                    message.denomExponent = reader.sint32();
                    break;
                case 4:
                    message.marketId = reader.uint32();
                    break;
                case 5:
                    message.feeMultiplierPpm = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseParams();
        message.treasuryAccount = (_a = object.treasuryAccount) !== null && _a !== void 0 ? _a : "";
        message.denom = (_b = object.denom) !== null && _b !== void 0 ? _b : "";
        message.denomExponent = (_c = object.denomExponent) !== null && _c !== void 0 ? _c : 0;
        message.marketId = (_d = object.marketId) !== null && _d !== void 0 ? _d : 0;
        message.feeMultiplierPpm = (_e = object.feeMultiplierPpm) !== null && _e !== void 0 ? _e : 0;
        return message;
    }
};
