// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/asset-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class AssetModel extends objection_1.Model {
    static get tableName() {
        return 'assets';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'symbol',
                'atomicResolution',
                'hasMarket',
            ],
            properties: {
                id: { type: 'string', pattern: validators_1.IntegerPattern },
                symbol: { type: 'string' },
                atomicResolution: { type: 'integer' },
                hasMarket: { type: 'boolean', default: false },
                marketId: { type: ['integer', 'null'] },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            symbol: 'string',
            atomicResolution: 'integer',
            hasMarket: 'boolean',
            marketId: 'integer',
        };
    }
}
exports.default = AssetModel;
AssetModel.relationMappings = {
    assetPosition: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'asset-position-model'),
        join: {
            from: 'assets.id',
            to: 'asset_positions.assetId',
        },
    },
    transfers: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'transfer-model'),
        join: {
            from: 'assets.id',
            to: 'transfers.assetId',
        },
    },
    markets: {
        relation: objection_1.Model.HasOneRelation,
        modelClass: path_1.default.join(__dirname, 'market-model'),
        join: {
            from: 'assets.marketId',
            to: 'markets.id',
        },
    },
};
