// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/asset-position-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
const base_model_1 = __importDefault(require("./base-model"));
class AssetPositionModel extends base_model_1.default {
    static get tableName() {
        return 'asset_positions';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'subaccountId',
                'assetId',
                'size',
                'isLong',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                subaccountId: { type: 'string', format: 'uuid' },
                assetId: { type: 'string', pattern: validators_1.IntegerPattern },
                size: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                isLong: { type: 'boolean', default: true },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            subaccountId: 'string',
            assetId: 'string',
            size: 'string',
            isLong: 'boolean',
        };
    }
}
exports.default = AssetPositionModel;
AssetPositionModel.relationMappings = {
    subaccount: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'subaccount-model'),
        join: {
            from: 'asset_positions.subaccountId',
            to: 'subaccounts.id',
        },
    },
    asset: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'asset-model'),
        join: {
            from: 'asset_positions.assetId',
            to: 'assets.id',
        },
    },
};
