// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/block-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class BlockModel extends objection_1.Model {
    static get tableName() {
        return 'blocks';
    }
    static get idColumn() {
        return 'blockHeight';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'blockHeight',
                'time',
            ],
            properties: {
                blockHeight: { type: 'string', pattern: validators_1.IntegerPattern },
                time: { type: 'string', format: 'date-time' },
            },
        };
    }
}
exports.default = BlockModel;
BlockModel.relationMappings = {
    fills: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'fill-model'),
        join: {
            from: 'blocks.blockHeight',
            to: 'fills.createdAtHeight',
        },
    },
    transfers: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'transfer-model'),
        join: {
            from: 'blocks.blockHeight',
            to: 'transfers.createdAtHeight',
        },
    },
    tendermintEvents: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'tendermint-event-model'),
        join: {
            from: 'blocks.blockHeight',
            to: 'tendermint_events.blockHeight',
        },
    },
    oraclePrices: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'oracle-price-model'),
        join: {
            from: 'blocks.blockHeight',
            to: 'oracle_prices.effectiveAtHeight',
        },
    },
};
