// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/candle-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
const types_1 = require("../types");
class CandleModel extends objection_1.Model {
    static get tableName() {
        return 'candles';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'startedAt',
                'ticker',
                'resolution',
                'low',
                'high',
                'open',
                'close',
                'baseTokenVolume',
                'trades',
                'usdVolume',
                'startingOpenInterest',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                startedAt: { type: 'string', format: 'date-time' },
                ticker: { type: 'string' },
                resolution: { type: 'string', enum: [...Object.values(types_1.CandleResolution)] },
                low: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                high: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                open: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                close: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                baseTokenVolume: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                usdVolume: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                trades: { type: 'integer' },
                startingOpenInterest: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
            },
        };
    }
}
exports.default = CandleModel;
CandleModel.relationMappings = {};
