// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/compliance-data-model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
const types_1 = require("../types");
class ComplianceDataModel extends objection_1.Model {
    static get tableName() {
        return 'compliance_data';
    }
    static get idColumn() {
        return ['address', 'provider'];
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'address',
                'provider',
                'blocked',
                'updatedAt',
            ],
            properties: {
                address: { type: 'string' },
                provider: { type: 'string', enum: [...Object.values(types_1.ComplianceProvider)] },
                chain: { type: ['string', 'null'], default: null },
                blocked: { type: 'boolean' },
                riskScore: { type: ['string', 'null'], pattern: validators_1.NumericPattern, default: null },
                updatedAt: { type: 'string', format: 'date-time' },
            },
        };
    }
}
exports.default = ComplianceDataModel;
ComplianceDataModel.relationMappings = {};
