// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/fill-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
const types_1 = require("../types");
class FillModel extends objection_1.Model {
    static get tableName() {
        return 'fills';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'subaccountId',
                'side',
                'liquidity',
                'type',
                'clobPairId',
                'orderId',
                'size',
                'price',
                'quoteAmount',
                'eventId',
                'transactionHash',
                'createdAt',
                'createdAtHeight',
                'fee',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                subaccountId: { type: 'string', format: 'uuid' },
                side: { type: 'string', enum: [...Object.values(types_1.OrderSide)] },
                liquidity: { type: 'string', enum: [...Object.values(types_1.Liquidity)] },
                type: { type: 'string', enum: [...Object.values(types_1.FillType)] },
                clobPairId: { type: 'string', pattern: validators_1.IntegerPattern },
                orderId: { type: ['string', 'null'], default: null, format: 'uuid' },
                size: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                price: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                quoteAmount: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                transactionHash: { type: 'string' },
                createdAt: { type: 'string', format: 'date-time' },
                createdAtHeight: { type: 'string', pattern: validators_1.IntegerPattern },
                clientMetadata: { type: ['string', 'null'], pattern: validators_1.IntegerPattern },
                fee: { type: 'string', pattern: validators_1.NumericPattern },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            subaccountId: 'string',
            side: 'string',
            liquidity: 'string',
            type: 'string',
            clobPairId: 'string',
            orderId: 'string',
            size: 'string',
            price: 'string',
            quoteAmount: 'string',
            eventId: 'hex-string',
            transactionHash: 'string',
            createdAt: 'date-time',
            createdAtHeight: 'string',
            clientMetadata: 'string',
            fee: 'string',
        };
    }
}
exports.default = FillModel;
FillModel.relationMappings = {
    subaccounts: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'subaccount-model'),
        join: {
            from: 'fills.subaccountId',
            to: 'subaccounts.id',
        },
    },
    tendermintEvents: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'tendermint-event-model'),
        join: {
            from: 'fills.eventId',
            to: 'tendermint_events.id',
        },
    },
    blocks: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'fills.createdAtHeight',
            to: 'blocks.blockHeight',
        },
    },
};
