// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/funding-index-updates-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class FundingIndexUpdatesModel extends objection_1.Model {
    static get tableName() {
        return 'funding_index_updates';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'perpetualId',
                'eventId',
                'rate',
                'oraclePrice',
                'fundingIndex',
                'effectiveAt',
                'effectiveAtHeight',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                perpetualId: { type: 'string', pattern: validators_1.IntegerPattern },
                rate: { type: 'string', pattern: validators_1.NumericPattern },
                oraclePrice: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                fundingIndex: { type: 'string', pattern: validators_1.NumericPattern },
                effectiveAt: { type: 'string', format: 'date-time' },
                effectiveAtHeight: { type: 'string', pattern: validators_1.IntegerPattern },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            perpetualId: 'string',
            eventId: 'hex-string',
            rate: 'string',
            oraclePrice: 'string',
            fundingIndex: 'string',
            effectiveAt: 'date-time',
            effectiveAtHeight: 'string',
        };
    }
}
exports.default = FundingIndexUpdatesModel;
FundingIndexUpdatesModel.relationMappings = {
    openTendermintEvent: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'tendermint-event-model'),
        join: {
            from: 'funding_index_updates.eventId',
            to: 'tendermint_events.id',
        },
    },
    blocks: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'funding_index_updates.effectiveAtHeight',
            to: 'blocks.blockHeight',
        },
    },
    perpetualMarkets: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-market-model'),
        join: {
            from: 'funding_index_updates.perpetualId',
            to: 'perpetual_markets.id',
        },
    },
};
