// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/liquidity-tiers-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validators_1 = require("../lib/validators");
const base_model_1 = __importDefault(require("./base-model"));
class LiquidityTiersModel extends base_model_1.default {
    static get tableName() {
        return 'liquidity_tiers';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'name',
                'initialMarginPpm',
                'maintenanceFractionPpm',
            ],
            properties: {
                id: { type: 'integer' },
                name: { type: 'string' },
                initialMarginPpm: { type: 'string', pattern: validators_1.IntegerPattern },
                maintenanceFractionPpm: { type: 'string', pattern: validators_1.IntegerPattern },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'integer',
            name: 'string',
            initialMarginPpm: 'string',
            maintenanceFractionPpm: 'string',
        };
    }
}
exports.default = LiquidityTiersModel;
LiquidityTiersModel.relationMappings = {};
