// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/market-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class MarketModel extends objection_1.Model {
    static get tableName() {
        return 'markets';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'pair',
                'exponent',
                'minPriceChangePpm',
            ],
            properties: {
                id: { type: 'integer' },
                pair: { type: 'string' },
                exponent: { type: 'integer' },
                minPriceChangePpm: { type: 'integer' },
                oraclePrice: { type: ['string', 'null'], pattern: validators_1.NonNegativeNumericPattern, default: null },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'integer',
            pair: 'string',
            exponent: 'integer',
            minPriceChangePpm: 'integer',
            oraclePrice: 'string',
        };
    }
}
exports.default = MarketModel;
MarketModel.relationMappings = {
    perpetualMarket: {
        relation: objection_1.Model.HasOneRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-market-model'),
        join: {
            from: 'markets.id',
            to: 'perpetual_markets.marketId',
        },
    },
    oraclePrices: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'oracle-price-model'),
        join: {
            from: 'markets.id',
            to: 'oracle_prices.marketId',
        },
    },
};
