// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/oracle-price-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class OraclePriceModel extends objection_1.Model {
    static get tableName() {
        return 'oracle_prices';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'marketId',
                'price',
                'effectiveAt',
                'effectiveAtHeight',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                marketId: { type: 'integer' },
                price: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                effectiveAt: { type: 'string', format: 'date-time' },
                effectiveAtHeight: { type: 'string', pattern: validators_1.IntegerPattern },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            marketId: 'integer',
            price: 'string',
            effectiveAt: 'date-time',
            effectiveAtHeight: 'string',
        };
    }
}
exports.default = OraclePriceModel;
OraclePriceModel.relationMappings = {
    market: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'market-model'),
        join: {
            from: 'oracle_prices.marketId',
            to: 'markets.id',
        },
    },
    blocks: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'oracle_prices.effectiveAtHeight',
            to: 'blocks.blockHeight',
        },
    },
};
