// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/order-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
const types_1 = require("../types");
const base_model_1 = __importDefault(require("./base-model"));
class OrderModel extends base_model_1.default {
    static get tableName() {
        return 'orders';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'subaccountId',
                'clientId',
                'clobPairId',
                'side',
                'size',
                'totalFilled',
                'price',
                'type',
                'status',
                'timeInForce',
                'reduceOnly',
                'orderFlags',
                'createdAtHeight',
                'clientMetadata',
                'triggerPrice',
                'updatedAt',
                'updatedAtHeight',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                subaccountId: { type: 'string', format: 'uuid' },
                clientId: { type: 'string', pattern: validators_1.IntegerPattern },
                clobPairId: { type: 'string', pattern: validators_1.IntegerPattern },
                side: { type: 'string', enum: [...Object.values(types_1.OrderSide)] },
                size: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                totalFilled: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                price: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                type: { type: 'string', enum: [...Object.values(types_1.OrderType)] },
                status: { type: 'string', enum: [...Object.values(types_1.OrderStatus)] },
                timeInForce: { type: 'string', enum: [...Object.values(types_1.TimeInForce)] },
                reduceOnly: { type: 'boolean' },
                orderFlags: { type: 'string', pattern: validators_1.IntegerPattern },
                goodTilBlock: { type: ['string', 'null'], default: null, pattern: validators_1.IntegerPattern },
                goodTilBlockTime: { type: ['string', 'null'], default: null, format: 'date-time' },
                createdAtHeight: { type: ['string', 'null'], default: null, pattern: validators_1.IntegerPattern },
                clientMetadata: { type: 'string', pattern: validators_1.IntegerPattern },
                triggerPrice: { type: ['string', 'null'], default: null, pattern: validators_1.NonNegativeNumericPattern },
                updatedAt: { type: 'string', format: 'date-time' },
                updatedAtHeight: { type: 'string', pattern: validators_1.IntegerPattern },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            subaccountId: 'string',
            clientId: 'string',
            clobPairId: 'string',
            side: 'string',
            size: 'string',
            totalFilled: 'string',
            price: 'string',
            type: 'string',
            status: 'string',
            timeInForce: 'string',
            reduceOnly: 'boolean',
            orderFlags: 'string',
            goodTilBlock: 'string',
            goodTilBlockTime: 'date-time',
            createdAtHeight: 'string',
            clientMetadata: 'string',
            triggerPrice: 'string',
            updatedAt: 'date-time',
            updatedAtHeight: 'string',
        };
    }
}
exports.default = OrderModel;
OrderModel.relationMappings = {
    subaccounts: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'subaccount-model'),
        join: {
            from: 'orders.subaccountId',
            to: 'subaccounts.id',
        },
    },
    fills: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'fill-model'),
        join: {
            from: 'orders.id',
            to: 'fills.orderId',
        },
    },
    blocks: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'orders.createdAtHeight',
            to: 'blocks.blockHeight',
        },
    },
};
