// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/perpetual-market-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
const types_1 = require("../types");
class PerpetualMarketModel extends objection_1.Model {
    static get tableName() {
        return 'perpetual_markets';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'clobPairId',
                'ticker',
                'marketId',
                'status',
                'priceChange24H',
                'volume24H',
                'trades24H',
                'nextFundingRate',
                'openInterest',
                'quantumConversionExponent',
                'atomicResolution',
                'subticksPerTick',
                'stepBaseQuantums',
                'liquidityTierId',
                'marketType',
            ],
            properties: {
                id: { type: 'string', pattern: validators_1.IntegerPattern },
                clobPairId: { type: 'string', pattern: validators_1.IntegerPattern },
                ticker: { type: 'string' },
                marketId: { type: 'integer' },
                status: { type: 'string', enum: [...Object.values(types_1.PerpetualMarketStatus)] },
                priceChange24H: { type: 'string', pattern: validators_1.NumericPattern },
                volume24H: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                trades24H: { type: 'integer' },
                nextFundingRate: { type: 'string', pattern: validators_1.NumericPattern },
                openInterest: { type: 'string', pattern: validators_1.NumericPattern },
                quantumConversionExponent: { type: 'integer' },
                atomicResolution: { type: 'integer' },
                subticksPerTick: { type: 'integer' },
                stepBaseQuantums: { type: 'integer' },
                liquidityTierId: { type: 'integer' },
                marketType: { type: 'string' },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            clobPairId: 'string',
            ticker: 'string',
            marketId: 'integer',
            status: 'string',
            priceChange24H: 'string',
            volume24H: 'string',
            trades24H: 'integer',
            nextFundingRate: 'string',
            openInterest: 'string',
            quantumConversionExponent: 'integer',
            atomicResolution: 'integer',
            subticksPerTick: 'integer',
            stepBaseQuantums: 'integer',
            liquidityTierId: 'integer',
            marketType: 'string',
        };
    }
}
exports.default = PerpetualMarketModel;
PerpetualMarketModel.relationMappings = {
    perpetualPosition: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-position-model'),
        join: {
            from: 'perpetual_markets.id',
            to: 'perpetual_positions.perpetualId',
        },
    },
    market: {
        relation: objection_1.Model.HasOneRelation,
        modelClass: path_1.default.join(__dirname, 'market-model'),
        join: {
            from: 'perpetual_markets.marketId',
            to: 'markets.id',
        },
    },
    liquidityTiers: {
        relation: objection_1.Model.HasOneRelation,
        modelClass: path_1.default.join(__dirname, 'liquidity-tiers-model'),
        join: {
            from: 'perpetual_markets.liquidityTierId',
            to: 'liquidity_tiers.id',
        },
    },
};
