// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/perpetual-position-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
const types_1 = require("../types");
class PerpetualPositionModel extends objection_1.Model {
    static get tableName() {
        return 'perpetual_positions';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'subaccountId',
                'perpetualId',
                'side',
                'status',
                'size',
                'maxSize',
                'entryPrice',
                'sumOpen',
                'createdAt',
                'createdAtHeight',
                'openEventId',
                'lastEventId',
                'settledFunding',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                subaccountId: { type: 'string', format: 'uuid' },
                perpetualId: { type: 'string', pattern: validators_1.IntegerPattern },
                side: { type: 'string', enum: [...Object.values(types_1.PositionSide)] },
                status: { type: 'string', enum: [...Object.values(types_1.PerpetualPositionStatus)] },
                size: { type: 'string', pattern: validators_1.NumericPattern },
                maxSize: { type: 'string', pattern: validators_1.NumericPattern },
                entryPrice: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                exitPrice: { type: ['string', 'null'], pattern: validators_1.NonNegativeNumericPattern, default: null },
                sumOpen: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                sumClose: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                createdAt: { type: 'string', format: 'date-time' },
                closedAt: { type: ['string', 'null'], format: 'date-time', default: null },
                createdAtHeight: { type: 'string', pattern: validators_1.IntegerPattern },
                closedAtHeight: { type: ['string', 'null'], default: null, pattern: validators_1.IntegerPattern },
                settledFunding: { type: 'string', pattern: validators_1.NumericPattern },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            subaccountId: 'string',
            perpetualId: 'string',
            side: 'string',
            status: 'string',
            size: 'string',
            maxSize: 'string',
            entryPrice: 'string',
            exitPrice: 'string',
            sumOpen: 'string',
            sumClose: 'string',
            createdAt: 'date-time',
            closedAt: 'date-time',
            createdAtHeight: 'string',
            closedAtHeight: 'string',
            openEventId: 'hex-string',
            closeEventId: 'hex-string',
            lastEventId: 'hex-string',
            settledFunding: 'string',
        };
    }
}
exports.default = PerpetualPositionModel;
PerpetualPositionModel.relationMappings = {
    subaccount: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'subaccount-model'),
        join: {
            from: 'perpetual_positions.subaccountId',
            to: 'subaccounts.id',
        },
    },
    perpetualMarket: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-market-model'),
        join: {
            from: 'perpetual_positions.perpetualId',
            to: 'perpetual_markets.id',
        },
    },
    openTendermintEvent: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'tendermint-event-model'),
        join: {
            from: 'perpetual_positions.openEventId',
            to: 'tendermint_events.id',
        },
    },
    closeTendermintEvent: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'tendermint-event-model'),
        join: {
            from: 'perpetual_positions.closeEventId',
            to: 'tendermint_events.id',
        },
    },
    lastTendermintEvent: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'tendermint-event-model'),
        join: {
            from: 'perpetual_positions.lastEventId',
            to: 'tendermint_events.id',
        },
    },
};
