// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/pnl-ticks-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class PnlTicksModel extends objection_1.Model {
    static get tableName() {
        return 'pnl_ticks';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'subaccountId',
                'equity',
                'totalPnl',
                'netTransfers',
                'createdAt',
                'blockHeight',
                'blockTime',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                subaccountId: { type: 'string', format: 'uuid' },
                equity: { type: 'string', pattern: validators_1.NumericPattern },
                totalPnl: { type: 'string', pattern: validators_1.NumericPattern },
                netTransfers: { type: 'string', pattern: validators_1.NumericPattern },
                createdAt: { type: 'string', format: 'date-time' },
                blockHeight: { type: 'string', pattern: validators_1.IntegerPattern },
                blockTime: { type: 'string', format: 'date-time' },
            },
        };
    }
}
exports.default = PnlTicksModel;
PnlTicksModel.relationMappings = {
    subaccount: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'subaccount-model'),
        join: {
            from: 'pnl_ticks.subaccountId',
            to: 'subaccounts.id',
        },
    },
    block_height: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'pnl_ticks.blockHeight',
            to: 'blocks.blockHeight',
        },
    },
    block_time: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'pnl_ticks.blockTime',
            to: 'blocks.time',
        },
    },
};
