// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/subaccount-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const base_model_1 = __importDefault(require("./base-model"));
class SubaccountModel extends base_model_1.default {
    static get tableName() {
        return 'subaccounts';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'address',
                'subaccountNumber',
                'updatedAt',
                'updatedAtHeight',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                address: { type: 'string' },
                subaccountNumber: { type: 'integer' },
                updatedAt: { type: 'string', format: 'date-time' },
                updatedAtHeight: { type: 'string' },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            address: 'string',
            subaccountNumber: 'integer',
            updatedAt: 'date-time',
            updatedAtHeight: 'string',
        };
    }
}
exports.default = SubaccountModel;
SubaccountModel.relationMappings = {
    orders: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'order-model'),
        join: {
            from: 'subaccounts.id',
            to: 'orders.subaccountId',
        },
    },
    height: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'subaccounts.updatedAtHeight',
            to: 'blocks.blockHeight',
        },
    },
    perpetualPositions: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-position-model'),
        join: {
            from: 'subaccounts.id',
            to: 'perpetual_positions.subaccountId',
        },
    },
    fills: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'fill-model'),
        join: {
            from: 'subaccounts.id',
            to: 'fills.subaccountId',
        },
    },
    assetPosition: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'asset-position-model'),
        join: {
            from: 'subaccounts.id',
            to: 'asset_positions.subaccountId',
        },
    },
    fromTransfer: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'transfer-model'),
        join: {
            from: 'subaccounts.id',
            to: 'transfers.senderSubaccountId',
        },
    },
    toTransfer: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'transfer-model'),
        join: {
            from: 'subaccounts.id',
            to: 'transfers.recipientSubaccountId',
        },
    },
};
