// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/tendermint-event-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class TendermintEventModel extends objection_1.Model {
    static get tableName() {
        return 'tendermint_events';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'blockHeight',
                'transactionIndex',
                'eventIndex',
            ],
            properties: {
                blockHeight: { type: 'string', pattern: validators_1.IntegerPattern },
                transactionIndex: { type: 'integer' },
                eventIndex: { type: 'integer' },
            },
        };
    }
}
exports.default = TendermintEventModel;
TendermintEventModel.relationMappings = {
    openPerpetualPosition: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-position-model'),
        join: {
            from: 'tendermint_events.id',
            to: 'perpetual_positions.openEventId',
        },
    },
    closePerpetualPosition: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-position-model'),
        join: {
            from: 'tendermint_events.id',
            to: 'perpetual_positions.closeEventId',
        },
    },
    lastPerpetualPosition: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'perpetual-position-model'),
        join: {
            from: 'tendermint_events.id',
            to: 'perpetual_positions.lastEventId',
        },
    },
    fills: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'fill-model'),
        join: {
            from: 'tendermint_events.id',
            to: 'fills.eventId',
        },
    },
    blocks: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'tendermint_events.blockHeight',
            to: 'blocks.blockHeight',
        },
    },
    transfers: {
        relation: objection_1.Model.HasManyRelation,
        modelClass: path_1.default.join(__dirname, 'transfer-model'),
        join: {
            from: 'tendermint_events.id',
            to: 'transfers.eventId',
        },
    },
};
