// @dydxprotocol-indexer/postgres@18.19.39/build/src/models/transfer-model.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const objection_1 = require("objection");
const validators_1 = require("../lib/validators");
class TransferModel extends objection_1.Model {
    static get tableName() {
        return 'transfers';
    }
    static get idColumn() {
        return 'id';
    }
    static get jsonSchema() {
        return {
            type: 'object',
            required: [
                'id',
                'senderSubaccountId',
                'recipientSubaccountId',
                'senderWalletAddress',
                'recipientWalletAddress',
                'assetId',
                'size',
                'eventId',
                'transactionHash',
                'createdAt',
                'createdAtHeight',
            ],
            properties: {
                id: { type: 'string', format: 'uuid' },
                senderSubaccountId: { type: ['string', 'null'], default: null, format: 'uuid' },
                recipientSubaccountId: { type: ['string', 'null'], default: null, format: 'uuid' },
                senderWalletAddress: { type: ['string', 'null'], default: null },
                recipientWalletAddress: { type: ['string', 'null'], default: null },
                assetId: { type: 'string', pattern: validators_1.IntegerPattern },
                size: { type: 'string', pattern: validators_1.NonNegativeNumericPattern },
                transactionHash: { type: 'string' },
                createdAt: { type: 'string', format: 'date-time' },
                createdAtHeight: { type: 'string', pattern: validators_1.IntegerPattern },
            },
        };
    }
    /**
     * A mapping from column name to JSON conversion expected.
     * See getSqlConversionForDydxModelTypes for valid conversions.
     *
     * TODO(IND-239): Ensure that jsonSchema() / sqlToJsonConversions() / model fields match.
     */
    static get sqlToJsonConversions() {
        return {
            id: 'string',
            senderSubaccountId: 'string',
            recipientSubaccountId: 'string',
            senderWalletAddress: 'string',
            recipientWalletAddress: 'string',
            assetId: 'string',
            size: 'string',
            eventId: 'hex-string',
            transactionHash: 'string',
            createdAt: 'date-time',
            createdAtHeight: 'string',
        };
    }
}
exports.default = TransferModel;
TransferModel.relationMappings = {
    recipientSubaccount: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'subaccount-model'),
        join: {
            from: 'transfers.recipientSubaccountId',
            to: 'subaccounts.id',
        },
    },
    senderSubaccount: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'subaccount-model'),
        join: {
            from: 'transfers.senderSubaccountId',
            to: 'subaccounts.id',
        },
    },
    asset: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'asset-model'),
        join: {
            from: 'transfers.assetId',
            to: 'assets.id',
        },
    },
    tendermintEvents: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'tendermint-event-model'),
        join: {
            from: 'transfers.eventId',
            to: 'tendermint_events.id',
        },
    },
    blocks: {
        relation: objection_1.Model.BelongsToOneRelation,
        modelClass: path_1.default.join(__dirname, 'block-model'),
        join: {
            from: 'transfers.createdAtHeight',
            to: 'blocks.blockHeight',
        },
    },
};
