// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/perpetuals/tx.rpc.msg.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClientImpl = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
const tx_1 = require("./tx");
class MsgClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.addPremiumVotes = this.addPremiumVotes.bind(this);
        this.createPerpetual = this.createPerpetual.bind(this);
        this.setLiquidityTier = this.setLiquidityTier.bind(this);
        this.updatePerpetualParams = this.updatePerpetualParams.bind(this);
        this.updateParams = this.updateParams.bind(this);
    }
    addPremiumVotes(request) {
        const data = tx_1.MsgAddPremiumVotes.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.perpetuals.Msg", "AddPremiumVotes", data);
        return promise.then(data => tx_1.MsgAddPremiumVotesResponse.decode(new _m0.Reader(data)));
    }
    createPerpetual(request) {
        const data = tx_1.MsgCreatePerpetual.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.perpetuals.Msg", "CreatePerpetual", data);
        return promise.then(data => tx_1.MsgCreatePerpetualResponse.decode(new _m0.Reader(data)));
    }
    setLiquidityTier(request) {
        const data = tx_1.MsgSetLiquidityTier.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.perpetuals.Msg", "SetLiquidityTier", data);
        return promise.then(data => tx_1.MsgSetLiquidityTierResponse.decode(new _m0.Reader(data)));
    }
    updatePerpetualParams(request) {
        const data = tx_1.MsgUpdatePerpetualParams.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.perpetuals.Msg", "UpdatePerpetualParams", data);
        return promise.then(data => tx_1.MsgUpdatePerpetualParamsResponse.decode(new _m0.Reader(data)));
    }
    updateParams(request) {
        const data = tx_1.MsgUpdateParams.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.perpetuals.Msg", "UpdateParams", data);
        return promise.then(data => tx_1.MsgUpdateParamsResponse.decode(new _m0.Reader(data)));
    }
}
exports.MsgClientImpl = MsgClientImpl;
