// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/asset-position-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findUsdcPositionForSubaccounts = exports.findById = exports.upsert = exports.findAll = exports.uuid = void 0;
const big_js_1 = __importDefault(require("big.js"));
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const uuid_1 = require("../helpers/uuid");
const asset_position_model_1 = __importDefault(require("../models/asset-position-model"));
const types_1 = require("../types");
function uuid(subaccountId, assetId) {
    // TODO(IND-483): Fix all uuid string substitutions to use Array.join.
    return (0, uuid_1.getUuid)(Buffer.from(`${subaccountId}-${assetId}`, constants_1.BUFFER_ENCODING_UTF_8));
}
exports.uuid = uuid;
async function findAll({ limit, id, subaccountId, assetId, size, isLong, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        limit,
        id,
        subaccountId,
        assetId,
        size,
        isLong,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(asset_position_model_1.default, options);
    if (id) {
        baseQuery = baseQuery.whereIn(types_1.AssetPositionColumns.id, id);
    }
    if (subaccountId) {
        baseQuery = baseQuery.whereIn(types_1.AssetPositionColumns.subaccountId, subaccountId);
    }
    if (assetId) {
        baseQuery = baseQuery.whereIn(types_1.AssetPositionColumns.assetId, assetId);
    }
    if (size) {
        baseQuery = baseQuery.where(types_1.AssetPositionColumns.size, size);
    }
    if (isLong) {
        baseQuery = baseQuery.where(types_1.AssetPositionColumns.isLong, isLong);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.AssetPositionColumns.assetId, types_1.Ordering.ASC);
    }
    if (limit) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function upsert(assetPositionToUpsert, options = { txId: undefined }) {
    const { subaccountId, assetId, } = assetPositionToUpsert;
    const createdUuid = uuid(subaccountId, assetId);
    const assetPositions = await asset_position_model_1.default.query(transaction_1.default.get(options.txId)).upsert({
        ...assetPositionToUpsert,
        id: createdUuid,
    }).returning('*');
    // should only ever be one asset position
    return assetPositions[0];
}
exports.upsert = upsert;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(asset_position_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
function convertToSubaccountUsdcMap(usdcPositions) {
    return lodash_1.default.mapValues(lodash_1.default.keyBy(usdcPositions, 'subaccountId'), (asset) => {
        return (0, big_js_1.default)(asset.isLong ? asset.size : -asset.size);
    });
}
async function findUsdcPositionForSubaccounts(subaccountIds, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const positions = await findAll({
        subaccountId: subaccountIds,
        assetId: [constants_1.USDC_ASSET_ID],
    }, [], options);
    if (positions.length === 0) {
        return {};
    }
    return convertToSubaccountUsdcMap(positions);
}
exports.findUsdcPositionForSubaccounts = findUsdcPositionForSubaccounts;
