// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/block-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLatest = exports.findByBlockHeight = exports.create = exports.findAll = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const block_model_1 = __importDefault(require("../models/block-model"));
const types_1 = require("../types");
async function findAll({ blockHeight, createdOnOrAfter, limit, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        blockHeight,
        createdOnOrAfter,
        limit,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(block_model_1.default, options);
    if (blockHeight !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.BlockColumns.blockHeight, blockHeight);
    }
    if (createdOnOrAfter !== undefined) {
        baseQuery = baseQuery.where(types_1.BlockColumns.time, '>=', createdOnOrAfter);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.BlockColumns.blockHeight, types_1.Ordering.ASC);
    }
    if (limit !== undefined) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(blockToCreate, options = { txId: undefined }) {
    return block_model_1.default.query(transaction_1.default.get(options.txId)).insert(blockToCreate).returning('*');
}
exports.create = create;
async function findByBlockHeight(blockHeight, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(block_model_1.default, options);
    return baseQuery
        .findById(blockHeight)
        .returning('*');
}
exports.findByBlockHeight = findByBlockHeight;
async function getLatest(options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(block_model_1.default, options);
    const results = await baseQuery
        .orderBy(types_1.BlockColumns.blockHeight, types_1.Ordering.DESC)
        .limit(1)
        .returning('*');
    const latestBlock = results[0];
    if (latestBlock === undefined) {
        base_1.logger.error({
            at: 'block-table#getLatest',
            message: 'Unable to find latest block',
        });
        throw new Error('Unable to find latest block');
    }
    return latestBlock;
}
exports.getLatest = getLatest;
