// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/candle-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findCandlesMap = exports.findLatest = exports.findById = exports.update = exports.create = exports.findAll = exports.uuid = void 0;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const uuid_1 = require("../helpers/uuid");
const candle_model_1 = __importDefault(require("../models/candle-model"));
const types_1 = require("../types");
function uuid(startedAt, ticker, resolution) {
    // TODO(IND-483): Fix all uuid string substitutions to use Array.join.
    return (0, uuid_1.getUuid)(Buffer.from(`${startedAt}-${ticker}-${resolution}`, constants_1.BUFFER_ENCODING_UTF_8));
}
exports.uuid = uuid;
/**
 * Find all candles that match the given query config. fromIso is inclusive and toISO is exclusive.
 */
async function findAll({ limit, id, ticker, resolution, fromISO, toISO, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        id,
        ticker,
        resolution,
        fromISO,
        toISO,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(candle_model_1.default, options);
    if (id) {
        baseQuery = baseQuery.whereIn(types_1.CandleColumns.id, id);
    }
    if (ticker) {
        baseQuery = baseQuery.whereIn(types_1.CandleColumns.ticker, ticker);
    }
    if (resolution) {
        baseQuery = baseQuery.where(types_1.CandleColumns.resolution, resolution);
    }
    if (fromISO) {
        baseQuery = baseQuery.where(types_1.CandleColumns.startedAt, '>=', fromISO);
    }
    if (toISO) {
        baseQuery = baseQuery.where(types_1.CandleColumns.startedAt, '<', toISO);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.CandleColumns.ticker, types_1.Ordering.DESC).orderBy(types_1.CandleColumns.resolution, types_1.Ordering.DESC).orderBy(types_1.CandleColumns.startedAt, types_1.Ordering.DESC);
    }
    if (limit) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(candle, options = { txId: undefined }) {
    return candle_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        id: uuid(candle.startedAt, candle.ticker, candle.resolution),
        ...candle,
    }).returning('*');
}
exports.create = create;
async function update({ id, ...fields }, options = { txId: undefined }) {
    const candle = await candle_model_1.default.query(transaction_1.default.get(options.txId)).findById(id);
    const updatedCandle = await candle.$query().patch(fields).returning('*');
    // The objection types mistakenly think the query returns an array of candles.
    return updatedCandle;
}
exports.update = update;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(candle_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
async function findLatest(ticker, resolution, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const candles = await findAll({
        ticker: [ticker],
        resolution,
        limit: 1,
    }, [types_1.QueryableField.TICKER, types_1.QueryableField.RESOLUTION], {
        ...options,
        // Ordered by 'startedAt' descending, so the most recent candle will be only candle returned
        orderBy: [[types_1.CandleColumns.startedAt, types_1.Ordering.DESC]],
    });
    if (candles.length === 1) {
        return candles[0];
    }
    return undefined;
}
exports.findLatest = findLatest;
async function findCandlesMap(tickers, resolutions, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const candlesMap = {};
    await Promise.all(lodash_1.default.map(tickers, async (ticker) => {
        candlesMap[ticker] = {};
        const findLatestCandles = resolutions.map((resolution) => findLatest(ticker, resolution, options));
        // Map each resolution to its respective candle
        const allLatestCandles = await Promise.all(findLatestCandles);
        lodash_1.default.forEach(allLatestCandles, (candle) => {
            if (candle !== undefined) {
                candlesMap[ticker][candle.resolution] = candle;
            }
        });
    }));
    return candlesMap;
}
exports.findCandlesMap = findCandlesMap;
