// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/compliance-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bulkUpsert = exports.findByAddressAndProvider = exports.upsert = exports.update = exports.create = exports.findAll = void 0;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const knex_1 = require("../helpers/knex");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const compliance_data_model_1 = __importDefault(require("../models/compliance-data-model"));
const types_1 = require("../types");
const compliance_data_types_1 = require("../types/compliance-data-types");
async function findAll({ address, updatedBeforeOrAt, provider, blocked, limit, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        address,
        updatedBeforeOrAt,
        provider,
        blocked,
        limit,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(compliance_data_model_1.default, options);
    if (address !== undefined) {
        baseQuery = baseQuery.whereIn(compliance_data_types_1.ComplianceDataColumns.address, address);
    }
    if (updatedBeforeOrAt !== undefined) {
        baseQuery = baseQuery.where(compliance_data_types_1.ComplianceDataColumns.updatedAt, '<=', updatedBeforeOrAt);
    }
    if (provider !== undefined) {
        baseQuery = baseQuery.where(compliance_data_types_1.ComplianceDataColumns.provider, provider);
    }
    if (blocked !== undefined) {
        baseQuery = baseQuery.where(compliance_data_types_1.ComplianceDataColumns.blocked, blocked);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(compliance_data_types_1.ComplianceDataColumns.updatedAt, types_1.Ordering.ASC);
    }
    if (limit !== undefined) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(complianceDataToCreate, options = { txId: undefined }) {
    return compliance_data_model_1.default.query(transaction_1.default.get(options.txId)).insert(complianceDataToCreate).returning('*');
}
exports.create = create;
async function update({ address, provider, ...fields }, options = { txId: undefined }) {
    const complianceData = await compliance_data_model_1.default.query(transaction_1.default.get(options.txId)).findById([address, provider]);
    const updatedComplianceData = await complianceData
        .$query()
        .patch(fields)
        .returning('*');
    // The objection types mistakenly think the query returns an array of ComplianceData.
    return updatedComplianceData;
}
exports.update = update;
async function upsert(complianceDataToUpsert, options = { txId: undefined }) {
    const complianceData = await findByAddressAndProvider(complianceDataToUpsert.address, complianceDataToUpsert.provider);
    if (complianceData === undefined) {
        return create({
            ...complianceDataToUpsert,
        }, options);
    }
    const updatedComplianceData = await update({
        ...complianceDataToUpsert,
    }, options);
    if (updatedComplianceData === undefined) {
        throw Error('order must exist after update');
    }
    return updatedComplianceData;
}
exports.upsert = upsert;
async function findByAddressAndProvider(address, provider, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(compliance_data_model_1.default, options);
    return baseQuery
        .findById([address, provider])
        .returning('*');
}
exports.findByAddressAndProvider = findByAddressAndProvider;
async function bulkUpsert(complianceObjects, options = { txId: undefined }) {
    if (complianceObjects.length === 0) {
        return;
    }
    complianceObjects.forEach((complianceObject) => (0, stores_helpers_1.verifyAllInjectableVariables)(Object.values(complianceObject)));
    const columns = lodash_1.default.keys(complianceObjects[0]);
    const rows = (0, stores_helpers_1.setBulkRowsForUpdate)({
        objectArray: complianceObjects,
        columns,
        booleanColumns: [
            compliance_data_types_1.ComplianceDataColumns.blocked,
        ],
        numericColumns: [
            compliance_data_types_1.ComplianceDataColumns.riskScore,
        ],
        stringColumns: [
            compliance_data_types_1.ComplianceDataColumns.address,
            compliance_data_types_1.ComplianceDataColumns.chain,
            compliance_data_types_1.ComplianceDataColumns.provider,
        ],
        timestampColumns: [
            compliance_data_types_1.ComplianceDataColumns.updatedAt,
        ],
    });
    const query = (0, stores_helpers_1.generateBulkUpsertString)({
        table: compliance_data_model_1.default.tableName,
        objectRows: rows,
        columns,
        uniqueIdentifiers: [compliance_data_types_1.ComplianceDataColumns.address, compliance_data_types_1.ComplianceDataColumns.provider],
    });
    const transaction = transaction_1.default.get(options.txId);
    return transaction
        ? knex_1.knexPrimary.raw(query).transacting(transaction)
        : knex_1.knexPrimary.raw(query);
}
exports.bulkUpsert = bulkUpsert;
