// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/funding-index-updates-table.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findFundingIndexMap = exports.findMostRecentMarketFundingIndexUpdate = exports.findById = exports.create = exports.findAll = exports.uuid = void 0;
const big_js_1 = __importDefault(require("big.js"));
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const uuid_1 = require("../helpers/uuid");
const funding_index_updates_model_1 = __importDefault(require("../models/funding-index-updates-model"));
const types_1 = require("../types");
const PerpetualMarketTable = __importStar(require("./perpetual-market-table"));
function uuid(blockHeight, eventId, perpetualId) {
    // TODO(IND-483): Fix all uuid string substitutions to use Array.join.
    return (0, uuid_1.getUuid)(Buffer.from(`${blockHeight}-${eventId.toString('hex')}-${perpetualId}`, constants_1.BUFFER_ENCODING_UTF_8));
}
exports.uuid = uuid;
async function findAll({ limit, id, perpetualId, eventId, effectiveAt, effectiveAtHeight, effectiveBeforeOrAt, effectiveBeforeOrAtHeight, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        limit,
        id,
        perpetualId,
        eventId,
        effectiveAt,
        effectiveAtHeight,
        effectiveBeforeOrAt,
        effectiveBeforeOrAtHeight,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(funding_index_updates_model_1.default, options);
    if (id !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.FundingIndexUpdatesColumns.id, id);
    }
    if (perpetualId !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.FundingIndexUpdatesColumns.perpetualId, perpetualId);
    }
    if (eventId !== undefined) {
        baseQuery = baseQuery.where(types_1.FundingIndexUpdatesColumns.eventId, eventId);
    }
    if (effectiveAt !== undefined) {
        baseQuery = baseQuery.where(types_1.FundingIndexUpdatesColumns.effectiveAt, effectiveAt);
    }
    if (effectiveAtHeight !== undefined) {
        baseQuery = baseQuery.where(types_1.FundingIndexUpdatesColumns.effectiveAtHeight, effectiveAtHeight);
    }
    if (effectiveBeforeOrAt !== undefined) {
        baseQuery = baseQuery.where(types_1.FundingIndexUpdatesColumns.effectiveAt, '<=', effectiveBeforeOrAt);
    }
    if (effectiveBeforeOrAtHeight !== undefined) {
        baseQuery = baseQuery.where(types_1.FundingIndexUpdatesColumns.effectiveAtHeight, '<=', effectiveBeforeOrAtHeight);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.FundingIndexUpdatesColumns.effectiveAtHeight, types_1.Ordering.DESC).orderBy(types_1.FundingIndexUpdatesColumns.eventId, types_1.Ordering.DESC);
    }
    if (limit !== undefined) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(fundingIndexUpdateToCreate, options = { txId: undefined }) {
    return funding_index_updates_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        id: uuid(fundingIndexUpdateToCreate.effectiveAtHeight, fundingIndexUpdateToCreate.eventId, fundingIndexUpdateToCreate.perpetualId),
        ...fundingIndexUpdateToCreate,
    }).returning('*');
}
exports.create = create;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(funding_index_updates_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
async function findMostRecentMarketFundingIndexUpdate(perpetualId, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(funding_index_updates_model_1.default, options);
    const fundingIndexUpdates = await baseQuery
        .where(types_1.FundingIndexUpdatesColumns.perpetualId, perpetualId)
        .orderBy(types_1.FundingIndexUpdatesColumns.effectiveAtHeight, types_1.Ordering.DESC)
        .limit(1)
        .returning('*');
    if (fundingIndexUpdates.length === 0) {
        return undefined;
    }
    return fundingIndexUpdates[0];
}
exports.findMostRecentMarketFundingIndexUpdate = findMostRecentMarketFundingIndexUpdate;
async function findFundingIndexMap(effectiveBeforeOrAtHeight, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    // TODO(IND-39): Remove this default when the database is seeded using events emitted from
    // protocol during genesis.
    // Default funding index per perpetual market is 0.
    const perpetualMarkets = await PerpetualMarketTable.findAll({}, [], options);
    const initialFundingIndexMap = lodash_1.default.reduce(perpetualMarkets, (acc, perpetualMarket) => {
        acc[perpetualMarket.id] = (0, big_js_1.default)(0);
        return acc;
    }, {});
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(funding_index_updates_model_1.default, options);
    const fundingIndexUpdates = await baseQuery
        .distinctOn(types_1.FundingIndexUpdatesColumns.perpetualId)
        .where(types_1.FundingIndexUpdatesColumns.effectiveAtHeight, '<=', effectiveBeforeOrAtHeight)
        .orderBy(types_1.FundingIndexUpdatesColumns.perpetualId)
        .orderBy(types_1.FundingIndexUpdatesColumns.effectiveAtHeight, types_1.Ordering.DESC)
        .returning('*');
    return lodash_1.default.reduce(fundingIndexUpdates, (acc, fundingIndexUpdate) => {
        acc[fundingIndexUpdate.perpetualId] = (0, big_js_1.default)(fundingIndexUpdate.fundingIndex);
        return acc;
    }, initialFundingIndexMap);
}
exports.findFundingIndexMap = findFundingIndexMap;
