// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/market-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMarketsMap = exports.findByPair = exports.findById = exports.update = exports.create = exports.findAll = void 0;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const market_model_1 = __importDefault(require("../models/market-model"));
const types_1 = require("../types");
async function findAll({ limit, id, pair, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        id,
        pair,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(market_model_1.default, options);
    if (id) {
        baseQuery = baseQuery.whereIn(types_1.MarketColumns.id, id);
    }
    if (pair) {
        baseQuery = baseQuery.whereIn(types_1.MarketColumns.pair, pair);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.MarketColumns.id, types_1.Ordering.ASC);
    }
    if (limit) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(MarketToCreate, options = { txId: undefined }) {
    return market_model_1.default.query(transaction_1.default.get(options.txId)).insert(MarketToCreate).returning('*');
}
exports.create = create;
async function update({ id, ...fields }, options = { txId: undefined }) {
    const market = await market_model_1.default.query(transaction_1.default.get(options.txId)).findById(id);
    const updatedMarket = await market
        .$query(transaction_1.default.get(options.txId))
        .patch(fields).returning('*');
    // The objection types mistakenly think the query returns an array of markets.
    return updatedMarket;
}
exports.update = update;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(market_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
async function findByPair(pair, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(market_model_1.default, options);
    const markets = await baseQuery
        .where(types_1.MarketColumns.pair, pair)
        .returning('*');
    if (markets.length === 0) {
        return undefined;
    }
    return markets[0];
}
exports.findByPair = findByPair;
async function getMarketsMap(options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const markets = await findAll({}, [], options);
    return lodash_1.default.keyBy(markets, types_1.MarketColumns.id);
}
exports.getMarketsMap = getMarketsMap;
