// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/oracle-price-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findLatestPrices = exports.getLatestPrices = exports.getPricesFrom24hAgo = exports.findMostRecentMarketOraclePrice = exports.findOraclePricesInReverseChronologicalOrder = exports.findById = exports.create = exports.findAll = exports.uuid = void 0;
const lodash_1 = __importDefault(require("lodash"));
const luxon_1 = require("luxon");
const constants_1 = require("../constants");
const knex_1 = require("../helpers/knex");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const uuid_1 = require("../helpers/uuid");
const oracle_price_model_1 = __importDefault(require("../models/oracle-price-model"));
const types_1 = require("../types");
function uuid(marketId, height) {
    // TODO(IND-483): Fix all uuid string substitutions to use Array.join.
    return (0, uuid_1.getUuid)(Buffer.from(`${marketId.toString()}-${height}`, constants_1.BUFFER_ENCODING_UTF_8));
}
exports.uuid = uuid;
async function findAll({ limit, id, marketId, price, effectiveAt, effectiveAtHeight, effectiveBeforeOrAt, effectiveBeforeOrAtHeight, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        limit,
        id,
        marketId,
        price,
        effectiveAt,
        effectiveAtHeight,
        effectiveBeforeOrAt,
        effectiveBeforeOrAtHeight,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(oracle_price_model_1.default, options);
    if (id !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.OraclePriceColumns.id, id);
    }
    if (marketId !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.OraclePriceColumns.marketId, marketId);
    }
    if (price !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.OraclePriceColumns.price, price);
    }
    if (effectiveAt !== undefined) {
        baseQuery = baseQuery.where(types_1.OraclePriceColumns.effectiveAt, effectiveAt);
    }
    if (effectiveAtHeight !== undefined) {
        baseQuery = baseQuery.where(types_1.OraclePriceColumns.effectiveAtHeight, effectiveAtHeight);
    }
    if (effectiveBeforeOrAt !== undefined) {
        baseQuery = baseQuery.where(types_1.OraclePriceColumns.effectiveAt, '<=', effectiveBeforeOrAt);
    }
    if (effectiveBeforeOrAtHeight !== undefined) {
        baseQuery = baseQuery.where(types_1.OraclePriceColumns.effectiveAtHeight, '<=', effectiveBeforeOrAtHeight);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.OraclePriceColumns.effectiveAtHeight, types_1.Ordering.DESC);
    }
    if (limit !== undefined) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(oraclePriceToCreate, options = { txId: undefined }) {
    return oracle_price_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        id: uuid(oraclePriceToCreate.marketId, oraclePriceToCreate.effectiveAtHeight),
        ...oraclePriceToCreate,
    }).returning('*');
}
exports.create = create;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(oracle_price_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
async function findOraclePricesInReverseChronologicalOrder(marketId, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(oracle_price_model_1.default, options);
    return baseQuery
        .where(types_1.OraclePriceColumns.marketId, marketId)
        .orderBy(types_1.OraclePriceColumns.effectiveAtHeight, types_1.Ordering.DESC)
        .returning('*');
}
exports.findOraclePricesInReverseChronologicalOrder = findOraclePricesInReverseChronologicalOrder;
async function findMostRecentMarketOraclePrice(marketId, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(oracle_price_model_1.default, options);
    const oraclePrices = await baseQuery
        .where(types_1.OraclePriceColumns.marketId, marketId)
        .orderBy(types_1.OraclePriceColumns.effectiveAtHeight, types_1.Ordering.DESC)
        .limit(1)
        .returning('*');
    if (oraclePrices.length === 0) {
        return undefined;
    }
    return oraclePrices[0];
}
exports.findMostRecentMarketOraclePrice = findMostRecentMarketOraclePrice;
function constructPriceMap(oraclePrices) {
    return lodash_1.default.reduce(oraclePrices, (acc, oraclePrice) => {
        acc[oraclePrice.marketId] = oraclePrice.price;
        return acc;
    }, {});
}
async function findLatestPricesByDateTime(latestDateTimeISO) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(oracle_price_model_1.default, { readReplica: true });
    const innerQuery = (0, stores_helpers_1.setupBaseQuery)(oracle_price_model_1.default, { readReplica: true });
    const subQuery = innerQuery
        .select('marketId')
        .max('effectiveAt as maxEffectiveAt')
        .where('effectiveAt', '<=', latestDateTimeISO)
        .groupBy('marketId');
    const oraclePrices = await baseQuery
        .innerJoin(subQuery.as('sub'), function () {
        this
            .on('oracle_prices.marketId', '=', 'sub.marketId')
            .andOn('oracle_prices.effectiveAt', '=', 'sub.maxEffectiveAt');
    })
        .returning('*');
    return constructPriceMap(oraclePrices);
}
async function getPricesFrom24hAgo() {
    const oneDayAgo = luxon_1.DateTime.utc().minus({ days: 1 }).toISO();
    return findLatestPricesByDateTime(oneDayAgo);
}
exports.getPricesFrom24hAgo = getPricesFrom24hAgo;
async function getLatestPrices() {
    const now = luxon_1.DateTime.utc().toISO();
    return findLatestPricesByDateTime(now);
}
exports.getLatestPrices = getLatestPrices;
async function findLatestPrices(effectiveBeforeOrAtHeight, transaction) {
    const query = `
    SELECT "marketId", "price"
    FROM "oracle_prices"
    WHERE ("marketId", "effectiveAtHeight") IN (
      SELECT "marketId", MAX("effectiveAtHeight")
      FROM "oracle_prices"
      WHERE "effectiveAtHeight" <= '${effectiveBeforeOrAtHeight}'
      GROUP BY "marketId");
  `;
    let result;
    if (transaction === undefined) {
        result = await knex_1.knexReadReplica.getConnection().raw(query);
    }
    else {
        result = await knex_1.knexReadReplica.getConnection().raw(query).transacting(transaction);
    }
    return constructPriceMap(result.rows);
}
exports.findLatestPrices = findLatestPrices;
