// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/order-table.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateStaleOrderStatusByIds = exports.findOpenLongTermOrConditionalOrders = exports.isLongTermOrConditionalOrder = exports.upsert = exports.findBySubaccountIdAndClobPairAfterHeight = exports.findBySubaccountIdAndClobPair = exports.findById = exports.update = exports.create = exports.findAll = exports.orderIdToUuid = exports.uuid = void 0;
const big_js_1 = __importDefault(require("big.js"));
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const uuid_1 = require("../helpers/uuid");
const order_model_1 = __importDefault(require("../models/order-model"));
const types_1 = require("../types");
const SubaccountTable = __importStar(require("./subaccount-table"));
function uuid(subaccountId, clientId, clobPairId, orderFlags) {
    // TODO(IND-483): Fix all uuid string substitutions to use Array.join.
    return (0, uuid_1.getUuid)(Buffer.from(`${subaccountId}-${clientId}-${clobPairId}-${orderFlags}`, constants_1.BUFFER_ENCODING_UTF_8));
}
exports.uuid = uuid;
/**
 * Expects orderId.subaccountId, and orderId.clientId to exist.
 * @param order
 */
function orderIdToUuid(orderId) {
    return uuid(SubaccountTable.subaccountIdToUuid(orderId.subaccountId), orderId.clientId.toString(), orderId.clobPairId.toString(), orderId.orderFlags.toString());
}
exports.orderIdToUuid = orderIdToUuid;
async function findAll({ limit, id, subaccountId, clientId, clobPairId, side, size, totalFilled, price, type, statuses, reduceOnly, orderFlags, goodTilBlockBeforeOrAt, goodTilBlockTimeBeforeOrAt, clientMetadata, triggerPrice, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        limit,
        id,
        subaccountId,
        clientId,
        clobPairId,
        side,
        size,
        totalFilled,
        price,
        type,
        statuses,
        reduceOnly,
        orderFlags,
        goodTilBlockBeforeOrAt,
        goodTilBlockTimeBeforeOrAt,
        clientMetadata,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(order_model_1.default, options);
    if (id !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.OrderColumns.id, id);
    }
    if (subaccountId !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.OrderColumns.subaccountId, subaccountId);
    }
    if (clientId !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.clientId, clientId);
    }
    if (clobPairId !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.clobPairId, clobPairId);
    }
    if (side !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.side, side);
    }
    if (size !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.size, size);
    }
    if (totalFilled !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.totalFilled, totalFilled);
    }
    if (price !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.price, price);
    }
    if (type !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.type, type);
    }
    if (statuses !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.OrderColumns.status, statuses);
    }
    if (reduceOnly !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.reduceOnly, reduceOnly);
    }
    if (orderFlags !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.orderFlags, orderFlags);
    }
    if (clientMetadata !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.clientMetadata, clientMetadata);
    }
    if (triggerPrice !== undefined) {
        baseQuery = baseQuery.where(types_1.OrderColumns.triggerPrice, triggerPrice);
    }
    // If filtering by `goodTilBlock`, filter out all rows with NULL `goodTilBlock`
    if (goodTilBlockBeforeOrAt !== undefined) {
        baseQuery = baseQuery.whereNotNull(types_1.OrderColumns.goodTilBlock).andWhere(types_1.OrderColumns.goodTilBlock, '<=', goodTilBlockBeforeOrAt);
    }
    // If filtering by `goodTilBlockTime`, filter out all rows with NULL `goodTilBlockTime`
    if (goodTilBlockTimeBeforeOrAt !== undefined) {
        baseQuery = baseQuery.whereNotNull(types_1.OrderColumns.goodTilBlockTime).andWhere(types_1.OrderColumns.goodTilBlockTime, '<=', goodTilBlockTimeBeforeOrAt);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    if (limit !== undefined) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(orderToCreate, options = { txId: undefined }) {
    return order_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        id: uuid(orderToCreate.subaccountId, orderToCreate.clientId, orderToCreate.clobPairId, orderToCreate.orderFlags),
        ...orderToCreate,
    }).returning('*');
}
exports.create = create;
function getOrderStatus(orderToUpsert, totalFilled) {
    if (orderToUpsert.status === types_1.OrderStatus.BEST_EFFORT_CANCELED) {
        return types_1.OrderStatus.BEST_EFFORT_CANCELED;
    }
    if ((0, big_js_1.default)(orderToUpsert.size).lte(totalFilled)) {
        return types_1.OrderStatus.FILLED;
    }
    return orderToUpsert.status;
}
async function update({ ...fields }, options = { txId: undefined }) {
    const order = await order_model_1.default.query(transaction_1.default.get(options.txId)).findById(fields.id).patch(fields).returning('*');
    // The objection types mistakenly think the query returns an array of orders.
    return order;
}
exports.update = update;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(order_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
async function findBySubaccountIdAndClobPair(subaccountId, clobPairId, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(order_model_1.default, options);
    const orders = await baseQuery
        .where(types_1.OrderColumns.subaccountId, subaccountId)
        .where(types_1.OrderColumns.clobPairId, clobPairId)
        .returning('*');
    return orders;
}
exports.findBySubaccountIdAndClobPair = findBySubaccountIdAndClobPair;
async function findBySubaccountIdAndClobPairAfterHeight(subaccountId, clobPairId, height, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(order_model_1.default, options);
    const orders = await baseQuery
        .where(types_1.OrderColumns.subaccountId, subaccountId)
        .where(types_1.OrderColumns.clobPairId, clobPairId)
        .where(types_1.OrderColumns.createdAtHeight, '>=', height)
        .returning('*');
    return orders;
}
exports.findBySubaccountIdAndClobPairAfterHeight = findBySubaccountIdAndClobPairAfterHeight;
async function upsert(orderToUpsert, options = { txId: undefined }) {
    const orderId = uuid(orderToUpsert.subaccountId, orderToUpsert.clientId, orderToUpsert.clobPairId, orderToUpsert.orderFlags);
    const order = await findById(orderId, options);
    if (order === undefined) {
        return create({
            ...orderToUpsert,
            status: getOrderStatus(orderToUpsert, orderToUpsert.totalFilled),
        }, options);
    }
    const updatedOrder = await update({
        ...orderToUpsert,
        status: getOrderStatus(orderToUpsert, orderToUpsert.totalFilled),
        id: orderId,
    }, options);
    if (updatedOrder === undefined) {
        throw Error('order must exist after update');
    }
    return updatedOrder;
}
exports.upsert = upsert;
/**
 * Checks if the order is a long term or conditional order.
 * @param orderFlags
 */
function isLongTermOrConditionalOrder(orderFlags) {
    const flags = parseInt(orderFlags, 10);
    const isLongTerm = Math.floor(flags / 64) % 2 === 1;
    const isConditional = Math.floor(flags / 32) % 2 === 1;
    return isLongTerm || isConditional;
}
exports.isLongTermOrConditionalOrder = isLongTermOrConditionalOrder;
/**
 * Finds all open long term or conditional orders.
 * @param options
 */
async function findOpenLongTermOrConditionalOrders(options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(order_model_1.default, options);
    /* eslint-disable */
    return baseQuery
        .where(types_1.OrderColumns.status, types_1.OrderStatus.OPEN)
        .andWhere(function () {
        this.whereRaw('FLOOR("orderFlags"::integer / 64)::integer % 2 = 1')
            .orWhereRaw('FLOOR("orderFlags"::integer / 32)::integer % 2 = 1');
    })
        .returning('*');
}
exports.findOpenLongTermOrConditionalOrders = findOpenLongTermOrConditionalOrders;
async function updateStaleOrderStatusByIds(oldStatus, newStatus, latestBlockHeight, ids, options = { txId: undefined }) {
    const updatedOrders = await order_model_1.default
        .query(transaction_1.default.get(options.txId))
        .where(types_1.OrderColumns.status, oldStatus)
        .whereNotNull(types_1.OrderColumns.goodTilBlock)
        .where(types_1.OrderColumns.goodTilBlock, '<', latestBlockHeight)
        .whereIn(types_1.OrderColumns.id, ids)
        .patch({ status: newStatus })
        .returning('*');
    return updatedOrders;
}
exports.updateStaleOrderStatusByIds = updateStaleOrderStatusByIds;
