// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/perpetual-market-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateMarketCheckerFields = exports.findByTicker = exports.findByMarketId = exports.findByClobPairId = exports.findById = exports.updateByMarketId = exports.update = exports.create = exports.findAll = void 0;
const base_1 = require("@dydxprotocol-indexer/base");
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const knex_1 = require("../helpers/knex");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const perpetual_market_model_1 = __importDefault(require("../models/perpetual-market-model"));
const types_1 = require("../types");
async function findAll({ id, marketId, liquidityTierId, limit, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        id,
        marketId,
        liquidityTierId,
        limit,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(perpetual_market_model_1.default, options);
    if (id !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.PerpetualMarketColumns.id, id);
    }
    if (marketId !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.PerpetualMarketColumns.marketId, marketId);
    }
    if (liquidityTierId !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.PerpetualMarketColumns.liquidityTierId, liquidityTierId);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.PerpetualMarketColumns.id, types_1.Ordering.ASC);
    }
    if (limit) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(perpetualMarketToCreate, options = { txId: undefined }) {
    return perpetual_market_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        ...perpetualMarketToCreate,
    }).returning('*');
}
exports.create = create;
async function update({ ...fields }, options = { txId: undefined }) {
    const perpetualMarket = await perpetual_market_model_1.default.query(transaction_1.default.get(options.txId)).findById(fields.id).patch(fields).returning('*');
    // The objection types mistakenly think the query returns an array of perpetualMarkets.
    return perpetualMarket;
}
exports.update = update;
async function updateByMarketId({ ...fields }, options = { txId: undefined }) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(perpetual_market_model_1.default, options);
    const perpetualMarkets = await baseQuery
        .where(types_1.PerpetualMarketColumns.marketId, fields.marketId)
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        .patch(fields)
        .returning('*');
    if (perpetualMarkets.length === 0) {
        return undefined;
    }
    return perpetualMarkets[0];
}
exports.updateByMarketId = updateByMarketId;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(perpetual_market_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
async function findByClobPairId(clobPairId, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(perpetual_market_model_1.default, options);
    const perpetualMarkets = await baseQuery
        .where(types_1.PerpetualMarketColumns.clobPairId, clobPairId)
        .returning('*');
    if (perpetualMarkets.length === 0) {
        return undefined;
    }
    return perpetualMarkets[0];
}
exports.findByClobPairId = findByClobPairId;
async function findByMarketId(marketId, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(perpetual_market_model_1.default, options);
    const perpetualMarkets = await baseQuery
        .where(types_1.PerpetualMarketColumns.marketId, marketId)
        .returning('*');
    if (perpetualMarkets.length === 0) {
        return undefined;
    }
    return perpetualMarkets[0];
}
exports.findByMarketId = findByMarketId;
async function findByTicker(ticker, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(perpetual_market_model_1.default, options);
    const perpetualMarkets = await baseQuery
        .where(types_1.PerpetualMarketColumns.ticker, ticker)
        .returning('*');
    if (perpetualMarkets.length === 0) {
        return undefined;
    }
    if (perpetualMarkets.length > 1) {
        base_1.logger.error({
            at: 'perpetualMarketTable#findByTicker',
            message: 'More than one market with ticker',
            ticker,
        });
    }
    return perpetualMarkets[0];
}
exports.findByTicker = findByTicker;
async function updateMarketCheckerFields(markets, transaction) {
    if (markets.length === 0) {
        return;
    }
    markets.forEach((market) => (0, stores_helpers_1.verifyAllInjectableVariables)(Object.values(market)));
    const columns = lodash_1.default.keys(markets[0]);
    const marketRows = (0, stores_helpers_1.setBulkRowsForUpdate)({
        objectArray: markets,
        columns,
        stringColumns: [
            types_1.PerpetualMarketColumns.status,
            types_1.PerpetualMarketColumns.ticker,
        ],
        numericColumns: [
            types_1.PerpetualMarketColumns.id,
            types_1.PerpetualMarketColumns.volume24H,
            types_1.PerpetualMarketColumns.trades24H,
            types_1.PerpetualMarketColumns.priceChange24H,
            types_1.PerpetualMarketColumns.openInterest,
            types_1.PerpetualMarketColumns.nextFundingRate,
        ],
    });
    const query = (0, stores_helpers_1.generateBulkUpdateString)({
        table: 'perpetual_markets',
        objectRows: marketRows,
        columns,
        isUuid: false,
        uniqueIdentifier: types_1.PerpetualMarketColumns.id,
    });
    if (transaction) {
        await knex_1.knexPrimary.raw(query).transacting(transaction);
    }
    else {
        await knex_1.knexPrimary.raw(query);
    }
}
exports.updateMarketCheckerFields = updateMarketCheckerFields;
