// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/pnl-ticks-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findMostRecentPnlTickForEachAccount = exports.findLatestProcessedBlocktime = exports.findById = exports.createMany = exports.create = exports.findAll = exports.uuid = void 0;
const lodash_1 = __importDefault(require("lodash"));
const constants_1 = require("../constants");
const knex_1 = require("../helpers/knex");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const uuid_1 = require("../helpers/uuid");
const pnl_ticks_model_1 = __importDefault(require("../models/pnl-ticks-model"));
const types_1 = require("../types");
function uuid(subaccountId, createdAt) {
    // TODO(IND-483): Fix all uuid string substitutions to use Array.join.
    return (0, uuid_1.getUuid)(Buffer.from(`${subaccountId}-${createdAt}`, constants_1.BUFFER_ENCODING_UTF_8));
}
exports.uuid = uuid;
async function findAll({ limit, id, subaccountId, createdAt, blockHeight, blockTime, createdBeforeOrAt, createdBeforeOrAtBlockHeight, createdOnOrAfter, createdOnOrAfterBlockHeight, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        limit,
        id,
        subaccountId,
        createdAt,
        blockHeight,
        blockTime,
        createdBeforeOrAt,
        createdBeforeOrAtBlockHeight,
        createdOnOrAfter,
        createdOnOrAfterBlockHeight,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(pnl_ticks_model_1.default, options);
    if (id !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.PnlTicksColumns.id, id);
    }
    if (subaccountId !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.PnlTicksColumns.subaccountId, subaccountId);
    }
    if (createdAt !== undefined) {
        baseQuery = baseQuery.where(types_1.PnlTicksColumns.createdAt, createdAt);
    }
    if (blockHeight !== undefined) {
        baseQuery = baseQuery.where(types_1.PnlTicksColumns.blockHeight, blockHeight);
    }
    if (blockTime !== undefined) {
        baseQuery = baseQuery.where(types_1.PnlTicksColumns.blockTime, blockTime);
    }
    if (createdBeforeOrAtBlockHeight !== undefined) {
        baseQuery = baseQuery.where(types_1.PnlTicksColumns.blockHeight, '<=', createdBeforeOrAtBlockHeight);
    }
    if (createdBeforeOrAt !== undefined) {
        baseQuery = baseQuery.where(types_1.PnlTicksColumns.createdAt, '<=', createdBeforeOrAt);
    }
    if (createdOnOrAfterBlockHeight !== undefined) {
        baseQuery = baseQuery.where(types_1.PnlTicksColumns.blockHeight, '>=', createdOnOrAfterBlockHeight);
    }
    if (createdOnOrAfter !== undefined) {
        baseQuery = baseQuery.where(types_1.PnlTicksColumns.createdAt, '>=', createdOnOrAfter);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.PnlTicksColumns.subaccountId, types_1.Ordering.ASC).orderBy(types_1.PnlTicksColumns.blockHeight, types_1.Ordering.DESC);
    }
    if (limit !== undefined) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(pnlTicksToCreate, options = { txId: undefined }) {
    return pnl_ticks_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        id: uuid(pnlTicksToCreate.subaccountId, pnlTicksToCreate.createdAt),
        ...pnlTicksToCreate,
    }).returning('*');
}
exports.create = create;
async function createMany(pnlTicks, options = { txId: undefined }) {
    const ticks = pnlTicks.map((tick) => ({
        ...tick,
        id: uuid(tick.subaccountId, tick.createdAt),
    }));
    return pnl_ticks_model_1.default
        .query(transaction_1.default.get(options.txId))
        .insert(ticks)
        .returning('*');
}
exports.createMany = createMany;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(pnl_ticks_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
function convertPnlTicksFromDatabaseToPnlTicksCreateObject(pnlTicksFromDatabase) {
    return lodash_1.default.omit(pnlTicksFromDatabase, types_1.PnlTicksColumns.id);
}
async function findLatestProcessedBlocktime() {
    const result = await knex_1.knexReadReplica.getConnection().raw(`
    SELECT MAX("blockTime")
    FROM "pnl_ticks"
    `);
    return result.rows[0].max || constants_1.ZERO_TIME_ISO_8601;
}
exports.findLatestProcessedBlocktime = findLatestProcessedBlocktime;
async function findMostRecentPnlTickForEachAccount(createdOnOrAfterHeight) {
    (0, stores_helpers_1.verifyAllInjectableVariables)([createdOnOrAfterHeight]);
    const result = await knex_1.knexReadReplica.getConnection().raw(`
    SELECT DISTINCT ON ("subaccountId") *
    FROM "pnl_ticks"
    WHERE "blockHeight" >= '${createdOnOrAfterHeight}'
    ORDER BY "subaccountId" ASC, "blockHeight" DESC, "createdAt" DESC;
    `);
    return lodash_1.default.keyBy(result.rows.map(convertPnlTicksFromDatabaseToPnlTicksCreateObject), 'subaccountId');
}
exports.findMostRecentPnlTickForEachAccount = findMostRecentPnlTickForEachAccount;
