// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/subaccount-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.update = exports.upsert = exports.findById = exports.create = exports.getSubaccountsWithTransfers = exports.findAll = exports.subaccountIdToUuid = exports.uuid = void 0;
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const uuid_1 = require("../helpers/uuid");
const subaccount_model_1 = __importDefault(require("../models/subaccount-model"));
const types_1 = require("../types");
function uuid(address, subaccountNumber) {
    // TODO(IND-483): Fix all uuid string substitutions to use Array.join.
    return (0, uuid_1.getUuid)(Buffer.from(`${address}-${subaccountNumber}`, constants_1.BUFFER_ENCODING_UTF_8));
}
exports.uuid = uuid;
function subaccountIdToUuid(subaccountId) {
    return uuid(subaccountId.owner, subaccountId.number);
}
exports.subaccountIdToUuid = subaccountIdToUuid;
async function findAll({ id, address, subaccountNumber, updatedBeforeOrAt, updatedOnOrAfter, limit, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        id,
        address,
        subaccountNumber,
        updatedBeforeOrAt,
        limit,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(subaccount_model_1.default, options);
    if (id) {
        baseQuery = baseQuery.whereIn(types_1.SubaccountColumns.id, id);
    }
    if (address) {
        baseQuery = baseQuery.where(types_1.SubaccountColumns.address, address);
    }
    if (subaccountNumber) {
        baseQuery = baseQuery.where(types_1.SubaccountColumns.subaccountNumber, subaccountNumber);
    }
    if (updatedBeforeOrAt) {
        baseQuery = baseQuery.where(types_1.SubaccountColumns.updatedAt, '<=', updatedBeforeOrAt);
    }
    if (updatedOnOrAfter) {
        baseQuery = baseQuery.where(types_1.SubaccountColumns.updatedAt, '>=', updatedOnOrAfter);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.SubaccountColumns.address, types_1.Ordering.ASC).orderBy(types_1.SubaccountColumns.subaccountNumber, types_1.Ordering.ASC);
    }
    if (limit) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function getSubaccountsWithTransfers(createdBeforeOrAtHeight, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const queryString = `
    SELECT *
    FROM subaccounts
    WHERE id IN (
      SELECT "senderSubaccountId" FROM transfers
      WHERE "createdAtHeight" <= '${createdBeforeOrAtHeight}'
      UNION
      SELECT "recipientSubaccountId" FROM transfers
      WHERE "createdAtHeight" <= '${createdBeforeOrAtHeight}'
    )
  `;
    const result = await (0, stores_helpers_1.rawQuery)(queryString, options);
    return result.rows;
}
exports.getSubaccountsWithTransfers = getSubaccountsWithTransfers;
async function create(subaccountToCreate, options = { txId: undefined }) {
    return subaccount_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        id: uuid(subaccountToCreate.address, subaccountToCreate.subaccountNumber),
        ...subaccountToCreate,
    }).returning('*');
}
exports.create = create;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const baseQuery = (0, stores_helpers_1.setupBaseQuery)(subaccount_model_1.default, options);
    return baseQuery
        .findById(id)
        .returning('*');
}
exports.findById = findById;
async function upsert(subaccountToUpsert, options = { txId: undefined }) {
    const { address, subaccountNumber, } = subaccountToUpsert;
    const createdUuid = uuid(address, subaccountNumber);
    const subaccounts = await subaccount_model_1.default.query(transaction_1.default.get(options.txId)).upsert({
        ...subaccountToUpsert,
        id: createdUuid,
    }).returning('*');
    // should only ever be one subaccount
    return subaccounts[0];
}
exports.upsert = upsert;
async function update({ id, ...fields }, options = { txId: undefined }) {
    const subaccount = await subaccount_model_1.default.query(transaction_1.default.get(options.txId)).findById(id);
    const updatedSubaccount = await subaccount.$query().patch(fields).returning('*');
    // The objection types mistakenly think the query returns an array of Subaccounts.
    return updatedSubaccount;
}
exports.update = update;
