// @dydxprotocol-indexer/postgres@18.19.39/build/src/stores/tendermint-event-table.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compare = exports.findById = exports.create = exports.findAll = exports.createEventId = void 0;
const big_js_1 = require("big.js");
const constants_1 = require("../constants");
const stores_helpers_1 = require("../helpers/stores-helpers");
const transaction_1 = __importDefault(require("../helpers/transaction"));
const tendermint_event_model_1 = __importDefault(require("../models/tendermint-event-model"));
const types_1 = require("../types");
const THIRTY_TWO_BITS_IN_BYTES = 4;
function createEventId(blockHeight, transactionIndex, eventIndex) {
    const buffer = Buffer.alloc(3 * THIRTY_TWO_BITS_IN_BYTES);
    buffer.writeUInt32BE(Number(blockHeight), 0);
    // transactionIndex is -2 for BEGIN_BLOCK events, and -1 for END_BLOCK events.
    // Increment by 2 to ensure result is >= 0.
    buffer.writeUInt32BE(transactionIndex + 2, THIRTY_TWO_BITS_IN_BYTES);
    buffer.writeUInt32BE(eventIndex, 2 * THIRTY_TWO_BITS_IN_BYTES);
    return buffer;
}
exports.createEventId = createEventId;
async function findAll({ id, blockHeight, transactionIndex, eventIndex, limit, }, requiredFields, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    (0, stores_helpers_1.verifyAllRequiredFields)({
        id,
        blockHeight,
        transactionIndex,
        eventIndex,
        limit,
    }, requiredFields);
    let baseQuery = (0, stores_helpers_1.setupBaseQuery)(tendermint_event_model_1.default, options);
    if (id !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.TendermintEventColumns.id, id);
    }
    if (blockHeight !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.TendermintEventColumns.blockHeight, blockHeight);
    }
    if (transactionIndex !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.TendermintEventColumns.transactionIndex, transactionIndex);
    }
    if (eventIndex !== undefined) {
        baseQuery = baseQuery.whereIn(types_1.TendermintEventColumns.eventIndex, eventIndex);
    }
    if (options.orderBy !== undefined) {
        for (const [column, order] of options.orderBy) {
            baseQuery = baseQuery.orderBy(column, order);
        }
    }
    else {
        baseQuery = baseQuery.orderBy(types_1.TendermintEventColumns.id, types_1.Ordering.ASC);
    }
    if (limit !== undefined) {
        baseQuery = baseQuery.limit(limit);
    }
    return baseQuery.returning('*');
}
exports.findAll = findAll;
async function create(tendermintEventToCreate, options = { txId: undefined }) {
    return tendermint_event_model_1.default.query(transaction_1.default.get(options.txId)).insert({
        id: createEventId(tendermintEventToCreate.blockHeight, tendermintEventToCreate.transactionIndex, tendermintEventToCreate.eventIndex),
        ...tendermintEventToCreate,
    }).returning('*');
}
exports.create = create;
async function findById(id, options = constants_1.DEFAULT_POSTGRES_OPTIONS) {
    const events = await findAll({ id: [id] }, [], options);
    if (events.length === 0) {
        return undefined;
    }
    else {
        return events[0];
    }
}
exports.findById = findById;
function compare(eventA, eventB) {
    if (eventA.blockHeight !== eventB.blockHeight) {
        return (0, big_js_1.Big)(eventA.blockHeight).minus((0, big_js_1.Big)(eventB.blockHeight)).toNumber();
    }
    if (eventA.transactionIndex !== eventB.transactionIndex) {
        return eventA.transactionIndex - eventB.transactionIndex;
    }
    if (eventA.eventIndex !== eventB.eventIndex) {
        return eventA.eventIndex - eventB.eventIndex;
    }
    return 0;
}
exports.compare = compare;
