// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/stats/stats.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserStats = exports.GlobalStats = exports.EpochStats_UserWithStats = exports.EpochStats = exports.StatsMetadata = exports.BlockStats_Fill = exports.BlockStats = void 0;
const timestamp_1 = require("../../google/protobuf/timestamp");
const _m0 = __importStar(require("protobufjs/minimal"));
const helpers_1 = require("../../helpers");
function createBaseBlockStats() {
    return {
        fills: []
    };
}
exports.BlockStats = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.fills) {
            exports.BlockStats_Fill.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBlockStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.fills.push(exports.BlockStats_Fill.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseBlockStats();
        message.fills = ((_a = object.fills) === null || _a === void 0 ? void 0 : _a.map(e => exports.BlockStats_Fill.fromPartial(e))) || [];
        return message;
    }
};
function createBaseBlockStats_Fill() {
    return {
        taker: "",
        maker: "",
        notional: helpers_1.Long.UZERO
    };
}
exports.BlockStats_Fill = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.taker !== "") {
            writer.uint32(10).string(message.taker);
        }
        if (message.maker !== "") {
            writer.uint32(18).string(message.maker);
        }
        if (!message.notional.isZero()) {
            writer.uint32(24).uint64(message.notional);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseBlockStats_Fill();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.taker = reader.string();
                    break;
                case 2:
                    message.maker = reader.string();
                    break;
                case 3:
                    message.notional = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseBlockStats_Fill();
        message.taker = (_a = object.taker) !== null && _a !== void 0 ? _a : "";
        message.maker = (_b = object.maker) !== null && _b !== void 0 ? _b : "";
        message.notional = object.notional !== undefined && object.notional !== null ? helpers_1.Long.fromValue(object.notional) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseStatsMetadata() {
    return {
        trailingEpoch: 0
    };
}
exports.StatsMetadata = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.trailingEpoch !== 0) {
            writer.uint32(8).uint32(message.trailingEpoch);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStatsMetadata();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.trailingEpoch = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseStatsMetadata();
        message.trailingEpoch = (_a = object.trailingEpoch) !== null && _a !== void 0 ? _a : 0;
        return message;
    }
};
function createBaseEpochStats() {
    return {
        epochEndTime: undefined,
        stats: []
    };
}
exports.EpochStats = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.epochEndTime !== undefined) {
            timestamp_1.Timestamp.encode((0, helpers_1.toTimestamp)(message.epochEndTime), writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.stats) {
            exports.EpochStats_UserWithStats.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEpochStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.epochEndTime = (0, helpers_1.fromTimestamp)(timestamp_1.Timestamp.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.stats.push(exports.EpochStats_UserWithStats.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseEpochStats();
        message.epochEndTime = (_a = object.epochEndTime) !== null && _a !== void 0 ? _a : undefined;
        message.stats = ((_b = object.stats) === null || _b === void 0 ? void 0 : _b.map(e => exports.EpochStats_UserWithStats.fromPartial(e))) || [];
        return message;
    }
};
function createBaseEpochStats_UserWithStats() {
    return {
        user: "",
        stats: undefined
    };
}
exports.EpochStats_UserWithStats = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.user !== "") {
            writer.uint32(10).string(message.user);
        }
        if (message.stats !== undefined) {
            exports.UserStats.encode(message.stats, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEpochStats_UserWithStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.user = reader.string();
                    break;
                case 2:
                    message.stats = exports.UserStats.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseEpochStats_UserWithStats();
        message.user = (_a = object.user) !== null && _a !== void 0 ? _a : "";
        message.stats = object.stats !== undefined && object.stats !== null ? exports.UserStats.fromPartial(object.stats) : undefined;
        return message;
    }
};
function createBaseGlobalStats() {
    return {
        notionalTraded: helpers_1.Long.UZERO
    };
}
exports.GlobalStats = {
    encode(message, writer = _m0.Writer.create()) {
        if (!message.notionalTraded.isZero()) {
            writer.uint32(8).uint64(message.notionalTraded);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGlobalStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.notionalTraded = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseGlobalStats();
        message.notionalTraded = object.notionalTraded !== undefined && object.notionalTraded !== null ? helpers_1.Long.fromValue(object.notionalTraded) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseUserStats() {
    return {
        takerNotional: helpers_1.Long.UZERO,
        makerNotional: helpers_1.Long.UZERO
    };
}
exports.UserStats = {
    encode(message, writer = _m0.Writer.create()) {
        if (!message.takerNotional.isZero()) {
            writer.uint32(8).uint64(message.takerNotional);
        }
        if (!message.makerNotional.isZero()) {
            writer.uint32(16).uint64(message.makerNotional);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUserStats();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.takerNotional = reader.uint64();
                    break;
                case 2:
                    message.makerNotional = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseUserStats();
        message.takerNotional = object.takerNotional !== undefined && object.takerNotional !== null ? helpers_1.Long.fromValue(object.takerNotional) : helpers_1.Long.UZERO;
        message.makerNotional = object.makerNotional !== undefined && object.makerNotional !== null ? helpers_1.Long.fromValue(object.makerNotional) : helpers_1.Long.UZERO;
        return message;
    }
};
