// @dydxprotocol/v4-proto/src/codegen/cosmos/rpc.tx.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRPCMsgClient = void 0;
const createRPCMsgClient = async ({ rpc }) => ({
    cosmos: {
        auth: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./auth/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        authz: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./authz/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        bank: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./bank/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        circuit: {
            v1: new (await Promise.resolve().then(() => __importStar(require("./circuit/v1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        consensus: {
            v1: new (await Promise.resolve().then(() => __importStar(require("./consensus/v1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        crisis: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./crisis/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        distribution: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./distribution/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        evidence: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./evidence/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        feegrant: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./feegrant/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        gov: {
            v1: new (await Promise.resolve().then(() => __importStar(require("./gov/v1/tx.rpc.msg")))).MsgClientImpl(rpc),
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./gov/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        group: {
            v1: new (await Promise.resolve().then(() => __importStar(require("./group/v1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        mint: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./mint/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        nft: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./nft/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        slashing: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./slashing/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        staking: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./staking/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        upgrade: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./upgrade/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        },
        vesting: {
            v1beta1: new (await Promise.resolve().then(() => __importStar(require("./vesting/v1beta1/tx.rpc.msg")))).MsgClientImpl(rpc)
        }
    }
});
exports.createRPCMsgClient = createRPCMsgClient;
