// @dydxprotocol-indexer/postgres@18.19.39/build/__tests__/helpers/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultTransaction = exports.defaultTendermintEventId4 = exports.defaultTendermintEventId3 = exports.defaultTendermintEventId2 = exports.defaultTendermintEventId = exports.defaultTendermintEvent4 = exports.defaultTendermintEvent3 = exports.defaultTendermintEvent2 = exports.defaultTendermintEvent = exports.defaultBlock2 = exports.defaultBlock = exports.isolatedMarketOrderId = exports.defaultConditionalOrderId = exports.defaultOrderGoodTilBlockTimeId = exports.defaultOrderId = exports.defaultConditionalOrder = exports.defaultOrderGoodTilBlockTime = exports.isolatedMarketOrder = exports.defaultOrder = exports.isolatedPerpetualMarket2 = exports.isolatedPerpetualMarket = exports.defaultPerpetualMarket3 = exports.defaultPerpetualMarket2 = exports.defaultPerpetualMarket = exports.isolatedSubaccountAssetPositionId = exports.isolatedSubaccountAssetPosition = exports.defaultAssetPositionId2 = exports.defaultAssetPosition2 = exports.defaultAssetPositionId = exports.defaultAssetPosition = exports.defaultAsset3 = exports.defaultAsset2 = exports.defaultAsset = exports.isolatedSubaccountId2 = exports.isolatedSubaccountId = exports.isolatedSubaccount2 = exports.isolatedSubaccount = exports.defaultSubaccountId3 = exports.defaultSubaccountId2 = exports.defaultSubaccountId = exports.defaultWalletAddress = exports.defaultSubaccount3 = exports.defaultSubaccount2 = exports.defaultSubaccount = exports.blockedAddress = exports.defaultAddress = exports.dydxChain = exports.invalidTicker = exports.createdHeight = exports.createdDateTime = void 0;
exports.nonBlockedComplianceData = exports.blockedComplianceData = exports.isolatedMarketFundingIndexUpdateId = exports.isolatedMarketFundingIndexUpdate = exports.defaultFundingIndexUpdateId = exports.defaultFundingIndexUpdate = exports.defaultPnlTick = exports.defaultCandleId = exports.defaultCandle = exports.defaultOraclePriceId2 = exports.defaultOraclePrice2 = exports.defaultOraclePriceId = exports.defaultOraclePrice = exports.defaultLiquidityTier2 = exports.defaultLiquidityTier = exports.isolatedMarket2 = exports.isolatedMarket = exports.defaultMarket3 = exports.defaultMarket2 = exports.defaultMarket = exports.defaultDepositId = exports.defaultDeposit = exports.defaultWithdrawalId = exports.defaultWithdrawal = exports.defaultTransferId = exports.defaultTransfer3 = exports.defaultTransfer2 = exports.defaultTransfer = exports.isolatedMarketFill2 = exports.isolatedMarketFill = exports.defaultFill = exports.isolatedPerpetualPositionId = exports.isolatedPerpetualPosition = exports.defaultPerpetualPositionId = exports.defaultPerpetualPosition = exports.defaultTransactionId = void 0;
const v4_proto_parser_1 = require("@dydxprotocol-indexer/v4-proto-parser");
const luxon_1 = require("luxon");
const AssetPositionTable = __importStar(require("../../src/stores/asset-position-table"));
const CandleTable = __importStar(require("../../src/stores/candle-table"));
const FundingIndexUpdatesTable = __importStar(require("../../src/stores/funding-index-updates-table"));
const OraclePriceTable = __importStar(require("../../src/stores/oracle-price-table"));
const OrderTable = __importStar(require("../../src/stores/order-table"));
const PerpetualPositionTable = __importStar(require("../../src/stores/perpetual-position-table"));
const SubaccountTable = __importStar(require("../../src/stores/subaccount-table"));
const TendermintEventTable = __importStar(require("../../src/stores/tendermint-event-table"));
const TransactionTable = __importStar(require("../../src/stores/transaction-table"));
const TransferTable = __importStar(require("../../src/stores/transfer-table"));
const types_1 = require("../../src/types");
exports.createdDateTime = luxon_1.DateTime.utc();
exports.createdHeight = '2';
exports.invalidTicker = 'INVALID-INVALID';
exports.dydxChain = 'dydx';
exports.defaultAddress = 'dydx1n88uc38xhjgxzw9nwre4ep2c8ga4fjxc565lnf';
exports.blockedAddress = 'dydx1f9k5qldwmqrnwy8hcgp4fw6heuvszt35egvtx2';
// ============== Subaccounts ==============
exports.defaultSubaccount = {
    address: exports.defaultAddress,
    subaccountNumber: 0,
    updatedAt: exports.createdDateTime.toISO(),
    updatedAtHeight: exports.createdHeight,
};
exports.defaultSubaccount2 = {
    address: exports.defaultAddress,
    subaccountNumber: 1,
    updatedAt: exports.createdDateTime.toISO(),
    updatedAtHeight: exports.createdHeight,
};
exports.defaultSubaccount3 = {
    address: exports.defaultAddress,
    subaccountNumber: 2,
    updatedAt: exports.createdDateTime.toISO(),
    updatedAtHeight: exports.createdHeight,
};
exports.defaultWalletAddress = 'defaultWalletAddress';
exports.defaultSubaccountId = SubaccountTable.uuid(exports.defaultAddress, exports.defaultSubaccount.subaccountNumber);
exports.defaultSubaccountId2 = SubaccountTable.uuid(exports.defaultAddress, exports.defaultSubaccount2.subaccountNumber);
exports.defaultSubaccountId3 = SubaccountTable.uuid(exports.defaultAddress, exports.defaultSubaccount3.subaccountNumber);
exports.isolatedSubaccount = {
    address: exports.defaultAddress,
    subaccountNumber: 128,
    updatedAt: exports.createdDateTime.toISO(),
    updatedAtHeight: exports.createdHeight,
};
exports.isolatedSubaccount2 = {
    address: exports.defaultAddress,
    subaccountNumber: 256,
    updatedAt: exports.createdDateTime.toISO(),
    updatedAtHeight: exports.createdHeight,
};
exports.isolatedSubaccountId = SubaccountTable.uuid(exports.defaultAddress, exports.isolatedSubaccount.subaccountNumber);
exports.isolatedSubaccountId2 = SubaccountTable.uuid(exports.defaultAddress, exports.isolatedSubaccount2.subaccountNumber);
// ============== Assets ==============
exports.defaultAsset = {
    id: '0',
    symbol: 'USDC',
    atomicResolution: -6,
    hasMarket: false,
};
exports.defaultAsset2 = {
    id: '1',
    symbol: 'DYDX',
    atomicResolution: 0,
    hasMarket: true,
    marketId: 1,
};
exports.defaultAsset3 = {
    id: '2',
    symbol: 'WBTC',
    atomicResolution: -8,
    hasMarket: false,
};
// ============== AssetPositions ==============
exports.defaultAssetPosition = {
    subaccountId: exports.defaultSubaccountId,
    assetId: '0',
    size: '10000',
    isLong: true,
};
exports.defaultAssetPositionId = AssetPositionTable.uuid(exports.defaultAssetPosition.subaccountId, exports.defaultAssetPosition.assetId);
exports.defaultAssetPosition2 = {
    subaccountId: exports.defaultSubaccountId2,
    assetId: '1',
    size: '10000',
    isLong: false,
};
exports.defaultAssetPositionId2 = AssetPositionTable.uuid(exports.defaultAssetPosition2.subaccountId, exports.defaultAssetPosition2.assetId);
exports.isolatedSubaccountAssetPosition = {
    subaccountId: exports.isolatedSubaccountId,
    assetId: '0',
    size: '5000',
    isLong: true,
};
exports.isolatedSubaccountAssetPositionId = AssetPositionTable.uuid(exports.isolatedSubaccountAssetPosition.subaccountId, exports.isolatedSubaccountAssetPosition.assetId);
// ============== PerpetualMarkets ==============
exports.defaultPerpetualMarket = {
    id: '0',
    clobPairId: '1',
    ticker: 'BTC-USD',
    marketId: 0,
    status: types_1.PerpetualMarketStatus.ACTIVE,
    priceChange24H: '23',
    volume24H: '1000000',
    trades24H: 250,
    nextFundingRate: '10.2',
    openInterest: '400000',
    quantumConversionExponent: -8,
    atomicResolution: -10,
    subticksPerTick: 100,
    stepBaseQuantums: 10,
    liquidityTierId: 0,
    marketType: types_1.PerpetualMarketType.CROSS,
};
exports.defaultPerpetualMarket2 = {
    id: '1',
    clobPairId: '2',
    ticker: 'ETH-USD',
    marketId: 1,
    status: types_1.PerpetualMarketStatus.ACTIVE,
    priceChange24H: '23',
    volume24H: '100000',
    trades24H: 200,
    nextFundingRate: '1.2',
    openInterest: '40000',
    quantumConversionExponent: -6,
    atomicResolution: -18,
    subticksPerTick: 10,
    stepBaseQuantums: 1,
    liquidityTierId: 0,
    marketType: types_1.PerpetualMarketType.CROSS,
};
exports.defaultPerpetualMarket3 = {
    id: '2',
    clobPairId: '3',
    ticker: 'SHIB-USD',
    marketId: 2,
    status: types_1.PerpetualMarketStatus.ACTIVE,
    priceChange24H: '0.000000001',
    volume24H: '10000000',
    trades24H: 200,
    nextFundingRate: '1.2',
    openInterest: '40000',
    quantumConversionExponent: -16,
    atomicResolution: -2,
    subticksPerTick: 10,
    stepBaseQuantums: 1,
    liquidityTierId: 0,
    marketType: types_1.PerpetualMarketType.CROSS,
};
exports.isolatedPerpetualMarket = {
    id: '3',
    clobPairId: '4',
    ticker: 'ISO-USD',
    marketId: 3,
    status: types_1.PerpetualMarketStatus.ACTIVE,
    priceChange24H: '0.000000001',
    volume24H: '10000000',
    trades24H: 200,
    nextFundingRate: '1.2',
    openInterest: '40000',
    quantumConversionExponent: -16,
    atomicResolution: -2,
    subticksPerTick: 10,
    stepBaseQuantums: 1,
    liquidityTierId: 0,
    marketType: types_1.PerpetualMarketType.ISOLATED,
};
exports.isolatedPerpetualMarket2 = {
    id: '4',
    clobPairId: '5',
    ticker: 'ISO2-USD',
    marketId: 4,
    status: types_1.PerpetualMarketStatus.ACTIVE,
    priceChange24H: '0.000000001',
    volume24H: '10000000',
    trades24H: 200,
    nextFundingRate: '1.2',
    openInterest: '40000',
    quantumConversionExponent: -16,
    atomicResolution: -2,
    subticksPerTick: 10,
    stepBaseQuantums: 1,
    liquidityTierId: 0,
    marketType: types_1.PerpetualMarketType.ISOLATED,
};
// ============== Orders ==============
exports.defaultOrder = {
    subaccountId: exports.defaultSubaccountId,
    clientId: '1',
    clobPairId: '1',
    side: types_1.OrderSide.BUY,
    size: '25',
    totalFilled: '0',
    price: '20000',
    type: types_1.OrderType.LIMIT,
    status: types_1.OrderStatus.OPEN,
    timeInForce: types_1.TimeInForce.FOK,
    reduceOnly: false,
    goodTilBlock: '100',
    orderFlags: v4_proto_parser_1.ORDER_FLAG_SHORT_TERM.toString(),
    clientMetadata: '0',
    updatedAt: '2023-01-22T00:00:00.000Z',
    updatedAtHeight: '1',
};
exports.isolatedMarketOrder = {
    subaccountId: exports.isolatedSubaccountId,
    clientId: '1',
    clobPairId: '4',
    side: types_1.OrderSide.BUY,
    size: '25',
    totalFilled: '0',
    price: '20000',
    type: types_1.OrderType.LIMIT,
    status: types_1.OrderStatus.OPEN,
    timeInForce: types_1.TimeInForce.FOK,
    reduceOnly: false,
    goodTilBlock: '100',
    orderFlags: v4_proto_parser_1.ORDER_FLAG_SHORT_TERM.toString(),
    clientMetadata: '0',
    updatedAt: '2023-01-22T00:00:00.000Z',
    updatedAtHeight: '1',
};
exports.defaultOrderGoodTilBlockTime = {
    ...exports.defaultOrder,
    clientId: '2',
    goodTilBlock: undefined,
    goodTilBlockTime: '2023-01-22T00:00:00.000Z',
    createdAtHeight: '1',
    orderFlags: v4_proto_parser_1.ORDER_FLAG_LONG_TERM.toString(),
};
exports.defaultConditionalOrder = {
    ...exports.defaultOrderGoodTilBlockTime,
    type: types_1.OrderType.STOP_LIMIT,
    clientId: '3',
    orderFlags: v4_proto_parser_1.ORDER_FLAG_CONDITIONAL.toString(),
    triggerPrice: '19000',
};
exports.defaultOrderId = OrderTable.uuid(exports.defaultOrder.subaccountId, exports.defaultOrder.clientId, exports.defaultOrder.clobPairId, exports.defaultOrder.orderFlags);
exports.defaultOrderGoodTilBlockTimeId = OrderTable.uuid(exports.defaultOrderGoodTilBlockTime.subaccountId, exports.defaultOrderGoodTilBlockTime.clientId, exports.defaultOrderGoodTilBlockTime.clobPairId, exports.defaultOrderGoodTilBlockTime.orderFlags);
exports.defaultConditionalOrderId = OrderTable.uuid(exports.defaultConditionalOrder.subaccountId, exports.defaultConditionalOrder.clientId, exports.defaultConditionalOrder.clobPairId, exports.defaultConditionalOrder.orderFlags);
exports.isolatedMarketOrderId = OrderTable.uuid(exports.isolatedMarketOrder.subaccountId, exports.isolatedMarketOrder.clientId, exports.isolatedMarketOrder.clobPairId, exports.isolatedMarketOrder.orderFlags);
// ============== Blocks ==============
exports.defaultBlock = {
    blockHeight: '1',
    time: luxon_1.DateTime.utc(2022, 6, 1).toISO(),
};
exports.defaultBlock2 = {
    blockHeight: '2',
    time: luxon_1.DateTime.utc(2022, 6, 2).toISO(),
};
// ============== TendermintEvents ==============
exports.defaultTendermintEvent = {
    blockHeight: '1',
    transactionIndex: -1,
    eventIndex: 0,
};
exports.defaultTendermintEvent2 = {
    blockHeight: '2',
    transactionIndex: -1,
    eventIndex: 0,
};
exports.defaultTendermintEvent3 = {
    blockHeight: '2',
    transactionIndex: 0,
    eventIndex: 0,
};
exports.defaultTendermintEvent4 = {
    blockHeight: '2',
    transactionIndex: 1,
    eventIndex: 1,
};
exports.defaultTendermintEventId = TendermintEventTable.createEventId(exports.defaultTendermintEvent.blockHeight, exports.defaultTendermintEvent.transactionIndex, exports.defaultTendermintEvent.eventIndex);
exports.defaultTendermintEventId2 = TendermintEventTable.createEventId(exports.defaultTendermintEvent2.blockHeight, exports.defaultTendermintEvent2.transactionIndex, exports.defaultTendermintEvent2.eventIndex);
exports.defaultTendermintEventId3 = TendermintEventTable.createEventId(exports.defaultTendermintEvent3.blockHeight, exports.defaultTendermintEvent3.transactionIndex, exports.defaultTendermintEvent3.eventIndex);
exports.defaultTendermintEventId4 = TendermintEventTable.createEventId(exports.defaultTendermintEvent4.blockHeight, exports.defaultTendermintEvent4.transactionIndex, exports.defaultTendermintEvent4.eventIndex);
// ============== Transactions ==============
exports.defaultTransaction = {
    blockHeight: '1',
    transactionIndex: 0,
    transactionHash: '3ac776f8-1900-43de-ac38-7fb516f7d6d0',
};
exports.defaultTransactionId = TransactionTable.uuid(exports.defaultTransaction.blockHeight, exports.defaultTransaction.transactionIndex);
// ============== PerpetualPositions ==============
exports.defaultPerpetualPosition = {
    subaccountId: exports.defaultSubaccountId,
    perpetualId: exports.defaultPerpetualMarket.id,
    side: types_1.PositionSide.LONG,
    status: types_1.PerpetualPositionStatus.OPEN,
    size: '10',
    maxSize: '25',
    entryPrice: '20000',
    sumOpen: '10',
    sumClose: '0',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
    openEventId: exports.defaultTendermintEventId,
    lastEventId: exports.defaultTendermintEventId2,
    settledFunding: '200000',
};
exports.defaultPerpetualPositionId = PerpetualPositionTable.uuid(exports.defaultPerpetualPosition.subaccountId, exports.defaultPerpetualPosition.openEventId);
exports.isolatedPerpetualPosition = {
    subaccountId: exports.isolatedSubaccountId,
    perpetualId: exports.isolatedPerpetualMarket.id,
    side: types_1.PositionSide.LONG,
    status: types_1.PerpetualPositionStatus.OPEN,
    size: '10',
    maxSize: '25',
    entryPrice: '20000',
    sumOpen: '10',
    sumClose: '0',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
    openEventId: exports.defaultTendermintEventId,
    lastEventId: exports.defaultTendermintEventId2,
    settledFunding: '200000',
};
exports.isolatedPerpetualPositionId = PerpetualPositionTable.uuid(exports.isolatedPerpetualPosition.subaccountId, exports.isolatedPerpetualPosition.openEventId);
// ============== Fills ==============
exports.defaultFill = {
    subaccountId: exports.defaultSubaccountId,
    side: types_1.OrderSide.BUY,
    liquidity: types_1.Liquidity.TAKER,
    type: types_1.FillType.LIMIT,
    clobPairId: '1',
    orderId: exports.defaultOrderId,
    size: '10',
    price: '20000',
    quoteAmount: '200000',
    eventId: exports.defaultTendermintEventId,
    transactionHash: '',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
    clientMetadata: '0',
    fee: '1.1',
};
exports.isolatedMarketFill = {
    subaccountId: exports.isolatedSubaccountId,
    side: types_1.OrderSide.BUY,
    liquidity: types_1.Liquidity.TAKER,
    type: types_1.FillType.LIMIT,
    clobPairId: '4',
    orderId: exports.isolatedMarketOrderId,
    size: '10',
    price: '20000',
    quoteAmount: '200000',
    eventId: exports.defaultTendermintEventId2,
    transactionHash: '',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
    clientMetadata: '0',
    fee: '1.1',
};
exports.isolatedMarketFill2 = {
    subaccountId: exports.isolatedSubaccountId2,
    side: types_1.OrderSide.BUY,
    liquidity: types_1.Liquidity.TAKER,
    type: types_1.FillType.LIMIT,
    clobPairId: '4',
    orderId: exports.isolatedMarketOrderId,
    size: '10',
    price: '20000',
    quoteAmount: '200000',
    eventId: exports.defaultTendermintEventId3,
    transactionHash: '',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
    clientMetadata: '0',
    fee: '1.1',
};
// ============== Transfers ==============
exports.defaultTransfer = {
    senderSubaccountId: exports.defaultSubaccountId,
    recipientSubaccountId: exports.defaultSubaccountId2,
    assetId: exports.defaultAsset.id,
    size: '10',
    eventId: exports.defaultTendermintEventId,
    transactionHash: '',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
};
exports.defaultTransfer2 = {
    ...exports.defaultTransfer,
    senderSubaccountId: exports.defaultSubaccountId3,
    size: '5',
};
exports.defaultTransfer3 = {
    ...exports.defaultTransfer2,
    assetId: exports.defaultAsset2.id,
};
exports.defaultTransferId = TransferTable.uuid(exports.defaultTransfer.eventId, exports.defaultTransfer.assetId, exports.defaultTransfer.senderSubaccountId, exports.defaultTransfer.recipientSubaccountId, exports.defaultTransfer.senderWalletAddress, exports.defaultTransfer.recipientWalletAddress);
exports.defaultWithdrawal = {
    senderSubaccountId: exports.defaultSubaccountId,
    recipientWalletAddress: exports.defaultWalletAddress,
    assetId: exports.defaultAsset.id,
    size: '10',
    eventId: exports.defaultTendermintEventId,
    transactionHash: '',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
};
exports.defaultWithdrawalId = TransferTable.uuid(exports.defaultWithdrawal.eventId, exports.defaultWithdrawal.assetId, exports.defaultWithdrawal.senderSubaccountId, exports.defaultWithdrawal.recipientSubaccountId, exports.defaultWithdrawal.senderWalletAddress, exports.defaultWithdrawal.recipientWalletAddress);
exports.defaultDeposit = {
    senderWalletAddress: exports.defaultWalletAddress,
    recipientSubaccountId: exports.defaultSubaccountId,
    assetId: exports.defaultAsset.id,
    size: '10',
    eventId: exports.defaultTendermintEventId,
    transactionHash: '',
    createdAt: exports.createdDateTime.toISO(),
    createdAtHeight: exports.createdHeight,
};
exports.defaultDepositId = TransferTable.uuid(exports.defaultDeposit.eventId, exports.defaultDeposit.assetId, exports.defaultDeposit.senderSubaccountId, exports.defaultDeposit.recipientSubaccountId, exports.defaultDeposit.senderWalletAddress, exports.defaultDeposit.recipientWalletAddress);
// ============== Markets ==============
exports.defaultMarket = {
    id: 0,
    pair: 'BTC-USD',
    exponent: -5,
    minPriceChangePpm: 50,
    oraclePrice: '15000',
};
exports.defaultMarket2 = {
    id: 1,
    pair: 'ETH-USD',
    exponent: -6,
    minPriceChangePpm: 50,
    oraclePrice: '1000',
};
exports.defaultMarket3 = {
    id: 2,
    pair: 'SHIB-USD',
    exponent: -12,
    minPriceChangePpm: 50,
    oraclePrice: '0.000000065',
};
exports.isolatedMarket = {
    id: 3,
    pair: 'ISO-USD',
    exponent: -12,
    minPriceChangePpm: 50,
    oraclePrice: '0.000000075',
};
exports.isolatedMarket2 = {
    id: 4,
    pair: 'ISO2-USD',
    exponent: -12,
    minPriceChangePpm: 50,
    oraclePrice: '0.000000085',
};
// ============== LiquidityTiers ==============
exports.defaultLiquidityTier = {
    id: 0,
    name: 'Large-Cap',
    initialMarginPpm: '50000',
    maintenanceFractionPpm: '600000', // 60%
};
exports.defaultLiquidityTier2 = {
    id: 1,
    name: 'Mid-Cap',
    initialMarginPpm: '100000',
    maintenanceFractionPpm: '500000', // 50%
};
// ============== OraclePrices ==============
exports.defaultOraclePrice = {
    marketId: exports.defaultMarket.id,
    price: '10000',
    effectiveAt: exports.createdDateTime.toISO(),
    effectiveAtHeight: exports.createdHeight,
};
exports.defaultOraclePriceId = OraclePriceTable.uuid(exports.defaultOraclePrice.marketId, exports.defaultOraclePrice.effectiveAtHeight);
exports.defaultOraclePrice2 = {
    marketId: exports.defaultMarket2.id,
    price: '500',
    effectiveAt: exports.createdDateTime.toISO(),
    effectiveAtHeight: exports.createdHeight,
};
exports.defaultOraclePriceId2 = OraclePriceTable.uuid(exports.defaultOraclePrice2.marketId, exports.defaultOraclePrice2.effectiveAtHeight);
// ============== Candle ==============
exports.defaultCandle = {
    startedAt: exports.createdDateTime.toISO(),
    ticker: exports.defaultPerpetualMarket.ticker,
    resolution: types_1.CandleResolution.ONE_MINUTE,
    low: '10000',
    high: '12000',
    open: '11000',
    close: '11500',
    baseTokenVolume: '400000',
    usdVolume: '2200000',
    trades: 300,
    startingOpenInterest: '200000',
};
exports.defaultCandleId = CandleTable.uuid(exports.defaultCandle.startedAt, exports.defaultCandle.ticker, exports.defaultCandle.resolution);
// ============== Pnl Ticks ==============
exports.defaultPnlTick = {
    subaccountId: exports.defaultSubaccountId,
    equity: '100000',
    totalPnl: '10000',
    netTransfers: '1000',
    createdAt: exports.createdDateTime.toISO(),
    blockHeight: exports.createdHeight,
    blockTime: exports.defaultBlock2.time,
};
// ========= Funding Index updates ==========
exports.defaultFundingIndexUpdate = {
    perpetualId: exports.defaultPerpetualMarket.id,
    eventId: exports.defaultTendermintEventId,
    rate: '0.0004',
    oraclePrice: '10000',
    fundingIndex: '10050',
    effectiveAt: exports.createdDateTime.toISO(),
    effectiveAtHeight: exports.createdHeight,
};
exports.defaultFundingIndexUpdateId = FundingIndexUpdatesTable.uuid(exports.defaultFundingIndexUpdate.effectiveAtHeight, exports.defaultFundingIndexUpdate.eventId, exports.defaultFundingIndexUpdate.perpetualId);
exports.isolatedMarketFundingIndexUpdate = {
    perpetualId: exports.isolatedPerpetualMarket.id,
    eventId: exports.defaultTendermintEventId,
    rate: '0.0004',
    oraclePrice: '10000',
    fundingIndex: '10200',
    effectiveAt: exports.createdDateTime.toISO(),
    effectiveAtHeight: exports.createdHeight,
};
exports.isolatedMarketFundingIndexUpdateId = FundingIndexUpdatesTable.uuid(exports.isolatedMarketFundingIndexUpdate.effectiveAtHeight, exports.isolatedMarketFundingIndexUpdate.eventId, exports.isolatedMarketFundingIndexUpdate.perpetualId);
// ========= Compliance Data ==========
exports.blockedComplianceData = {
    address: exports.blockedAddress,
    provider: types_1.ComplianceProvider.ELLIPTIC,
    chain: exports.dydxChain,
    blocked: true,
    riskScore: '100.00',
    updatedAt: exports.createdDateTime.toISO(),
};
exports.nonBlockedComplianceData = {
    address: exports.defaultAddress,
    provider: types_1.ComplianceProvider.ELLIPTIC,
    chain: exports.dydxChain,
    blocked: false,
    riskScore: '10.00',
    updatedAt: exports.createdDateTime.plus(1).toISO(),
};
