// @osmonauts/ast@0.68.0/main/registry/registry.spec.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mutations = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _generator = _interopRequireDefault(require("@babel/generator"));

var _registry = require("./registry");

var mutations = [{
  methodName: 'joinPool',
  typeUrl: '/cosmos.pools.transfer.v1.MsgJoinPool',
  TypeName: 'MsgJoinPool'
}, {
  methodName: 'exitPool',
  typeUrl: '/cosmos.pools.transfer.v1.MsgExitPool',
  TypeName: 'MsgExitPool'
}];
exports.mutations = mutations;
var context = {
  addUtil: function addUtil() {}
};

var expectCode = function expectCode(ast) {
  expect((0, _generator["default"])(ast).code).toMatchSnapshot();
};

var printCode = function printCode(ast) {
  console.log((0, _generator["default"])(ast).code);
};

it('createTypeRegistry', /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
  return _regenerator["default"].wrap(function _callee$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          expectCode((0, _registry.createTypeRegistry)(context, mutations));

        case 1:
        case "end":
          return _context.stop();
      }
    }
  }, _callee);
})));
it('createRegistryLoader', /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
  return _regenerator["default"].wrap(function _callee2$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          expectCode((0, _registry.createRegistryLoader)(context));

        case 1:
        case "end":
          return _context2.stop();
      }
    }
  }, _callee2);
})));
