// @osmonauts/telescope@0.80.0/main/generators/create-rpc-query-client-scoped.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.plugin = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var dotty = _interopRequireWildcard(require("dotty"));

var _protoParser = require("@osmonauts/proto-parser");

var _path = require("path");

var _ast = require("@osmonauts/ast");

var _utils = require("../utils");

var _build = require("../build");

var _imports = require("../imports");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var plugin = function plugin(builder, bundler) {
  var _builder$options, _builder$options$rpcC, _builder$options2, _builder$options2$rpc;

  // if not enabled, exit
  if (!((_builder$options = builder.options) !== null && _builder$options !== void 0 && (_builder$options$rpcC = _builder$options.rpcClients) !== null && _builder$options$rpcC !== void 0 && _builder$options$rpcC.enabled)) {
    return;
  }

  if ((_builder$options2 = builder.options) !== null && _builder$options2 !== void 0 && (_builder$options2$rpc = _builder$options2.rpcClients) !== null && _builder$options2$rpc !== void 0 && _builder$options2$rpc.inline) {
    return;
  } // if no scopes, do them all!


  if (!builder.options.rpcClients.scoped || !builder.options.rpcClients.scoped.length) {
    // TODO inefficient
    // WE SHOULD NOT DO THIS IN A BUNDLER LOOP
    // MAKE SEPARATE PLUGIN
    return makeAllRPCBundles(builder, bundler);
  }

  if (!builder.options.rpcClients.scopedIsExclusive) {
    // TODO inefficient
    // WE SHOULD NOT DO THIS IN A BUNDLER LOOP
    // MAKE SEPARATE PLUGIN
    makeAllRPCBundles(builder, bundler);
  } // we have scopes!


  builder.options.rpcClients.scoped.forEach(function (rpc) {
    if (rpc.dir !== bundler.bundle.base) return;
    makeRPC(builder, bundler, rpc);
  });
};

exports.plugin = plugin;

var getFileName = function getFileName(dir, filename) {
  filename = filename.replace(/\.ts$/, '');
  var localname = (0, _path.join)(dir, filename + '.query');
  return localname + '.ts';
};

var makeRPC = function makeRPC(builder, bundler, rpc) {
  var _rpc$methodNameQuery, _rpc$filename;

  var dir = rpc.dir;
  var packages = rpc.packages;
  var methodName = (_rpc$methodNameQuery = rpc.methodNameQuery) !== null && _rpc$methodNameQuery !== void 0 ? _rpc$methodNameQuery : 'createRPCQueryClient';
  var localname = getFileName(dir, (_rpc$filename = rpc.filename) !== null && _rpc$filename !== void 0 ? _rpc$filename : 'rpc');
  var obj = {};
  builder.rpcQueryClients.forEach(function (file) {
    // ADD all option
    // which defaults to including cosmos 
    // and defaults to base for each
    if (!packages.includes(file["package"])) {
      return;
    }

    var f = localname;
    var f2 = file.localname;
    var importPath = (0, _utils.getRelativePath)(f, f2);
    dotty.put(obj, file["package"], importPath);
  });
  var ctx = new _build.TelescopeParseContext({
    absolute: '',
    filename: localname,
    proto: {
      "package": dir,
      imports: null,
      root: {},
      importNames: null
    },
    traversed: {
      "package": dir,
      imports: null,
      root: {},
      importNames: null
    }
  }, builder.store, builder.options); // TODO add addUtil to generic context

  ctx.proto.addUtil('Rpc');
  var rpcast = (0, _ast.createScopedRpcTmFactory)(ctx.proto, obj, methodName // 'QueryClientImpl' // make option later
  );
  var serviceImports = (0, _imports.getDepsFromQueries)(ctx.queries, localname);
  var imports = (0, _imports.aggregateImports)(ctx, serviceImports, localname);
  var importStmts = (0, _imports.getImportStatements)(localname, (0, _toConsumableArray2["default"])((0, _utils.fixlocalpaths)(imports)));
  var prog = [].concat(importStmts).concat(rpcast);
  var filename = bundler.getFilename(localname);
  bundler.writeAst(prog, filename);

  if (rpc.addToBundle) {
    bundler.addToBundleToPackage("".concat(dir, ".ClientFactory"), localname);
  }
}; // TODO

/*
 move all options for rpc into previous `rpc` prop and 
 clean up all these many options for one nested object full of options
*/


var makeAllRPCBundles = function makeAllRPCBundles(builder, bundler) {
  if (!builder.options.rpcClients.bundle) return; // [x] loop through every bundle 
  // [x] if not cosmos, add all cosmos
  // [x] call makeRPC

  var dir = bundler.bundle.base;
  var filename = 'rpc'; ///
  ///
  ///
  // refs with services

  var refs = builder.store.getProtos().filter(function (ref) {
    var proto = (0, _protoParser.getNestedProto)(ref.traversed); //// Anything except Msg Service OK...

    var allowedRpcServices = builder.options.rpcClients.enabledServices.filter(function (a) {
      return a !== 'Msg';
    });
    var found = allowedRpcServices.some(function (svc) {
      var _proto$svc;

      return (proto === null || proto === void 0 ? void 0 : proto[svc]) && ((_proto$svc = proto[svc]) === null || _proto$svc === void 0 ? void 0 : _proto$svc.type) === 'Service';
    });

    if (!found) {
      return;
    } ///


    return true;
  });
  var check = refs.filter(function (ref) {
    var _ref$proto$package$sp = ref.proto["package"].split('.'),
        _ref$proto$package$sp2 = (0, _slicedToArray2["default"])(_ref$proto$package$sp, 1),
        base = _ref$proto$package$sp2[0];

    return base === bundler.bundle.base;
  });

  if (!check.length) {
    // if there are no services
    // exit the plugin
    return;
  }

  var packages = refs.reduce(function (m, ref) {
    var _ref$proto$package$sp3 = ref.proto["package"].split('.'),
        _ref$proto$package$sp4 = (0, _slicedToArray2["default"])(_ref$proto$package$sp3, 1),
        base = _ref$proto$package$sp4[0];

    if (base === 'cosmos' || base === bundler.bundle.base) return (0, _toConsumableArray2["default"])(new Set([].concat((0, _toConsumableArray2["default"])(m), [ref.proto["package"]])));
    return m;
  }, []);
  makeRPC(builder, bundler, {
    dir: dir,
    filename: filename,
    packages: packages,
    addToBundle: true,
    methodNameQuery: 'createRPCQueryClient'
  });
};
