// @dydxprotocol/v4-proto/src/codegen/dydxprotocol/blocktime/query.rpc.Query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRpcQueryExtension = exports.QueryClientImpl = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
const stargate_1 = require("@cosmjs/stargate");
const query_1 = require("./query");
class QueryClientImpl {
    constructor(rpc) {
        this.rpc = rpc;
        this.downtimeParams = this.downtimeParams.bind(this);
        this.previousBlockInfo = this.previousBlockInfo.bind(this);
        this.allDowntimeInfo = this.allDowntimeInfo.bind(this);
    }
    downtimeParams(request = {}) {
        const data = query_1.QueryDowntimeParamsRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.blocktime.Query", "DowntimeParams", data);
        return promise.then(data => query_1.QueryDowntimeParamsResponse.decode(new _m0.Reader(data)));
    }
    previousBlockInfo(request = {}) {
        const data = query_1.QueryPreviousBlockInfoRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.blocktime.Query", "PreviousBlockInfo", data);
        return promise.then(data => query_1.QueryPreviousBlockInfoResponse.decode(new _m0.Reader(data)));
    }
    allDowntimeInfo(request = {}) {
        const data = query_1.QueryAllDowntimeInfoRequest.encode(request).finish();
        const promise = this.rpc.request("dydxprotocol.blocktime.Query", "AllDowntimeInfo", data);
        return promise.then(data => query_1.QueryAllDowntimeInfoResponse.decode(new _m0.Reader(data)));
    }
}
exports.QueryClientImpl = QueryClientImpl;
const createRpcQueryExtension = (base) => {
    const rpc = (0, stargate_1.createProtobufRpcClient)(base);
    const queryService = new QueryClientImpl(rpc);
    return {
        downtimeParams(request) {
            return queryService.downtimeParams(request);
        },
        previousBlockInfo(request) {
            return queryService.previousBlockInfo(request);
        },
        allDowntimeInfo(request) {
            return queryService.allDowntimeInfo(request);
        }
    };
};
exports.createRpcQueryExtension = createRpcQueryExtension;
